---
title: "1-diff-express-analysis"
author: "Laura H Spencer"
date: "11/17/2025"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
      number_sections: true
---

This notebook contains all gene expression analyses for the juvenile snow crab study, in which they were exposed to OA treatments (pH 7.8, 7.5) and sampled at 8 hr and 88 days. There was also a time-0 control at ambient pH (8.0). 

I aligned to the snow crab draft genome with STAR and used their annotation file to get gene counts.  I also had experimented with the tanner genome, but alignment rate was much lower, so opted with the snow genome.  

```{r include = FALSE}
knitr::opts_chunk$set(echo=TRUE, warning = F, message = F)
```

### Load libraries and source scripts 

```{r, message=FALSE, warning=FALSE, results=FALSE}
source("../references/biostats.R")
source("../references/iLOO.R")
`%!in%` = Negate(`%in%`)

# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("RCurl", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "cowplot", "clipr", "janitor", "ggpubr", "forcats", "apeglm", "car", "vsn", "devtools", "grid", "gridGraphics", "Rfast", "dendextend", "RColorBrewer", "scales", "VennDiagram", "colorspace", "edgeR", "devtools", "ggpattern", "eulerr", "Rmisc", "pairwiseAdonis", "plotly", "tximport", "GenomicFeatures",  "readr", "readxl", "limma", "ggfortify", "ggrepel", "matrixStats", "ggvenn", "ashr", "truncnorm", "ape", "purrr", "patchwork", "tidyverse")

# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("DESeq2", "WGCNA", "sva", "clusterProfiler", "apeglm", "EnhancedVolcano")

# Install BiocManager if needed
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]

# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

all.packages <- c(list.of.packages, bioconductor.packages)

# Load all required libraries
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})

# # Github packages 
# install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
```

### Obtain session information

```{r}
sessionInfo()
getwd()
```

## ANNOTATION: Using snow crab genome 

Here I pull gene annotations from the snow crab coding sequence fasta headers, which I saved to file GCA_016584305.1_ASM1658430v1_cds_headers.txt. I then augment them with DIAMOND Blast results, filling in putative functions for genes that were not annotated. 

#### First get gene ID -> GO term maps 

<!-- ```{bash} -->
<!-- wget https://current.geneontology.org/ontology/external2go/interpro2go -->
<!-- wget https://current.geneontology.org/ontology/external2go/pfam2go -->
<!-- ``` -->

#### Pull header lines with key fields
Format: gene_id  protein_id  token
token is like GO:0004930 or InterPro:IPR006026 or PFAM:PF00075 or CDD:cd04280

<!-- ```{bash} -->
<!-- # 1a) Pull header lines → tokens -->
<!-- awk -F'[][]' ' -->
<!--   /^>/{ -->
<!--     hdr=$0 -->
<!--     sub(/^>/,"",hdr) -->

<!--     # locus_tag (gene key) -->
<!--     gene="" -->
<!--     if (match(hdr, /locus_tag=([^]]*)/, a)) gene=a[1] -->

<!--     # optional protein_id -->
<!--     prot="" -->
<!--     if (match(hdr, /protein_id=([^]]*)/, b)) prot=b[1] -->

<!--     # Split on any [db_xref=...] -->
<!--     n=split(hdr, arr, /\[db_xref=/) -->
<!--     if (n>1){ -->
<!--       for (i=2; i<=n; i++){ -->
<!--         split(arr[i], right, "]")             # take content up to the closing ] -->
<!--         split(right[1], tok, ",")             # tokens are comma-separated -->
<!--         for (j in tok){ -->
<!--           if (tok[j] != ""){ -->
<!--             printf("%s\t%s\t%s\n", gene, prot, tok[j]) -->
<!--           } -->
<!--         } -->
<!--       } -->
<!--     } -->
<!--   } -->
<!-- ' ../references/GCA_016584305.1_ASM1658430v1_cds_headers.txt > ../references/snow_cds_tokens.tsv -->

<!-- # (Optional) keep only explicit GO tokens (directly present in headers) -->
<!-- awk '$3 ~ /^GO:/' ../references/snow_cds_tokens.tsv > ../references/snow_cds_GO.tsv -->
<!-- ``` -->

#### Build TERM2GENE / TERM2NAME from headers + external2go maps

```{r}
tokens <- read_tsv("../references/snow_cds_GO.tsv",
                   col_names = c("gene_id","protein_id","token"),
                   col_types = "ccc")

tokens <- tokens %>%
  mutate(type = case_when(
           str_starts(token, "GO:") ~ "GO",
           str_starts(token, "InterPro:") ~ "InterPro",
           str_starts(token, "PFAM:") ~ "PFAM",
           TRUE ~ "OTHER"),
         id = case_when(
           type=="GO" ~ token,
           type=="InterPro" ~ str_remove(token, "^InterPro:"),
           type=="PFAM" ~ str_remove(token, "^PFAM:"),
           TRUE ~ token))

# ---- parse interpro2go / pfam2go (each line contains external id + GO:xxxxxx tokens)
parse_map <- function(path, pattern_id){
  x <- read_lines(path)
  tibble(raw = x) %>%
    mutate(go = str_extract_all(raw, "GO:\\d+"),
           external = str_match(raw, pattern_id)[,2]) %>%
    filter(!is.na(external), lengths(go) > 0) %>%
    unnest(go) %>%
    distinct(external, go)
}

map_interpro <- parse_map("../references/interpro2go", "InterPro:(IPR\\d+)")
map_pfam     <- parse_map("../references/pfam2go",     "Pfam:(PF\\d+)")

# Direct GO from headers
gene_go_direct <- tokens %>%
  filter(type=="GO") %>%
  transmute(gene_id, GO = id)

# InterPro → GO
gene_go_ipr <- tokens %>%
  filter(type=="InterPro") %>%
  inner_join(map_interpro, by = c("id" = "external")) %>%
  transmute(gene_id, GO = go)

# Pfam → GO
gene_go_pfam <- tokens %>%
  filter(type=="PFAM") %>%
  inner_join(map_pfam, by = c("id" = "external")) %>%
  transmute(gene_id, GO = go)

# Union (deduplicate)
gene2go_all <- bind_rows(gene_go_direct, gene_go_ipr, gene_go_pfam) %>%
  distinct()

# Optional: add GO term names for nicer outputs
suppressPackageStartupMessages({
  if (!requireNamespace("GO.db", quietly = TRUE)) {
    stop("Please install GO.db: BiocManager::install('GO.db')")
  }
  library(GO.db)
})
go_names <- AnnotationDbi::select(GO.db,
                                  keys = unique(gene2go_all$GO),
                                  keytype = "GOID",
                                  columns = "TERM") %>%
  as_tibble() %>%
  dplyr::rename(GO = GOID, term_name = TERM) %>%
  distinct()

TERM2GENE <- gene2go_all %>% dplyr::select(GO, gene_id)
TERM2NAME <- go_names

message("Genes with any GO (after InterPro/Pfam expansion): ",
        n_distinct(TERM2GENE$gene_id), " / total genes in headers: ",
        n_distinct(tokens$gene_id))

# Save for posterity! 
write_tsv(gene2go_all, "../references/gene2go_all.tsv")
write_tsv(go_names,    "../references/go_term_names.tsv")
```

#### Read in the annotation, add DIAMOND Blast & GO terms

```{r}
snow.annot <- read_delim(file = "../references/GCA_016584305.1_ASM1658430v1_cds_headers.txt", delim = "nothing", col_names = "all") %>% 
  mutate(cds=gsub(">| \\[", "", str_extract(all, ">lcl\\|(.*?) \\[")),
         gene_id=gsub("locus_tag=|\\]", "", str_extract(all, "locus_tag=(.*?)\\]")),
         gene_name=gsub("gene=|\\]", "", str_extract(all, "gene=(.*?)\\]")),
         protein=gsub("protein=|\\]", "", str_extract(all, "protein=(.*?)\\]")),
         protein_id=gsub("protein_id=|\\]", "", str_extract(all, "protein_id=(.*?)\\]"))) %>% dplyr::select(-all) %>% 
    left_join(gene2go_all, by="gene_id") %>% 

    ### Fill in "hypothetical protein" entries with diamond blast results 
    left_join(
        read_delim("../references/GCA_016584305.1_ASM1658430v1_cds_diamond_blast.tab", delim="\t", 
           col_names = c("transcript", "saccver", "pident", "length", "mismatch", "gapopen", 
                         "qstart", "qend", "sstart", "send", "evalue", "bitscore")) %>%
    separate(saccver, sep="\\|", into=c("na", "spid", "gene_uni"), remove = F) %>%
    dplyr::select(-na) %>% separate(gene_uni, sep="_", into=c("gene_uni", "species"), remove=T) %>% 
    dplyr::select(transcript,spid,gene_uni,species,pident,length,evalue), by=c("cds"="transcript")) 

# Copy and save to Uniprot 
#snow.annot %>% dplyr::select(spid) %>% filter(!is.na(spid)) %>% distinct() %>% write_clip(allow_non_interactive = T)

# Read in Uniprot & GO information for SPID genes
snow.annot <- snow.annot %>% 
    left_join(read_delim("../references/GCA_016584305.1_ASM1658430v1_cds_diamond_blast_uniprot.tsv",delim="\t") %>% 
                  clean_names() %>% dplyr::select(-from), by=c("spid"="entry"))

# Save for posterity 
write_delim(snow.annot, file="../references/snow_annotation_spencer.tab", col_names = T, delim="\t")
```

```{r}
# this is the GTF, by the way. 
# snow.genome.gtf <- read.gff(file = "../references/snowcrab_ASM1658430v1.gtf", GFF3 = F) %>% filter(feature=="gene") %>%  
#   mutate(gene_id=gsub("gene_id |;|\"", "", str_extract(attributes, "gene_id (.*?);")),
#        ncbi_id=gsub("db_xref |;|\"|GeneID:", "", str_extract(attributes, "db_xref (.*?);")),
#        biotype=gsub("transcript_biotype |;|\"", "", str_extract(attributes, "transcript_biotype (.*?);")),
#        exon=gsub("exon_number |;|\"", "", str_extract(attributes, "exon_number (.*?);")))
```

## Load RNASeq Data & Stats 

### Get read stats prior to the alignment step 

```{r}
read.stats.raw <- read_delim("../fastqc/raw/multiqc_data/multiqc_general_stats.txt") %>% 
    rename_with(~ gsub("FastQC_mqc-generalstats-fastqc-", "", .x)) %>% 
        mutate(Sample=gsub("5010_|5329_","sample.",Sample)) %>% 
        mutate(Sample=gsub("_.*_L003|_.*_L004|_001", "", Sample))
read.stats.trimmed <- read_delim("../fastqc/trimmed/multiqc_data/multiqc_general_stats.txt") %>% 
    rename_with(~ gsub("FastQC_mqc-generalstats-fastqc-", "", .x)) %>% 
    mutate(Sample=gsub(".trimmed.", "_",Sample)) %>% 
    mutate(Sample=gsub(".v1","",Sample)) %>% 
    mutate(Sample=gsub("sample_","sample.",Sample))

sum(read.stats.raw$total_sequences) #total raw sequences
sum(read.stats.trimmed$total_sequences) #total trimmed sequences
```


<!-- ### Alternate option (not used for publication): read in aligned data - Salmon transcript counts  -->

<!-- ```{bash} -->
<!-- rm ../results/salmon_snow_transcriptome/filenames.txt -->
<!-- for file in ../results/salmon_snow_transcriptome/* -->
<!-- do -->
<!-- filename="$(echo $file)" -->
<!-- sample="$(basename -a $filename | cut -d "/" -f 3  | cut -d "." -f 1)" -->
<!-- printf "%s\t%s\n" "$filename/quant.sf" "$sample" >> ../results/salmon_snow_transcriptome/filenames.txt -->
<!-- done -->
<!-- ``` -->

<!-- ```{r} -->
<!-- filenames <- read_delim(file="../results/salmon_snow_transcriptome/filenames.txt", col_names = c("filename", "sample")) -->
<!-- files <- file.path(filenames$filename) #extract vector of filenames  -->
<!-- all(file.exists(files)) #easy code to check that all files exist!  -->

<!-- # Create sample.info object with annotation name for each sample/file, this will be used throughout the analysis   -->
<!-- sample.info <- read_excel("../data/Crabs for RNA analysis.xlsx", sheet = "Sheet1",  -->
<!--                           skip = 8, col_names = c("sample", "pH", "duration"), na = "N/A") %>%  -->
<!--           mutate_at(vars(pH, duration), as.factor) %>% -->
<!--   mutate(OA=as.factor(case_when( -->
<!--     pH=="pH 7.5" ~ "severe", -->
<!--     pH=="pH 7.8" ~ "moderate", -->
<!--     pH=="Ambient" ~ "ambient"))) %>% -->
<!--   mutate(treatment=as.factor(case_when( -->
<!--     (OA=="severe" & duration=="long") ~ "severe_long", -->
<!--     (OA=="severe" & duration=="short") ~ "severe_short", -->
<!--     (OA=="moderate" & duration=="long") ~ "moderate_long", -->
<!--     (OA=="moderate" & duration=="short") ~ "moderate_short", -->
<!--     TRUE ~ "ambient"))) %>% -->
<!--     mutate(sample=paste0("sample_",sample)) %>%  -->
<!--   left_join(filenames, by="sample") -->
<!-- ``` -->

<!-- _The tximport package has a single function for importing transcript-level estimates. The type argument is used to specify what software was used for estimation (“kallisto”, “salmon”, “sailfish”, and “rsem” are implemented). A simple list with matrices, “abundance”, “counts”, and “length”, is returned, where the transcript level information is summarized to the gene-level. The “length” matrix can be used to generate an offset matrix for downstream gene-level differential analysis of count matrices, as shown below._ -->

<!-- ### Generate gene-level abundance count matrix  -->

<!-- ```{r} -->
<!-- # Trinity map file has: gene \t transcript -->

<!-- tx2gene <- read_tsv("../results/trinity/Trinity-GG.fasta.gene_trans_map", col_names=c("gene","transcript")) %>%  -->
<!--     dplyr::select(transcript, gene) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- txi.gen <- tximport(files, type = "salmon", txIn=TRUE, txOut = FALSE, tx2gene=tx2gene , geneIdCol = "gene", txIdCol = "transcript", countsFromAbundance = "no") -->
<!-- colnames(txi.gen$counts) <- filenames$sample -->
<!-- colnames(txi.gen$abundance) <- filenames$sample -->
<!-- ``` -->

<!-- ## Generate gene-level abundance count matrix  -->

<!-- ```{r} -->
<!-- paste("No. transcripts identified by Salmon across all samples: ", nrow(txi.gen$counts), sep="") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- save(txi.gen, file = "../results/txi.gen") -->
<!-- ``` -->

<!-- ### Check out total fragment count by sample -->

<!-- ```{r} -->
<!-- fragments <- data.frame(colSums(txi.gen$counts)) %>% -->
<!--                   dplyr::rename(read.total = 1) %>% -->
<!--                   rownames_to_column(var="sample") %>% -->
<!--     #mutate(sample=as.numeric(sample)) %>% -->
<!--   left_join(sample.info, by=c("sample")) %>% -->
<!--   mutate(treatment=factor(treatment, ordered = T, levels=c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long"))) -->


<!-- #ggplotly( -->
<!-- ggplot(data = fragments %>% -->
<!--          arrange(OA)) + -->
<!--            geom_bar(aes(x=sample, y=read.total, fill=treatment), stat = "identity") + -->
<!--   ggtitle("Total # fragments by sample") + -->
<!--              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + theme_minimal()#) -->

<!-- fragments%>%arrange(read.total) #sample 45 has very little data -->
<!-- sum(fragments$read.total) -->
<!-- ``` -->
<!-- #### Generate DESeq dataset object from tximport object -->

<!-- ```{r} -->
<!-- dds.treatment <- DESeqDataSetFromTximport(txi.gen, -->
<!--                                           colData = left_join(data.frame(sample=colnames(txi.gen$counts)), sample.info)%>%  -->
<!--                                               column_to_rownames("sample")%>%dplyr::select(OA,treatment),  -->
<!--                                           design = ~ treatment) -->

<!-- nrow(counts(dds.treatment)) #number genes before filtering -->
<!-- sum(counts(dds.treatment)) #number of fragments before filtering -->
<!-- ``` -->

## Read in gene count data from STAR quantification

### Generate tab-separated file that lists data file names (one file for each library) and sample number

```{bash}
rm ../results/star_snowcrab_genome/countsfilenames.txt
for file in ../results/star_snowcrab_genome/*_ReadsPerGene.out.tab
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 2)"
printf "%s\t%s\n" "$filename" "$sample" >> ../results/star_snowcrab_genome/countsfilenames.txt
done
```

### Preview the contents of the countsfilenames.txt file 

```{bash}
head -n 5 ../results/star_snowcrab_genome/countsfilenames.txt
```

### Generate object with filenames and sample annotation information, then join 

```{r}
# Create object from the countsfilenames.txt file 
filenames <- read_delim(file="../results/star_snowcrab_genome/countsfilenames.txt", delim = "\t", col_names = c("filename", "sample")) %>%
  mutate(sampleID=paste("S", sample, sep=""))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

# Create sample.info object with annotation name for each sample/file, this will be used throughout the analysis  
sample.info <- read_excel("../data/Crabs for RNA analysis.xlsx", sheet = "Sheet1", 
                          skip = 8, col_names = c("sample", "pH", "duration"), na = "N/A") %>% 
          mutate_at(vars(pH, duration), as.factor) %>%
  mutate(OA=as.factor(case_when(
    pH=="pH 7.5" ~ "severe",
    pH=="pH 7.8" ~ "moderate",
    pH=="Ambient" ~ "ambient"))) %>%
  mutate(treatment=as.factor(case_when(
    (OA=="severe" & duration=="long") ~ "severe_long",
    (OA=="severe" & duration=="short") ~ "severe_short",
    (OA=="moderate" & duration=="long") ~ "moderate_long",
    (OA=="moderate" & duration=="short") ~ "moderate_short",
    TRUE ~ "ambient"))) %>%
  left_join(filenames)

sample.info %>% filter(sample%!in%c(45,60)) %>% group_by(treatment) %>% tally()

```

### Import count data into list of dataframes (one for each library)

Selecting only the first 2 columns (1=gene ID, 2=count); rename columns 

```{r}
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$sampleID)

# Check out number of counts & number of genes for each sammple 
for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read.delim(file=files[i], header = F))[-1:-4,1:2]
    names(file_list[[i]]) <- c("gene", filenames$sampleID[i])
    print(paste("Total COUNTS,", names(file_list[[i]][2]), ":", prettyNum(sum(file_list[[i]][2]), big.mark=","), sep=" "))
    print(paste("Total GENES,", names(file_list[[i]][2]), ":", prettyNum(nrow(file_list[[i]] %>% filter(.[[2]] != 0)), big.mark=","), sep=" "))
}
```

### Preview counts for one sample to see what gene count data looks like 

```{r}
file_list[[1]] %>% head(n=20)
```

### Merge data from all samples into one dataframe 

```{r}
counts <- file_list %>% purrr::reduce(full_join, by = "gene") %>% column_to_rownames(var="gene")
head(counts) #resulting dataframe has genes by row, and samples by column
```

## Summarize counts and visualize

```{r}
print(paste("Number of samples:", ncol(counts), sep=" "))
print(paste("Total number of genes in dataframe:", prettyNum(nrow(counts), big.mark = ","), sep=" "))
print(paste("Average number of genes per sample:", prettyNum(mean(colSums(counts != 0)), big.mark = ","), sep=" "))
print(paste("Total counts, all samples:", prettyNum(sum(colSums(counts)), big.mark = ","), sep=" "))
#print(paste("Counts for", colnames(counts %>% select(contains("Tank"))), ":",  prettyNum(colSums(counts %>% select(contains("Tank"))), big.mark = ","), sep=" "))


#inspect total counts by sample - this generates an interactive plot, hover over bars to identify sample 
 ggplotly(
   ggplot(data.frame(colSums(counts)) %>% 
            dplyr::rename(count.total = 1) %>% rownames_to_column(var="sample")) + 
     geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample") + 
              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```

#### Create DESeq count matrix

NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

#### Generate DESeq dataset object 

```{r,warning=F,message=FALSE}
dds.treatment <- DESeqDataSetFromMatrix(countData = counts,
                                          colData = (left_join(data.frame(sampleID=colnames(counts)), sample.info) %>%
                                              column_to_rownames("sampleID")%>%dplyr::select(OA,treatment)),
                                          design = ~ treatment)

nrow(counts(dds.treatment)) #number genes before filtering
sum(counts(dds.treatment)) #number of fragments before filtering
```

### Drop outlier sample and low-frequency genes 

```{r}
exclude<-c("S45","S60")
dds.treatment <- dds.treatment[, colnames(dds.treatment) != exclude]  # by name
```

### Remove low-frequency genes 

```{r,warning=F,message=FALSE}
keep <- rowSums(counts(dds.treatment) >= 10) >= 0.1*62
dds.treatment <- dds.treatment[keep, ]
nrow(counts(dds.treatment)) #number genes after filtering
sum(counts(dds.treatment)) #number of fragments after filtering

# Mean by sample 
data.frame(colSums(counts(dds.treatment))) %>%
                  dplyr::rename(read.total = 1) %>%
                   rownames_to_column(var="sample") %>% #summary()
 dplyr::summarise(mean=round(mean(read.total, na.rm=T)),
           sd=round(sd(read.total, na.rm=T)),
           se=round(sd/sqrt(length(sample))),
           median=median(read.total),
           min=min(read.total),
           max=max(read.total)) %>% t() %>% as.data.frame() %>%
   dplyr::mutate(V1=prettyNum(V1, big.mark = ",")) %>%
   dplyr::rename("fragments.per.sample"="V1") #use this to average across all samples
```
```{r}
counts(dds.treatment) %>%
  as.data.frame() %>%
  dplyr::summarise(dplyr::across(everything(), ~ sum(.x != 0))) %>%
  tidyr::pivot_longer(everything(), names_to = "column", values_to = "nonzero_count") %>%
  dplyr::summarise(
    ave = mean(nonzero_count),
    sd  = sd(nonzero_count))
```

#### Transform data for visualization 

The variance stabilizing transformation is recommended prior to visualizing expression data and running PCA's

```{r}
vsd.treatment <- varianceStabilizingTransformation(dds.treatment, blind=FALSE)
save(vsd.treatment, file = "../results/vsd.treatment")

# Save variance-stabilization normalized gene counts x sample matrix, annotated for later use 
counts.trans <- assay(vsd.treatment) %>% t() %>% as.matrix() %>% as.data.frame() %>% 
  rownames_to_column(var = "sampleID") %>% 
  left_join(sample.info[,c("sampleID", "treatment")]) %>%
  column_to_rownames(var = "sampleID") %>% 
  dplyr::select(treatment, everything()) %>% 
  mutate(treatment=factor(treatment, ordered = T, 
                          levels=c("ambient","moderate_short","moderate_long","severe_short","severe_long")))
save(counts.trans, file = "../results/counts.trans")
```

## Perform PCA's

```{r}
#Set color scale 
color_scale <- c("ambient"= "green",
                      "moderate_short"="yellow",
                      "moderate_long"="#ffc043",
                      "severe_short"="red",
                      "severe_long"="darkred")
```

### PCAs using DESeq2

```{r,warning=F, message=F}
# To view the the guts of the PlotPCA function in DESeq2, execute this code chunk. It reveals that, by default, it only uses the top 500 most influential genes to plot the PCA! 

#getMethod("plotPCA","DESeqTransform")
# Generate PCA with points color coded by pH Treatment with various number of top genes to use for principal components, selected by highest row variance  

# Top 500 most influential genes
plotPCA(vsd.treatment, intgroup="treatment", ntop=500) +
           ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 500 genes") +
    geom_point(size=2) +
   scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
      theme_minimal()+ stat_ellipse() 

# Top 2000 most influential genes
plotPCA(vsd.treatment, intgroup="treatment", ntop=2000) +
           ggtitle("PC1 x PC2") +
    geom_point(size=1.5) +
      scale_color_manual(values=color_scale) +
      theme_minimal()+ stat_ellipse() 

# Top 2000 most influential genes
plotPCA(vsd.treatment, intgroup="treatment", ntop=nrow(assay(vsd.treatment))) +
           ggtitle("PC1 x PC2") +
    geom_point(size=1.5) +
      scale_color_manual(values=color_scale) +
      theme_minimal()+ stat_ellipse() 
```

```{r}
# All genes!  Top 3 PCs! 
print(pca_1 <- plotPCA(vsd.treatment, intgroup="treatment", ntop=nrow(assay(vsd.treatment)), pcsToUse = c(1,2)) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\nall genes") + 
    geom_point(size=1.5) +
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
    theme_minimal()+ stat_ellipse())

print(pca_2 <- plotPCA(vsd.treatment, intgroup="treatment", ntop=nrow(assay(vsd.treatment)), pcsToUse = c(1,3)) + 
           ggtitle("PC1 x PC3") + 
    geom_point(size=1.5) +
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
    theme_minimal()+ stat_ellipse())

print(pca_3 <- plotPCA(vsd.treatment, intgroup="treatment", ntop=nrow(assay(vsd.treatment)), pcsToUse = c(2,3)) + 
           ggtitle("PC2 x PC3") + 
    geom_point(size=1.5) +
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
    theme_minimal()+ stat_ellipse())

print(pca_6 <- plotPCA(vsd.treatment, intgroup="treatment", ntop=nrow(assay(vsd.treatment)), pcsToUse = c(4,6)) + 
           ggtitle("PC4 x PC6") + 
    geom_point(size=1.5) +
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
    theme_minimal()+ stat_ellipse())

# pdf(file = "../results/pcas.pdf", width = 8, height = 11)
# pca_1 / pca_2 / pca_3
# dev.off()
```

### Run PCAs using prcomp to get scores, scree plot to assess sign. components, etc. 

This enables screeplot to ID significant PCs, calculates variance, etc. 

```{r, warning=F, message=F}
#pca.princomp <- prcomp(cov(assay(vsd.pH)), scale=F) #scale=F for variance-covariance matrix
pca.princomp <- prcomp(t(assay(vsd.treatment))) # using the same code that's under the hood of PlotPCA but using all genes 
pca.eigenval(pca.princomp) #The Proporation of Variance = %variance 
pc.percent <- pca.eigenval(pca.princomp)[2,1:6]*100
screeplot(pca.princomp, bstick=TRUE, npcs = 50) #looks like PC 1-4 are significant 
pc.percent[1:6] %>% sum() #total variance explained by first 2 (significant) axes
```

#### Annotate PCA results

```{r}
#### Generate dataframe with prcomp results 
tab.expr.annot <- data.frame(sampleID = colnames(assay(vsd.treatment)),
    PC1.all = pca.princomp$x[,1],    
    PC2.all = pca.princomp$x[,2],    
    PC3.all = pca.princomp$x[,3],    
    PC4.all = pca.princomp$x[,4],    
    PC5.all = pca.princomp$x[,5],    
    PC6.all = pca.princomp$x[,6],    
    stringsAsFactors = FALSE) %>% 
    left_join(sample.info, by=c("sampleID")) %>% droplevels() #add treatment info to PC results

#pdf(file="../results/PCA-1x2.pdf", height = 5.25, width = 7.5)
ggscatter(tab.expr.annot, 
          x="PC1.all", y="PC2.all", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = F, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)) #+ 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17, 17, 15, 15))))

#dev.off()
```
```{r}
#pdf(file="../results/PCA-1x3.pdf", height = 5.25, width = 7.5)
ggscatter(tab.expr.annot, 
          x="PC1.all", y="PC3.all", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC3") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)) #+ 
#  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17, 17, 15, 15))))

#dev.off()
```
```{r}
ggscatter(tab.expr.annot, 
          x="PC1.all", y="PC4.all", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC4") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)) #+ 

ggscatter(tab.expr.annot, 
          x="PC1.all", y="PC5.all", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC5") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC5 (", round(pc.percent[5], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)) #+ 

ggscatter(tab.expr.annot, 
          x="PC1.all", y="PC6.all", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC6") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC6 (", round(pc.percent[6], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)) #+ 
```

## Try to find culprit of horseshoe pattern 

- # fragments 
- extraction batch 
- RQN
- Ave. size 
- Insert size median/IQR
- GC% 
- 5’/3’ bias
- % uniquely mapped
- % multimapped 
- % rRNA 
- % mitochondrial
- Exonic/intronic/intergenic fraction
- Duplication rate
- Number of detected genes (counts ≥10)

### Fragment total (after filtering)

```{r}
tab.expr.annot %>% left_join(data.frame(read.total=colSums(counts))%>%rownames_to_column("sampleID")) %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    ggplot(aes(x=score, y=read.total,color=treatment)) + geom_point() + facet_wrap(~PC,scales = "free") +
              scale_color_manual(name="Treatment",
                     values=color_scale) +
    theme_minimal() +
    ggtitle("Number of fragments ~PC score")
```

### pH 

```{r}
tab.expr.annot %>% left_join(data.frame(read.total=colSums(counts))%>%rownames_to_column("sampleID")) %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    ggplot(aes(y=score, x=pH,color=treatment)) + geom_boxplot() +
    geom_jitter(width = 0.2) + facet_wrap(~PC) + 
              scale_color_manual(name="Treatment",
                     values=color_scale) +
    theme_minimal() +
    ggtitle("PC score ~ pH")
```

### Explore Library prep stats and whether they are related to PC scores 

```{r}
# read in library prep data 
prep_data <- read_xlsx("../data/GC3F-IS-_4710__Crabs_for_RNA_analysis.xlsx", sheet = "Prep_data",skip = 2) %>% 
    clean_names() %>% 
    mutate(sampleID=gsub("IS #","S",number)) %>% 
    left_join(tab.expr.annot,by="sampleID") 
```

### RNA Quality 

```{r}
prep_data %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    ggplot(aes(x=score, y=rqn,color=treatment)) + 
    geom_point() + facet_wrap(~PC,scales = "free") + 
    theme_minimal() +
    scale_color_manual(name="Treatment",  values=color_scale) +
    ggtitle("PC score ~ RQN")
```

### Extraction batch 
_silly that they extracted them in batches numerically which coincided with treatment_ 

```{r}
prep_data %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    ggplot(aes(x=extraction_batch, y=score, color=treatment)) + 
    geom_point() + facet_wrap(~PC,scales="free") + 
    theme_minimal() +
    scale_color_manual(name="Treatment",  values=color_scale) +
    ggtitle("PC score ~ extraction batch")
```

Sample number (i.e. order of things)
```{r}
prep_data %>% 
    mutate(sample_number=as.numeric(gsub("sample_","",sample))) %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    ggplot(aes(x=sample_number, y=score, color=treatment)) + 
    geom_point() + facet_wrap(~PC,scales="free") + 
    theme_minimal() + 
    scale_color_manual(name="Treatment",  values=color_scale) +
    ggtitle("PC score ~ sample number")
```

### Average cDNA library length (in bp)

```{r}
prep_data %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    filter(!is.na(score)) %>% 
    ggplot(aes(x=avg_length_bp_in_50_1000_bp, y=score,color=treatment)) + 
    geom_point() + facet_wrap(~PC, scales = "free") + 
    theme_minimal() +
    scale_color_manual(name="Treatment",  values=color_scale) +
    ggtitle("PC score ~ ave bp") 
```

### RNA concentration 

```{r}
prep_data %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    filter(!is.na(score)) %>% 
    ggplot(aes(x=conc_n_m_in_50_1000_bp, y=score,color=treatment)) + 
    geom_point() + facet_wrap(~PC, scales = "free") + 
    theme_minimal() + 
    scale_color_manual(name="Treatment",  values=color_scale) +
    ggtitle("PC score ~ ave bp") 
```

### Read in mapping stats 

```{r}
map_stats <- read_delim("../results/star_snowcrab_genome/multiqc_data/multiqc_general_stats.txt") %>% 
    mutate(sampleID=gsub("sample_","S",Sample))%>%
    dplyr::rename("uniquely_mapped"="STAR_mqc-generalstats-star-uniquely_mapped") %>% 
    dplyr::rename("percent_mapped"="STAR_mqc-generalstats-star-uniquely_mapped_percent") %>% 
    left_join(tab.expr.annot,by="sampleID") 

sum(map_stats$uniquely_mapped)
mean(map_stats$percent_mapped)
sd(map_stats$percent_mapped)
```

### % Mapped - seems to be a pattern here 

```{r}
map_stats %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    ggplot(aes(x=score, y=percent_mapped,color=treatment)) + 
    geom_point() + facet_wrap(~PC, scales = "free") + 
    theme_minimal() + 
              scale_color_manual(name="Treatment",
                     values=color_scale) +
    ggtitle("PC score ~ % map")

# Zoom in on PC1, exclude one sample with low mapping rate
map_stats %>% filter(percent_mapped>75) %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score") %>% 
    filter(PC=="PC1.all") %>% 
    ggplot(aes(x=score, y=percent_mapped,color=treatment)) + 
    geom_point() + facet_wrap(~PC, scales = "free") + 
    theme_minimal() + 
              scale_color_manual(name="Treatment",
                     values=color_scale) +
    ggtitle("PC1 score ~ % map")

# There seems to be some relationship between mapping rate and PC1 score 
```

### Stats from MultiQC of trimmed data 

```{r,warning=F, message=F}
multiqc_trimmed <- read_delim("../fastqc/trimmed/multiqc_data/multiqc_general_stats.txt", delim = "\t") %>% 
    mutate(Sample=gsub(".trimmed.R1.v1|.trimmed.R2.v1","",Sample)) %>%     clean_names() %>%
    mutate(sampleID=gsub("sample_","S",sample))%>%
    left_join(tab.expr.annot, by="sampleID") %>% 
    pivot_longer(starts_with("PC"), names_to = "PC", values_to = "score")

multiqc_trimmed %>% 
    filter(fast_qc_mqc_generalstats_fastqc_percent_duplicates>70) %>% 
    ggplot(aes(x=score, y=fast_qc_mqc_generalstats_fastqc_percent_duplicates,color=treatment)) + 
    geom_point() + facet_wrap(~PC, scales = "free") + 
    theme_minimal() +
    scale_color_manual(name="Treatment", values=color_scale) +
    ggtitle("PC score ~ Duplication rate")

multiqc_trimmed %>% 
    #filter(fast_qc_mqc_generalstats_fastqc_percent_duplicates>70) %>% 
    ggplot(aes(x=score, y=fast_qc_mqc_generalstats_fastqc_percent_gc,color=treatment)) + 
    geom_point() + facet_wrap(~PC, scales = "free") + 
    theme_minimal() +
    scale_color_manual(name="Treatment", values=color_scale) +
    ggtitle("PC score ~ Percent GC")

multiqc_trimmed %>% 
    ggplot(aes(x=score, y=fast_qc_mqc_generalstats_fastqc_avg_sequence_length,color=treatment)) + 
    geom_point() + facet_wrap(~PC, scales = "free") + 
    theme_minimal() +
    scale_color_manual(name="Treatment", values=color_scale) +
    ggtitle("PC score ~ Ave seq length")
```

### I cannot find a technical / sequencing cause of the horseshoe shape in the PCA. It could be caused by a biological factor - e.g. molt status? 

## Differential Expression Analysis with Surrogate Variable Analysis (SVA)

Before running DESeq2 to identify DEGs, I am concerned about the potential unexplained factor influencing the horseshoe shape in PC space. Here I add surrogate variables to my design to account for this variance. 

This script removes unwanted technical variation from RNA-seq data using surrogate variable analysis (SVA). 
In some datasets, hidden batch effects or other unmeasured factors can distort PCA plots (e.g., a "horseshoe" effect) 
and confound differential expression results. SVA estimates these hidden factors (surrogate variables, SVs) 
from the variance-stabilized counts while controlling for the variable of interest (here: treatment). 
The SVs are then added as covariates to a copy of the original DESeqDataSet, creating a new object for analysis. 
For visualization, counts are adjusted by regressing out the SVs before PCA, allowing inspection of how many SVs 
are needed to remove unwanted structure without overfitting. This preserves the original dataset while enabling 
sensitivity tests of different numbers of SVs in the model.

### A. Explore (visualize) expression after removing SVs — OK for PCA, not for testing

```{r}
# --- Define variables to PROTECT during SVA estimation ---
#    Add any known biology you don't want SVA to "clean away"
#    e.g., sex, RIN, library_size_factor, batch (if known), etc.
protect_vars <- ~ treatment  # expand like: ~ treatment + sex + RIN

# --- Create model matrices for SVA ---
mod  <- model.matrix(protect_vars, data = colData(dds.treatment))
mod0 <- model.matrix(~ 1,         data = colData(dds.treatment))

# --- Use transformed counts for SVA estimation (common practice with DESeq2) ---
Y    <- SummarizedExperiment::assay(vsd.treatment)

# --- Automatically choose number of SVs, then estimate them ---
n.sv <- sva::num.sv(Y, mod, method = "be")  # "be" = Buja & Eyuboglu permutations
svobj <- sva::svaseq(Y, mod, mod0, n.sv = n.sv)  # returns svobj$sv, samples x n.sv
sv_df <- setNames(as.data.frame(svobj$sv), paste0("SV", seq_len(ncol(svobj$sv))))

# --- Visualization ONLY: PCA after regressing out SVs (OK for plots, NOT for DE) ---
#    This helps you see whether latent structure (batch-ish) is gone while treatment remains.
sv_mat <- as.matrix(sv_df)
Y_adj  <- limma::removeBatchEffect(Y, covariates = sv_mat)  # visualization only
pca    <- prcomp(t(Y_adj), scale. = FALSE)
autoplot(pca, data = as.data.frame(colData(dds.treatment)), colour = "treatment") +
  theme_minimal() + scale_color_manual(values=color_scale)   
```

#### Assess changes in PCA based on number of SVs

```{r}
try_k <- function(k) {
  if (k == 0L) {
    d <- dds.treatment
    return(d)
  }
  d <- dds.treatment
  colData(d) <- cbind(colData(d), sv_df[, paste0("SV", seq_len(k)), drop = FALSE])
  rhs <- paste(c(paste0("SV", seq_len(k)), all.vars(protect_vars)[-1]), collapse = " + ")
  design(d) <- as.formula(paste("~", rhs))
  d
}

# example:
dds_k8  <- try_k( min(8L, ncol(sv_df)) )
dds_k8  <- DESeq(dds_k8)
DESeq2::plotPCA(dds_k8)

res_k8  <- results(dds_k8, contrast = c("treatment","ambient","moderate_short"))


plotPCA(varianceStabilizingTransformation(dds_k8), 
        intgroup="treatment",  pcsToUse = c(1,2)) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\nall genes") + 
    geom_point(size=1.5) +
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
    theme_minimal()+ stat_ellipse()

```

### Look for correlations among surrogate variables, library prep, and PCs 

```{r}
sva_meta <- cbind(sampleID=colnames(Y), sv_df) %>% 
    left_join(data.frame(read.total=colSums(counts))%>%rownames_to_column("sampleID")) %>% 
    left_join(prep_data%>%dplyr::select(sampleID,extraction_batch,rqn,
                                        adapter_well_used_new_unique_dual_matched_plate,
                                        actual_conc_qubit_ng_u_l:avg_length_bp_130_bp,PC1.all:PC3.all), 
              by="sampleID") %>%
    left_join(map_stats%>%dplyr::select(percent_mapped:sampleID), by="sampleID") %>% 
    separate(adapter_well_used_new_unique_dual_matched_plate,into = c("plate_row","plate_column"),sep = 1) %>% 
    arrange(plate_column, plate_row) %>% mutate(plate_order=1:nrow(.))

# Compute correlations
sva_cor <- sva_meta %>%
  column_to_rownames("sampleID") %>% 
  dplyr::select(where(is.numeric)) %>% 
  dplyr::select(-PC1.all:-PC3.all) %>% 
  cor(use = "pairwise.complete.obs") %>%
  as.data.frame() %>%
  rownames_to_column("var1") %>%
  pivot_longer(-var1, names_to = "var2", values_to = "correlation")

# Plot as a heatmap
ggplot(sva_cor, aes(var1, var2, fill = correlation)) +
  geom_tile() +
  #geom_text(aes(label = round(correlation, 2)), color = "black") +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

Explore some correlations more closely if needed
SV1 = correlates with % mapped. Otherwise nothing else. 

```{r}
sva_meta %>% dplyr::select(read.total, starts_with("SV")) %>% 
    pivot_longer(starts_with("SV")) %>% 
    ggplot(aes(x=value,y=read.total)) + geom_point() + facet_wrap(~name, scales = "free") + theme_minimal()

sva_meta %>% dplyr::select(percent_mapped, starts_with("SV")) %>% 
    pivot_longer(starts_with("SV")) %>% 
    filter(name=="SV1",percent_mapped>65)%>%
    ggplot(aes(x=value,y=percent_mapped)) + geom_point() + facet_wrap(~name, scales = "free") + theme_minimal()
```

### Which Genes are correlated with SVs?

```{r, message=F, error=F, warning=F}
library(limma)

meta <- cbind(colData(dds.treatment), sv_df)
design_sv <- model.matrix(~ treatment + SV1 + SV2 + SV3 + SV4 + SV5 + SV6 + SV7, data = meta)
fit <- lmFit(Y, design_sv)
fit <- eBayes(fit)

sv1_res <- topTable(fit, coef = "SV1", number = Inf, sort.by = "P") 
sv2_res <- topTable(fit, coef = "SV2", number = Inf, sort.by = "P")
sv3_res <- topTable(fit, coef = "SV3", number = Inf, sort.by = "P")
sv4_res <- topTable(fit, coef = "SV4", number = Inf, sort.by = "P")
sv5_res <- topTable(fit, coef = "SV5", number = Inf, sort.by = "P")
sv6_res <- topTable(fit, coef = "SV6", number = Inf, sort.by = "P")
sv7_res <- topTable(fit, coef = "SV7", number = Inf, sort.by = "P")


# Explore expression along SVs in some important ion transport genes 

sv1_genes <- sv1_res %>% filter(adj.P.Val<0.05) %>% rownames_to_column("gene_id") %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue) %>% 
                  distinct() %>% dplyr::select(gene_id, gene_name, protein), by="gene_id") %>% 
    distinct()

sv2_genes <- sv2_res %>% filter(adj.P.Val<0.05) %>% rownames_to_column("gene_id") %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue) %>% 
                  distinct() %>% dplyr::select(gene_id, gene_name, protein), by="gene_id") %>% 
    distinct()

bind_cols(as.data.frame(meta),
    as.data.frame(t(Y[sv1_genes$gene_id,]))) %>% 
    pivot_longer(starts_with("GWK"), names_to="gene_id",values_to = "expr") %>% 
#    filter(gene_id%in%c("GWK47_017801", "GWK47_045539", "GWK47_045540", "GWK47_000782","GWK47_045267","GWK47_022853")) %>%
    left_join(sv1_genes[c("gene_id","protein")], by="gene_id") %>% 
#    filter(grepl("Sodium/potassium-transporting ATPase",protein)) %>%
    filter(treatment %!in% c("moderate_long","severe_long")) %>% 
    filter(grepl("V-type",protein)) %>%
#    filter(SV1>0) %>% 
    ggplot(aes(x=SV1, y=expr, color=treatment)) + geom_point(size=2.5) + theme_bw() + 
    geom_smooth(se=F, method = "gam") + ggtitle("ATPases and membrane transporters") + 
    facet_wrap(~str_wrap(protein,width = 30), scales="free") + 
    theme(legend.position = "bottom", legend.title = element_blank()) 
    ```


### PCA after regressing out SVs
This enables screeplot to ID significant PCs, calculates variance, etc. 

```{r}
pca.princomp.sva <- prcomp(t(Y_adj), scale.=F) #scale=F for variance-covariance matrix
pca.eigenval(pca.princomp.sva) #The Proporation of Variance = %variance 
pc.percent.sva <- pca.eigenval(pca.princomp.sva)[2,1:6]*100
screeplot(pca.princomp.sva, bstick=TRUE, npcs = 50) 
pc.percent.sva[1:2] %>% sum() 

#### Annotate PCA results

#### Generate dataframe with prcomp results 
tab.expr.sva <- data.frame(sampleID = colnames(Y_adj),
    PC1 = pca.princomp.sva$x[,1],    # the first eigenvector
    PC2 = pca.princomp.sva$x[,2],    # the second eigenvector
    PC3 = pca.princomp.sva$x[,3],    # the third eigenvector
    stringsAsFactors = FALSE) %>% 
    left_join(sample.info, by=c("sampleID")) %>%
    mutate(treatment=factor(treatment, ordered = T, 
                            levels=c("ambient",
                              "moderate_short",
                              "moderate_long",
                              "severe_short",
                              "severe_long")))

print(pca_sva_12 <- ggscatter(tab.expr.sva, 
          x="PC1", y="PC2", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = T, star.plot = F) +
  theme_minimal() + #ggtitle("Gene expression PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.sva[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.sva[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)))

print(pca_sva_13 <- ggscatter(tab.expr.sva, 
          x="PC1", y="PC3", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = F, star.plot = T) +
  theme_minimal() + ggtitle("Gene expression PC1xPC3") + 
  xlab(paste("PC1 (", round(pc.percent.sva[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent.sva[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)))

print(pca_sva_23 <- ggscatter(tab.expr.sva, 
          x="PC2", y="PC3", col="treatment", shape="OA", size=3.5, alpha=0.85, 
          ellipse = F, star.plot = T) +
  theme_minimal() + ggtitle("Gene expression PC2xPC3") + 
  xlab(paste("PC2 (", round(pc.percent.sva[2], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent.sva[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(guide="none", values=c("ambient"=19, "moderate"=17, "severe"=15)))

pdf(file="../results/Fig_PCA12_with-sva.pdf", height = 5.25, width = 7.5)
print(pca_sva_12)
dev.off()
```

### perMANOVA - multivariate analysis of variance

```{r}
# Effect of temperature
vsd.t <- t(Y_adj) %>% as.data.frame()
perm <- vegan::adonis2(vsd.t ~ treatment, data=left_join(data.frame(sampleID=colnames(counts)), sample.info) %>% 
            filter(sampleID%!in%c("S45","S60")) %>% 
                column_to_rownames("sampleID")%>%dplyr::select(OA,treatment), permutations = 1000, method="bray")
perm #result - effects 

# Now pairwise comparisons
vsd.t <- left_join(data.frame(sampleID=colnames(counts)), sample.info)%>%
          filter(sampleID%!in%c("S45","S60")) %>% dplyr::select(sampleID, OA,treatment) %>%
          left_join(t(Y_adj) %>% as.data.frame() %>% rownames_to_column("sampleID"),
                    by="sampleID")

# Clustering by treatment? yes all different. 
pairwise.adonis(vsd.t[,-1:-3], vsd.t$treatment, perm = 1000) %>% arrange(p.adjusted) %>%
  #filter(p.adjusted<0.01) %>% 
    dplyr::select(pairs, p.adjusted)
```

### Examine homogeneity of multivariate dispersions using all data 

```{r}
# calculate centroids for each 
b <- tab.expr.sva %>% ungroup() %>% group_by(treatment) %>% 
    dplyr::summarize(PC1.mean=mean(PC1), PC1.sd=sd(PC1), PC2.mean=mean(PC2), PC2.sd=sd(PC2))

# calculate euclidean distance from treatment centroid for each sample 
c <- tab.expr.sva %>% left_join(b, by = "treatment") %>% 
  mutate(dist.cent=sqrt((PC1-PC1.mean)^2+(PC2-PC2.mean)^2))

# examine euclidean distances from centroid 
ggplot(c, aes(x=treatment, y=dist.cent, colour=treatment)) + 
  geom_boxplot() + geom_jitter(width = .2) +
  ggtitle("Euclidean distances to centroid of PCA by treatment\nAll genes") +
  theme_minimal() + theme(legend.position = "none", axis.title.x = element_blank()) +
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure"))

ggplot(data=c, aes(x=PC1, y=PC2, group=treatment, colour=treatment)) + 
  geom_point() + 
    theme_minimal()+ stat_ellipse() +
  geom_segment(aes(x=PC1.mean, y=PC2.mean, xend=PC1, yend=PC2)) +
      geom_point(aes(x=PC1.mean, y=PC2.mean,fill=treatment), col="black", shape=22, size=4) +
   ggtitle("PCA by Treatment, All genes\n(var-stabilizing transformed)") + theme(legend.position = "none") +
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_fill_manual(guide=NULL, values=color_scale)

# Examine homogeneity of multivariate dispersions - distances to centroids 
# By examining distance to centroid on PCA
hist(log(c$dist.cent), breaks = 50) #normal distribution? 
shapiro.test(log(c$dist.cent)) #not really 

# By OA treatment, regardless of time 
summary(aov(log(dist.cent)~OA, data=c)) # yes diff
TukeyHSD(aov(log(dist.cent)~OA, data=c)) # no diff 

# By OA treatment, time points separately 
summary(aov(log(dist.cent)~treatment, data=c)) # yes diff
(TukeyHSD(aov(log(dist.cent)~treatment, data=c)))[[1]] %>% as.data.frame()  %>% 
    clean_names() %>% filter(p_adj<0.1)

c %>% group_by(treatment) %>% dplyr::summarize(mean=round(mean(dist.cent),1),
                                               sd=round(sd(dist.cent),1))
```

### PC Loadings - preview top 100 genes most influential on variance / PC structure 

Interpretation:
- Large positive loadings → genes contributing to the positive side of that PC axis.
- Large negative loadings → genes contributing to the negative side.
- Use the signed table if you want “PC1-high” vs “PC1-low” gene sets for enrichment / exploration 

```{r}
## 1) Function to get top loadings for a given PC
top_loadings <- function(pca, pc = 1, n = 50, signed = FALSE) {
  stopifnot(pc >= 1, pc <= ncol(pca$rotation))
  load <- pca$rotation[, pc]                       # numeric vector, names = genes
  if (signed) {
    # Return top positive and top negative separately
    pos <- sort(load, decreasing = TRUE)
    neg <- sort(load, decreasing = FALSE)
    tibble::tibble(
      gene  = c(names(pos)[seq_len(min(n, length(pos)))],
                names(neg)[seq_len(min(n, length(neg)))]),
      loading = c(pos[seq_len(min(n, length(pos)))],
                  neg[seq_len(min(n, length(neg)))]),
      sign = ifelse(loading >= 0, "positive", "negative"),
      PC   = paste0("PC", pc)
    )
  } else {
    ord <- order(abs(load), decreasing = TRUE)
    idx <- ord[seq_len(min(n, length(ord)))]
    tibble::tibble(
      gene    = names(load)[idx],
      loading = load[idx],
      abs_loading = abs(load[idx]),
      PC      = paste0("PC", pc)
    )
  }
}

## 2) Get top genes for multiple PCs at once
pcs_to_take <- 1:3   # change as needed
n_each      <- 50    # top-N per PC

# (a) by absolute loading, annotate!
top_abs <- dplyr::bind_rows(lapply(pcs_to_take, function(pc)
  top_loadings(pca.princomp.sva, pc = pc, n = n_each, signed = FALSE) %>% 
      left_join(snow.annot %>% dplyr::select(-cds), by=c("gene"="gene_id"))))

# (b) signed: top positive and top negative separately, , annotate!
top_signed <- dplyr::bind_rows(lapply(pcs_to_take, function(pc)
  top_loadings(pca.princomp.sva, pc = pc, n = n_each, signed = TRUE)%>% 
      left_join(snow.annot %>% dplyr::select(-cds), 
                by=c("gene"="gene_id"))))

## 5) Quick peek
dplyr::slice_head(dplyr::arrange(top_abs, PC, dplyr::desc(abs_loading)), n = 10)
dplyr::slice_head(dplyr::group_by(top_signed, PC, sign), n = 5)

# # 4) Write to CSV
# write.csv(top_abs, "../results/pca/top_genes_by_abs_loading_PC1-3.csv", row.names = FALSE)
# write.csv(top_signed, "../results/pca/top_genes_by_signed_loading_PC1-3.csv", row.names = FALSE)
```

### Save list of annotated genes that have highest PC loadings - for DAVID 

```{r}
# ### Enrichment analysis, top 100 genes per PC, make list to then save to file 
# top_pc_loadings_4David <- list(
# 
# # Top 100, both positive & negative PC loadings 
#   "PC1_top100" = 
#     (top_abs%>%filter(PC=="PC1") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC1, 49 annotated
#   
#   "PC2_top100" = 
#     (top_abs%>%filter(PC=="PC2") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC2, 38 annotated
#   
#   "PC3_top100" = 
#     (top_abs%>%filter(PC=="PC3") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC3, 33 annotated
# 
# # Positive PC Loadings only 
#   "PC1_top100_positive" = 
#     (top_signed%>%filter(PC=="PC1" & sign=="positive") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC1, 31 annotated
#   
#   "PC2_top100_positive" = 
#     (top_signed%>%filter(PC=="PC2" & sign=="positive") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC2, 33 annotated
#   
#   "PC3_top100_positive" = 
#     (top_signed%>%filter(PC=="PC3" & sign=="positive") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC3, 25 annotated
# 
#     "PC1_top100_negative" = 
#     (top_signed%>%filter(PC=="PC1" & sign=="negative") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC1, 34 annotated
#   
#   "PC2_top100_negative" = 
#     (top_signed%>%filter(PC=="PC2" & sign=="negative") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid, # Top for PC2, 32 annotated
#   
#   "PC3_top100_negative" = 
#     (top_signed%>%filter(PC=="PC3" & sign=="negative") %>% dplyr::select(spid) %>% filter(!is.na(spid)))$spid) # Top for PC3, 37 annotated
# 
# 
# # Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs 
# write_delim(x = plyr::ldply(top_pc_loadings_4David, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", 
#             file = "../results/pca/top_pc_loadings_4David.txt")
# 
# (snow_blast_top_annot%>% filter(!is.na(spid)))$spid %>% write_clip(allow_non_interactive = T) #background -- all annotated genes in transcriptome 
# 
# # Read in DAVID results 
# read_delim("../results/pca/DAVID_pc1_top100.txt") %>% clean_names() %>%
#     filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT")) %>% 
#     filter(fdr<0.1) %>% arrange(p_value)
# 
# read_delim("../results/pca/DAVID_pc2_top100.txt") %>% clean_names() %>%
#     filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT")) %>% 
#     filter(fdr<0.1) %>% arrange(p_value)
# 
# read_delim("../results/pca/DAVID_pc3_top100.txt") %>% clean_names() %>%
#     filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT")) %>% 
#     filter(fdr<0.1) %>% arrange(p_value)

```

### PCA with top loadings 

```{r}
# Packages
# library(dplyr)
# library(ggplot2)
# library(ggrepel)
#library(tibble)
#library(grid)   # for unit() in arrow()

## --- 1) % variance for axis labels
pct <- round(pc.percent.sva[1:2], 1)

## --- 2) Symmetric biplot scaling (alpha = 1/2) for scores & loadings
eig <- pca.princomp.sva$sdev[1:2]
scores_raw <- as.data.frame(pca.princomp.sva$x[, 1:2, drop = FALSE])
colnames(scores_raw) <- c("PC1","PC2")

scores_sym <- as.data.frame(sweep(scores_raw, 2, eig^0.5, `*`)) %>%
  mutate(sampleID = rownames(pca.princomp.sva$x)) %>%
  left_join(sample.info, by = "sampleID") %>%
  mutate(
    treatment = factor(treatment,
      levels = c("ambient","moderate_short","moderate_long","severe_short","severe_long"),
      ordered = TRUE
    )
  )

loadings_sym <- as.data.frame(sweep(pca.princomp.sva$rotation[, 1:2, drop = FALSE], 2, eig^0.5, `*`))
# Keep rownames as gene IDs for lookup
loadings_sym$gene <- rownames(loadings_sym)

## --- 3) Pick top genes (by |loading| across PC1/PC2), keep annotated ones
pcs_to_take <- 1
n_each <- 6  # top absolute loadings 

top_abs <- dplyr::bind_rows(lapply(pcs_to_take, function(pc)
  top_loadings(pca.princomp.sva, pc = pc, n = n_each, signed = TRUE) %>%
    left_join(
      snow.annot %>% dplyr::select(-cds),
      by = c("gene" = "gene_id"))))

annot_top <- top_abs %>%
  filter(!if_all(c(gene_name), is.na)) %>%  # keep only annotated
  distinct(gene, .keep_all = TRUE) #%>%
  #dplyr::arrange(desc(abs_loading)) %>%
  #slice_head(n = 10)  # total arrows; reduce if cluttered

## --- 4) Arrow endpoints (scale relative to score cloud)
score_extent <- max(apply(scores_sym[, c("PC1","PC2")], 2, function(v) diff(range(v))))
arrow_scale  <- 0.8 * score_extent

load_df <- annot_top %>%
  dplyr::select(gene, loading, PC, gene_name) %>% #, abs_loading
  left_join(loadings_sym, by = "gene") %>%
  transmute(
    gene,
    PC1_end = PC1 * arrow_scale,
    PC2_end = PC2 * arrow_scale,
    label = dplyr::coalesce(gene_name)
  )

## --- 5) PC1 × PC2 biplot with loadings
## --- 5) PC1 × PC2 biplot with loadings
p_biplot <- ggplot(scores_sym, aes(PC1, PC2)) +
  geom_hline(yintercept = 0, linewidth = 0.2, color = "gray70") +
  geom_vline(xintercept = 0, linewidth = 0.2, color = "gray70") +

  # --- NEW: shaded ellipses by treatment (draw first so points/arrows sit on top)
  stat_ellipse(
    aes(fill = treatment, color = treatment),
    type = "norm", level = 0.68, geom = "polygon",
    alpha = 0.18, linewidth = 0.4
  ) +

  # sample points (color + shape both mapped to treatment)
  geom_point(aes(color = treatment, shape = treatment), size = 3.5, alpha = 0.9) +

  # # loading vectors (from origin)
  # geom_segment(
  #   data = load_df,
  #   aes(x = 0, y = 0, xend = PC1_end, yend = PC2_end),
  #   inherit.aes = FALSE,
  #   linewidth = 0.5, alpha = 0.6,
  #   arrow = arrow(length = unit(0.02, "npc"), type = "closed"),
  #   color = "gray30"
  # ) +
  # 
  # # labels for loading tips
  # ggrepel::geom_text_repel(
  #   data = load_df,
  #   aes(x = PC1_end, y = PC2_end, label = label),
  #   inherit.aes = FALSE,
  #   size = 4, max.overlaps = Inf,
  #   box.padding = 0.25, point.padding = 0.05,
  #   segment.alpha = 0.3, min.segment.length = 0
  # ) +

  # scales (match points & ellipses)
  scale_color_manual(
    name = "Treatment",
    values = color_scale,
    labels = c(
      "ambient"         = "Control",
      "moderate_short"  = "pH 7.8, Short Exposure",
      "moderate_long"   = "pH 7.8, Long Exposure",
      "severe_short"    = "pH 7.5, Short Exposure",
      "severe_long"     = "pH 7.5, Long Exposure"
    )
  ) +
  scale_fill_manual(
    name = "Treatment",
    values = color_scale,
    labels = c(
      "ambient"         = "Control",
      "moderate_short"  = "pH 7.8, Short Exposure",
      "moderate_long"   = "pH 7.8, Long Exposure",
      "severe_short"    = "pH 7.5, Short Exposure",
      "severe_long"     = "pH 7.5, Long Exposure"
    )
  ) +
  scale_shape_manual(
    name = "Treatment",
    values = c(
      "ambient"        = 19,
      "moderate_short" = 15,
      "moderate_long"  = 15,
      "severe_short"   = 17,
      "severe_long"    = 17
    )
  ) +

  # force one combined legend (color legend shows both shape + color)
  guides(
    color = guide_legend(
      override.aes = list(
        shape = c(19, 15, 15, 17, 17),
        fill  = NA  # keeps points open, avoids filled symbols in legend
      )
    ),
    shape = "none",
    fill  = "none"
  ) +

  labs(
    x = paste0("PC1 (", pct[1], "%)"),
    y = paste0("PC2 (", pct[2], "%)"),
    title = "PCA biplot (PC1 × PC2)" # with top loadings
  ) +
  theme_minimal(base_size = 12) +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 10),
    legend.title = element_text(size = 11)
  )

p_biplot
```

### PCA with heatmap of top loading genes 

```{r}
## 0) Inputs assumed available:
## - pca.princomp.sva (from prcomp on t(Y_adj))
## - sample.info (data.frame with 'sample', 'treatment', optional 'OA')
## - Y_adj: genes x samples (rows = genes, cols = sample IDs)
## - top_loadings() helper from your earlier snippet

## 1) PC scores per sample
# tab.expr.sva

## 2) Top genes by PC1 loading
n_top <- 10
top_pc1 <- top_loadings(pca.princomp.sva, pc = 1, n = n_top, signed = TRUE) %>%
  dplyr::select(gene, loading, sign) %>%
  dplyr::arrange(desc(abs(loading)))   # keep your magnitude sort

## ---- NEW: build a label map and compute the label order ----
label_map <- snow.annot %>%
  dplyr::select(gene_id, gene_name) %>%
  mutate(label = coalesce(gene_name, gene_id))

# order by your chosen criterion (here: loading desc)
order_tbl <- top_pc1 %>%
  left_join(label_map, by = c("gene" = "gene_id")) %>%
  mutate(label = coalesce(label, gene)) %>%
  dplyr::arrange(desc(loading))

# handle duplicated labels (same symbol for multiple IDs)
order_tbl <- order_tbl %>%
  dplyr::add_count(label, name = "n_label") %>%
  dplyr::mutate(label = dplyr::if_else(n_label > 1,
                                       paste0(label, " [", gene, "]"),
                                       label)) %>%
  dplyr::select(-n_label)

label_levels <- order_tbl$label   # this is the y-axis order you want

## 3) Expression matrix (still by IDs)
stopifnot(all(top_pc1$gene %in% rownames(Y_adj)))
mat_top <- Y_adj[top_pc1$gene, tab.expr.sva$sampleID, drop = FALSE]
mat_top_z <- t(scale(t(mat_top))); mat_top_z[is.na(mat_top_z)] <- 0

## 4) Long format + metadata + labels (apply the factor levels!)
# --- de-duplicate keys before joining ---
#top_pc1      <- top_pc1      %>% distinct(gene,     .keep_all = TRUE)
label_map    <- label_map    %>% distinct(gene_id,  .keep_all = TRUE)
order_tbl    <- order_tbl    %>% distinct(gene,     .keep_all = TRUE)

# --- unique levels for the factor ---
label_levels <- unique(label_levels)

# --- build long table ---
top_pc1_long <- as.data.frame(mat_top_z) %>%
  tibble::rownames_to_column("gene") %>%
  pivot_longer(-gene, names_to = "sampleID", values_to = "z") %>%
  left_join(tab.expr.sva, by = "sampleID") %>%                 # or relationship = "many-to-one"
  left_join(top_pc1,      by = "gene") %>%                     # or relationship = "many-to-one"
  left_join(label_map,    by = c("gene" = "gene_id")) %>%      # or relationship = "many-to-one"
  mutate(label = coalesce(label, gene)) %>%
  # keep the same disambiguation as in order_tbl (now one row per gene)
  left_join(order_tbl %>% dplyr::select(gene, label), by = "gene",
            suffix = c("", ".ord")) %>%
  mutate(label = coalesce(label.ord, label)) %>%
  dplyr::select(-label.ord) %>%
  mutate(
    sign  = factor(sign, levels = c("positive","negative")),
    label = factor(label, levels = label_levels)   # no duplicates now
  )

## 5) Tile width
dx <- diff(range(top_pc1_long$PC1, na.rm = TRUE))
nS <- length(unique(top_pc1_long$sample))
tile_w <- if (nS > 1) dx / (nS - 1) else dx * 0.05

## 6) Plot (now ordered by loading even though y=label)
tile_pc1 <- ggplot(top_pc1_long, aes(x = PC1, y = label, fill = z)) +
  geom_vline(xintercept = 0, linewidth = 0.2, color = "gray80") +
  geom_tile(width = tile_w, height = 0.9) +
  scale_fill_gradientn(
    name = "Row‑z(VST)",
    colours = c( "navy","white","chocolate"),
    values  = scales::rescale(c(-3,-1,1,3)),
    limits  = c(-3,3), oob = scales::squish
  ) +
  labs(
    x = "PC1 score",
    y = NULL,
#    y = "Top PC1‑loading genes",
#    title = "Expression of top PC1‑influential genes (ordered by loading)"
  ) +
  theme_minimal(base_size = 6) +
  theme(panel.grid = element_blank(), axis.text.y = element_text(size = 6),
        legend.position = "right")
tile_pc1
```

### How many genes should we include in our "Top Loadings" heatmap? 

1% = 111 genes 

```{r}
pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC1) %>% mutate(abs=abs(PC1)) %>% 
    #filter(abs>0.04)
    slice_max(abs,prop = 0.01) %>% summary()
hist(abs(pca.princomp.sva$rotation[,1]),breaks = 1000, main="PC1 loadings", xlab = "Abs(PC1 loadings)")
abline(v = 0.03127)

pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC2) %>% mutate(abs=abs(PC2))  %>% 
    #filter(abs>0.04)
    slice_max(abs,prop = 0.01) %>% summary()
hist(abs(pca.princomp.sva$rotation[,2]),breaks = 1000, main="PC2 loadings", xlab = "Abs(PC2 loadings)")
abline(v = 0.03273)

citation("vegan")
```

### Enrichment analysis, top 1% of loadings 

#### Prep list to start adding for DAVID enrichment analysis 

```{r}
david.lists <- vector(mode = "list",length = 20)

names(david.lists) <- c("pc1_pos_loadings", "pc1_neg_loadings", "pc2_pos_loadings", "pc2_neg_loadings",
                        "mod_short_all", "mod_short_up", "mod_short_down",
                        "sev_short_all", "sev_short_up", "sev_short_down",
                        "intersect_short_all", "intersect_short_up", "intersect_short_down",
                        "mod_persist_all", "sev_persist_all", "intersect_persist_all",
                        "long_all","long_up","long_down", "background")
```


```{r}
#TERM2GENE 

# gene universe = all genes that have any GO term, filtered for only analyzed genes 
universe <- (TERM2GENE%>%filter(gene_id %in% rownames(counts(dds.treatment))))$gene_id %>% unique()

# Genes with top positive loadings on PC1 (higher expression at day 84)
pc1_genes_pos <- rownames(
    pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC1) %>% 
    #mutate(abs=abs(PC1))  %>% 
    slice_max(PC1,prop = 0.005))

summary(pc1_pos_ego <- enricher(
  gene      = pc1_genes_pos,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=3)) %>% filter(p.adjust<0.05) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

david.lists[["pc1_pos_loadings"]] <- data.frame("gene_id"=pc1_genes_pos) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

# Copy to save to spreadsheet
david.lists[["pc1_pos_loadings"]] %>% 
    left_join(pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC1) %>% rownames_to_column("gene_id"), 
              by="gene_id") %>% arrange(desc(PC1)) %>% select(PC1, everything()) %>% 
    write_clip(allow_non_interactive = T)

# Genes with top negative loadings on PC1 (higher expression in 8hr OA exposures)
pc1_genes_neg <- rownames(
    pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC1) %>% 
    #mutate(abs=abs(PC1))  %>% 
    slice_min(PC1,prop = 0.005))

summary(pc1_neg_ego <- enricher(
  gene      = pc1_genes_neg,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize = 3)) %>% filter(p.adjust<0.05) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

david.lists[["pc1_neg_loadings"]] <- data.frame("gene_id"=pc1_genes_neg) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

# Copy to save to spreadsheet
david.lists[["pc1_neg_loadings"]] %>% 
    left_join(pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC1) %>% rownames_to_column("gene_id"), 
              by="gene_id") %>% arrange(PC1) %>% select(PC1, everything()) %>% 
    write_clip(allow_non_interactive = T)

# List of top 10 genes per sign per PC 
pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC1) %>% 
    slice_max(PC1,n = 10) %>% rownames_to_column("gene_id") %>% 
    left_join(snow.annot%>%dplyr::select(-cds)) %>% 
    select(gene_id:protein,spid,protein_names) %>% distinct() #%>% filter(!is.na(gene_name) & !is.na(spid))

pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC1) %>% 
    slice_min(PC1,n = 5) %>% rownames_to_column("gene_id") %>% 
    left_join(snow.annot%>%dplyr::select(-cds)) %>% 
    select(gene_id:protein,GO,spid,protein_names) %>% distinct() #%>% filter(!is.na(gene_name) & !is.na(spid))

# Also add PC2 loadings
pc2_genes_pos <- rownames(
    pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC2) %>% 
    #mutate(abs=abs(PC1))  %>% 
    slice_max(PC2,prop = 0.005))

summary(pc2_pos_ego <- enricher(
  gene      = pc2_genes_pos,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize = 3)) %>% filter(p.adjust<0.05) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

pc2_genes_neg <- rownames(
    pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC2) %>% 
    #mutate(abs=abs(PC1))  %>% 
    slice_min(PC2,prop = 0.005))

summary(pc2_neg_ego <- enricher(
  gene      = pc2_genes_neg,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize = 3)) %>% filter(p.adjust<0.05) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())


# Save to DAVID lists

# Genes with top positive loadings on PC2 
pc2_genes_pos <- rownames(
    pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC2) %>% 
    slice_max(PC2,prop = 0.005))
pc2_genes_neg <- rownames(
    pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC2) %>% 
    slice_min(PC2,prop = 0.005))

david.lists[["pc2_pos_loadings"]] <- data.frame("gene_id"=pc2_genes_pos) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()
david.lists[["pc2_neg_loadings"]] <- data.frame("gene_id"=pc2_genes_neg) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

# Copy to save to spreadsheet
david.lists[["pc2_pos_loadings"]] %>% 
    left_join(pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC2) %>% rownames_to_column("gene_id"), 
              by="gene_id") %>% arrange(desc(PC2)) %>% select(PC2, everything()) %>% 
    write_clip(allow_non_interactive = T)

david.lists[["pc2_neg_loadings"]] %>% 
    left_join(pca.princomp.sva$rotation %>% as.data.frame() %>% dplyr::select(PC2) %>% rownames_to_column("gene_id"), 
              by="gene_id") %>% arrange(PC2) %>% select(PC2, everything()) %>% 
    write_clip(allow_non_interactive = T)
```



```{r}
make_pc_tile <- function(pc = 1, n_top = 20, pos_labels = "left", pos_legend = "right",
                         palette = c("blue","cyan","yellow","red")) {
  stopifnot(pc %in% c(1,2,3))
  pc_col    <- paste0("PC", pc)
  xlab_text <- paste0("PC", pc, " score")

  ## Top genes by signed loading (ensure one row per gene)
  top_pc <- top_loadings(pca.princomp.sva, pc = pc, n = n_top, signed = TRUE) %>%
    dplyr::select(gene, loading, sign) %>%
    dplyr::arrange(dplyr::desc(loading)) %>%
    dplyr::distinct(gene, .keep_all = TRUE)

  ## Annotation map (may be NA); one name per gene
  label_map <- snow.annot %>%
    dplyr::select(gene_id, gene_name, protein) %>%
    dplyr::distinct(gene_id, .keep_all = TRUE)

  ## Order table and unique ID levels
  order_tbl <- top_pc %>%
    dplyr::left_join(label_map, by = c("gene" = "gene_id")) %>%
    dplyr::arrange(dplyr::desc(loading)) %>%
    dplyr::distinct(gene, .keep_all = TRUE)

  id_levels <- unique(order_tbl$gene)

  ## Build y-axis labels:
  ##  - use gene_names when present, else gene ID
  ##  - if a name is duplicated, append " [gene]" only for the duplicates
  name_map <- order_tbl %>% dplyr::select(gene, gene_name)
  label_tbl <- dplyr::tibble(gene = id_levels) %>%
    dplyr::left_join(name_map, by = "gene") %>%
    dplyr::mutate(label = dplyr::if_else(is.na(gene_name) | gene_name == "", gene, gene_name)) %>%
    dplyr::group_by(label) %>%
    dplyr::mutate(label = dplyr::if_else(dplyr::n() > 1, paste0(label, " [", gene, "]"), label)) %>%
    dplyr::ungroup()

  lab_vec <- setNames(label_tbl$label, label_tbl$gene)

  ## Expression (IDs)
  stopifnot(all(top_pc$gene %in% rownames(Y_adj)))
  tab_expr_unique <- tab.expr.sva %>% dplyr::distinct(sampleID, .keep_all = TRUE)

  mat_top <- Y_adj[top_pc$gene, tab_expr_unique$sampleID, drop = FALSE]
  mat_top_z <- t(scale(t(mat_top))); mat_top_z[is.na(mat_top_z)] <- 0

  ## Long format
  long_df <- as.data.frame(mat_top_z) %>%
    tibble::rownames_to_column("gene") %>%
    tidyr::pivot_longer(-gene, names_to = "sampleID", values_to = "z") %>%
    dplyr::left_join(tab_expr_unique, by = "sampleID") %>%
    dplyr::left_join(top_pc %>% dplyr::distinct(gene, .keep_all = TRUE), by = "gene") %>%
    dplyr::mutate(
      gene = factor(gene, levels = id_levels),   # order by loading
      PC   = .data[[pc_col]]
    )

  ## Tile width
  dx  <- diff(range(long_df$PC, na.rm = TRUE))
  nS  <- length(unique(long_df$sample))
  tile_w <- if (nS > 1) dx / (nS - 1) else max(dx * 0.05, 0.05)

  ## Plot
  ggplot(long_df, aes(x = PC, y = gene, fill = z)) +
    geom_vline(xintercept = 0, linewidth = 0.2, color = "gray80") +
    geom_tile(width = tile_w, height = 0.9) +
    scale_y_discrete(limits = id_levels, labels = lab_vec, position = pos_labels) +
    scale_fill_gradientn(
      name   = "Row-z(VST)",
      colours = palette,
      values  = scales::rescale(c(-3, -1, 1, 3)),
      limits  = c(-3, 3),
      oob     = scales::squish
    ) +
    labs(x = xlab_text, y = NULL) +
    theme_minimal(base_size = 5) +
    theme(panel.grid = element_blank(),
          axis.text.y = element_text(size = 5),
          legend.position = pos_legend)
}

## Generate tiles
pc1_tile <- make_pc_tile(pc = 1, n_top = 55, pos_labels = "right", pos_legend = "right",
                         palette = c("blue","cyan","yellow","red"))
pc2_tile <- make_pc_tile(pc = 2, n_top = 55, pos_labels = "right", pos_legend = "right",
                         palette = c("blue","cyan","yellow","red"))
pc3_tile <- make_pc_tile(pc = 3, n_top = 55, pos_labels = "left", pos_legend = "left",
                         palette = c("blue","cyan","yellow","red"))

pdf("../results/pca/Figure_heatmap_pc1.pdf", width = 7, height = 2)
pc1_tile
dev.off()

pdf("../results/pca/Figure_heatmap_pc2.pdf", width = 7, height = 2)
pc2_tile
dev.off()

pc3_tile

pdf("../results/pca/Figure_pca.pdf", width = 9, height = 7)
#pca_sva_12
p_biplot
dev.off()

```

```{r}
# (a / d) + plot_layout(axis_titles = "collect")
((pc2_tile+coord_flip()) + pca_sva_12 / pc1_tile) 
```

## C.  Differential Expression Analysis with SVs as covariates  

```{r}
# --- Attach SVs to a COPY of your DESeqDataSet ---
dds_sva <- dds.treatment
colData(dds_sva) <- cbind(colData(dds_sva), sv_df)

# (Optional) relevel treatment so your contrasts are intuitive
dds_sva$treatment <- relevel(dds_sva$treatment, ref = "ambient")

# --- Build the DESeq2 design: SVs + protected biological terms ---
#    Keep the same variables you protected in (A1), and prepend all SVs.
# collect the SV terms
sv_terms <- paste0("SV", seq_len(ncol(sv_df)))

# add your biological variables of interest directly
bio_terms <- c("treatment")   # add "sex", "RIN", etc. here if needed

# full right-hand side of the formula
rhs <- paste(c(sv_terms, bio_terms), collapse = " + ")

# set the design
design(dds_sva) <- as.formula(paste("~", rhs))

# --- Fit DESeq2 !---
dds_sva <- DESeq2::DESeq(dds_sva)

# --- Differential Expression PAIRWISE CONTRASTS --- 
res_amb_mod_short <- results(dds_sva, contrast = c("treatment", "ambient", "moderate_short"), alpha = 0.1)
res_amb_sev_short   <- results(dds_sva, contrast = c("treatment", "ambient", "severe_short"),   alpha = 0.1)
res_OA_short          <- results(dds_sva, contrast = c("treatment", "severe_short", "moderate_short"), alpha = 0.1)
res_amb_mod_long <- results(dds_sva, contrast = c("treatment", "ambient", "moderate_long"), alpha = 0.1)
res_amb_sev_long   <- results(dds_sva, contrast = c("treatment", "ambient", "severe_long"),   alpha = 0.1)
res_OA_long           <- results(dds_sva, contrast = c("treatment", "moderate_long", "severe_long"),   alpha = 0.1)
# res_mod <- results(dds_sva, contrast = c("treatment", "moderate_short", "moderate_long"), alpha = 0.1)
# res_sev <- results(dds_sva, contrast = c("treatment", "severe_short", "severe_long"), alpha = 0.1)

# ####  Use this code to NOT include surrogate variables in design 
# dds.treatment <- DESeq2::DESeq(dds.treatment)
# 
# # --- Differential Expression PAIRWISE CONTRASTS --- 
# res_amb_mod_short <- results(dds.treatment, contrast = c("treatment", "ambient", "moderate_short"), alpha = 0.1)
# res_amb_sev_short   <- results(dds.treatment, contrast = c("treatment", "ambient", "severe_short"),   alpha = 0.1)
# res_OA_short          <- results(dds.treatment, contrast = c("treatment", "severe_short", "moderate_short"), alpha = 0.1)
# res_amb_mod_long <- results(dds.treatment, contrast = c("treatment", "ambient", "moderate_long"), alpha = 0.1)
# res_amb_sev_long   <- results(dds.treatment, contrast = c("treatment", "ambient", "severe_long"),   alpha = 0.1)
# res_OA_long           <- results(dds.treatment, contrast = c("treatment", "moderate_long", "severe_long"),   alpha = 0.1)
# ####  

# --- Post-filtering examples ---
keep_fun <- function(res) subset(res, padj < 0.05) #& padj < 0.05 & abs(log2FoldChange) > 0.5 
diffex_amb_mod_short <- keep_fun(res_amb_mod_short)
diffex_amb_sev_short <- keep_fun(res_amb_sev_short)
diffex_OA_short     <- keep_fun(res_OA_short)
diffex_amb_mod_long <- keep_fun(res_amb_mod_long) # <-- not ideal, ctrl at time-0
diffex_amb_sev_long <- keep_fun(res_amb_sev_long) # <-- not ideal, ctrl at time-0
diffex_OA_long     <- keep_fun(res_OA_long)
# diffex_mod     <- keep_fun(res_mod)
# diffex_sev     <- keep_fun(res_sev)

# Example look for intersecting genes and with larger effect size too 
intersect(rownames(diffex_amb_mod_short), rownames(diffex_amb_sev_short))
intersect(rownames(subset(diffex_amb_mod_short, abs(log2FoldChange) > 0.5)), rownames(subset(diffex_amb_sev_short, abs(log2FoldChange) > 0.5)))
```

### Examine potential influence of outliers 

Inspect Cook's distances for all genes. 

```{r}
par(mar = c(8,5,2,2))
boxplot(log10(assays(dds_sva)[["cooks"]]), range = 0, las = 2)
```

### Identify outliers using iLOO 

DESeq2 has integrated outlier detection. But, after inspecting DEG's, there do appear to be a few influenced by outlier samples. So, I also perform outlier detection using the iterative Leave-One-Out method (iLOO) and script from [George et al. 2015](https://doi.org/10.1371/journal.pone.0125224), see [iLOO.R script](../references/iLOO.R)

Note: The function iLOO requires an input  matrix, which represents a matrix of read counts for a treatment or homogeneous group containing   samples and  features. iLOO returns a  matrix, where only the outlier read counts are present (all non-outlier values have been NA’ed). 

```{r}
#sample.info$treatment %>% levels()

# Create vectors of sample IDs per treatment
samples.ambient <- (sample.info[c("sampleID", "treatment")] %>% filter(treatment=="ambient"))$sampleID
samples.moderate_long <- (sample.info[c("sampleID", "treatment")] %>% filter(treatment=="moderate_long", sampleID!="S45"))$sampleID
samples.moderate_short <- (sample.info[c("sampleID", "treatment")] %>% filter(treatment=="moderate_short", sampleID!="S60"))$sampleID
samples.severe_long <- (sample.info[c("sampleID", "treatment")] %>% filter(treatment=="severe_long"))$sampleID
samples.severe_short <- (sample.info[c("sampleID", "treatment")] %>% filter(treatment=="severe_short"))$sampleID

# Create vector of all DEGs identified among any contrast 
degs <- unique(c(rownames(diffex_amb_mod_short),rownames(diffex_amb_sev_short),rownames(diffex_OA_short), rownames(diffex_amb_mod_long), rownames(diffex_amb_sev_long)))
length(degs)

# Extract gene counts that have outliers replaced with trimmed means, and filter for DEGs. NOTE: the iLOO function requires raw counts. 
degs.counts <- assay(dds_sva) %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(gene %in% degs)

# Extract counts for 
counts.ambient <- degs.counts %>% dplyr::select(gene, all_of(samples.ambient)) %>% column_to_rownames("gene")
counts.moderate_long <- degs.counts %>% dplyr::select(gene, all_of(samples.moderate_long)) %>% column_to_rownames("gene")
counts.moderate_short <- degs.counts %>% dplyr::select(gene, all_of(samples.moderate_short)) %>% column_to_rownames("gene")
counts.severe_long <- degs.counts %>% dplyr::select(gene, all_of(samples.severe_long)) %>% column_to_rownames("gene")
counts.severe_short <- degs.counts %>% dplyr::select(gene, all_of(samples.severe_short)) %>% column_to_rownames("gene")

# Run the iterative Leave-One-Out function on gene counts for each treatment 
iLOO.ambient <- iLOO(counts.ambient)
iLOO.moderate_long <- iLOO(counts.moderate_long)
iLOO.moderate_short <- iLOO(counts.moderate_short)
iLOO.severe_long <- iLOO(counts.severe_long)
iLOO.severe_short <- iLOO(counts.severe_short)

#### Generate lists of degs that have an outlier in at least one sample

# Ambient vs. moderate-short
outs.diffex.mod.short <- full_join(
    
  iLOO.ambient[rowSums(is.na(iLOO.ambient)) != ncol(iLOO.ambient), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_mod_short)),

iLOO.moderate_short[rowSums(is.na(iLOO.moderate_short)) != ncol(iLOO.moderate_short), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_mod_short)),
by="gene")

# Ambient vs. severe-short
outs.diffex.sev.short <- full_join(
    
iLOO.ambient[rowSums(is.na(iLOO.ambient)) != ncol(iLOO.ambient), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_sev_short)),

iLOO.severe_short[rowSums(is.na(iLOO.severe_short)) != ncol(iLOO.severe_short), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_sev_short)),
by="gene")

# moderate-short vs. severe-short
outs.diffex.short <- full_join(
  iLOO.moderate_short[rowSums(is.na(iLOO.moderate_short)) != ncol(iLOO.moderate_short), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_OA_short)),

iLOO.severe_short[rowSums(is.na(iLOO.severe_short)) != ncol(iLOO.severe_short), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_OA_short)),
by="gene")

# Ambient vs. moderate-long
outs.diffex.mod.long <- full_join(
    
  iLOO.ambient[rowSums(is.na(iLOO.ambient)) != ncol(iLOO.ambient), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_mod_long)),

iLOO.moderate_short[rowSums(is.na(iLOO.moderate_long)) != ncol(iLOO.moderate_long), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_mod_long)),
by="gene")

# Ambient vs. severe-long
outs.diffex.sev.long <- full_join(
    
  iLOO.ambient[rowSums(is.na(iLOO.ambient)) != ncol(iLOO.ambient), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_sev_long)),

iLOO.moderate_short[rowSums(is.na(iLOO.severe_long)) != ncol(iLOO.severe_long), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_amb_sev_long)),
by="gene")

# moderate-long vs. severe-long
outs.diffex.long <- full_join(
  iLOO.moderate_long[rowSums(is.na(iLOO.moderate_long)) != ncol(iLOO.moderate_long), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_OA_long)),

iLOO.severe_long[rowSums(is.na(iLOO.moderate_long)) != ncol(iLOO.moderate_long), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex_OA_long)),
by="gene")
```

#### How many genes could be removed due to potential outlier effects? 114

```{r}
data.frame("outliers"=
               c(outs.diffex.mod.short$gene, 
  outs.diffex.sev.short$gene,
  outs.diffex.short$gene,
  outs.diffex.mod.long$gene,
  outs.diffex.sev.long$gene,
  outs.diffex.long$gene) %>% unique())
```

### Generate final list of DEGs that will be used for subsequent analyses & results! 

Remove genes from DEG lists that contain iLOO-identified outliers from DEG objects, and save to file. 

```{r}
(diffex_amb_mod_short_filt <- subset(diffex_amb_mod_short, rownames(diffex_amb_mod_short) %!in% c((outs.diffex.mod.short%>%filter(gene!="GWK47_053415"))$gene)))
(diffex_amb_sev_short_filt <- subset(diffex_amb_sev_short, rownames(diffex_amb_sev_short) %!in% c(outs.diffex.sev.short$gene)))
(diffex_OA_short_filt <- subset(diffex_OA_short, rownames(diffex_OA_short) %!in% c(outs.diffex.short$gene)))
(diffex_amb_mod_long_filt <- subset(diffex_amb_mod_long, rownames(diffex_amb_mod_long) %!in% c(outs.diffex.mod.long$gene)))
(diffex_amb_sev_long_filt <- subset(diffex_amb_sev_long, rownames(diffex_amb_sev_long) %!in% c(outs.diffex.sev.long$gene)))
(diffex_OA_long_filt <- subset(diffex_OA_long, rownames(diffex_OA_long) %!in% c(outs.diffex.long$gene)))

save(diffex_amb_mod_short_filt, file = "../results/deseq2/diffex_amb_mod_short_filt")
save(diffex_amb_sev_short_filt, file = "../results/deseq2/diffex.sev.short.filt")
save(diffex_OA_short_filt, file = "../results/deseq2/diffex_OA_short_filt")
save(diffex_amb_mod_long_filt, file = "../results/deseq2/diffex_amb_mod_long_filt")
save(diffex_amb_sev_long_filt, file = "../results/deseq2/diffex_amb_sev_long_filt")
save(diffex_OA_long_filt, file = "../results/deseq2/diffex_OA_long_filt")

all_degs <- c(rownames(diffex_amb_mod_short_filt),
              rownames(diffex_amb_sev_short_filt),
              rownames(diffex_OA_short_filt),
              rownames(diffex_amb_mod_long_filt),
              rownames(diffex_amb_sev_long_filt),
              rownames(diffex_OA_long_filt)) %>% unique()
```



```{r}
# Add background list of Uniprot SPID to DAVID list 
david.lists[["background"]] <- as.data.frame(res_OA_long) %>% rownames_to_column("gene_id") %>% 
    left_join(snow.annot%>%dplyr::select(gene_id:protein,spid,evalue)%>%group_by(gene_id)%>%slice_min(evalue),
              by="gene_id") %>% distinct() 
```


## Create visualizations 

```{r}
as.data.frame(diffex_amb_sev_short_filt) %>% rownames_to_column("gene") %>%
    filter(gene %in% c((as.data.frame(diffex_amb_sev_long_filt) %>% rownames_to_column("gene")))$gene) %>% 
    left_join(snow.annot, by=c("gene"="gene_id"))
```

### Venn Diagrams of DEGs by pairwise contrast

```{r}
# Generate counts matrix With DEGs among any treatment 
sample_meta_figures <- sample.info %>% filter(sampleID%!in%c("S45","S60")) %>% mutate(treatment=factor(treatment, ordered = T,
                          levels=c("ambient","moderate_short","severe_short","moderate_long","severe_long"))) %>% 
                              arrange(treatment)
```

### Plots to look at acclimation response to short-term OA exposure

```{r, warning=F, message=F}
# res_amb_mod_short_shrink <- lfcShrink(dds_sva,
#     contrast = c("treatment", "ambient", "moderate_short"), 
#     res=res_amb_mod_short, type = 'normal')

res_amb_mod_short_shrink <- lfcShrink(
#    dds_sva, 
    dds.treatment, 
    coef = "treatment_moderate_short_vs_ambient", type = 'apeglm')

# res_amb_sev_short_shrink <- lfcShrink(dds_sva,
#     contrast = c("treatment", "ambient", "severe_short"), 
#     res=res_amb_sev_short, type = 'normal')

res_amb_sev_short_shrink <- lfcShrink(
#    dds_sva, 
    dds.treatment, 
    coef = "treatment_severe_short_vs_ambient", type = 'apeglm')

as.data.frame(res_amb_mod_short_shrink) %>% arrange(padj) %>% rownames_to_column("gene_id") %>% 
    mutate(log=-log10(padj)) %>% #left_join(snow.annot[c("gene_id","gene_name")]%>%distinct()) %>% 
    mutate(size = case_when(
    padj < 0.05 & abs(log2FoldChange) > 0.5 ~ 2.5,
    TRUE ~ 1.5))

(as.data.frame(res_amb_mod_short_shrink) %>% mutate(size = case_when(padj < 0.05 & abs(log2FoldChange) > 0.5 ~ 2.5,TRUE ~ 1.5)))$size
```

```{r, warning=F, message=F}
intersect(rownames(diffex_amb_mod_short_filt), rownames(diffex_amb_sev_short_filt))
intersect(rownames(subset(diffex_amb_mod_short_filt, abs(log2FoldChange) > 1)), rownames(subset(diffex_amb_sev_short_filt, abs(log2FoldChange) > 1)))

# probably need to remake this object
res_amb_mod_short_shrink

EnhancedVolcano(res_amb_mod_short_shrink,
    lab = (res_amb_mod_short_shrink%>%as.data.frame()%>%rownames_to_column("gene_id") %>% left_join(snow.annot[c("gene_id","gene_name")]%>%distinct()))$gene_name,
#    lab = rep("", nrow(res_amb_mod_short_shrink)),  # no labels
    x = 'log2FoldChange',
    y = 'padj',
#    title = '8hr Acclimation, pH 7.8',
    title=NULL,
    subtitle = NULL, caption=NULL,
    legendPosition = "none",
    pointSize = (as.data.frame(res_amb_mod_short_shrink) %>% mutate(size = case_when(padj < 0.05 ~ 1.5,TRUE ~ 1.5)))$size, # & Changeling(log2FoldChange) > 0.5
    pCutoff = 0.05,
    pCutoffCol = "padj",
    FCcutoff = 0.58496,
    labSize = 3,
    col=c('black', 'black', 'chocolate3', 'gold3'),
    colAlpha = 1,
    xlim = c(-2.5,3.5),
    ylim = c(0, 3))

EnhancedVolcano(res_amb_sev_short_shrink,
    lab = (res_amb_sev_short_shrink%>%as.data.frame()%>%rownames_to_column("gene_id") %>%
               left_join(snow.annot[c("gene_id","gene_name")]%>%distinct()))$gene_name,
#    lab = rep("", nrow(res_amb_sev_short_shrink)),  # no labels
    x = 'log2FoldChange',
    y = 'padj',
#    title = '8hr Acclimation, pH 7.5',
    title=NULL,
    subtitle = NULL, caption=NULL,
    legendPosition = "none",
    pCutoff = 0.05,
    pCutoffCol = "padj",
    FCcutoff = 0.58496,
    pointSize = (as.data.frame(res_amb_sev_short_shrink) %>% mutate(size = case_when(padj < 0.05 ~ 1.5,TRUE ~ 1.5)))$size, # & abs(log2FoldChange) > 0.5
    labSize = 5,
    col=c('black', 'black', 'coral4', 'red2'),
    colAlpha = 1,
    xlim = c(-2,2.75),
    ylim = c(0, 3.5))


# VOLCANO PLOTS FOR PAPER 
## This includes points in blue that are DEGs in both short-term OA treatments (compared to control)

## Step 1: define overlap genes
overlap_genes <- intersect(
  rownames(diffex_amb_mod_short_filt),
  rownames(diffex_amb_sev_short_filt))

overlap_genes

## Step 2: base color mapping (gray / brown / gold)
# turn DESeqResults into a data frame so we can work with it
df_mod <- as.data.frame(res_amb_mod_short_shrink) 
df_sev <- as.data.frame(res_amb_sev_short_shrink)

colCustom_mod <- rep("gray60", nrow(df_mod))  # default = nonsig
names(colCustom_mod) <- rownames(df_mod)

colCustom_mod[df_mod$padj < 0.05] <- "tan4"  # sig padj only
colCustom_mod[df_mod$padj < 0.05 & abs(df_mod$log2FoldChange) > 0.58496] <- "darkgoldenrod2"  # sig + big FC
colCustom_mod[overlap_genes] <- "dodgerblue3"  # force overlaps to blue

colCustom_sev <- rep("gray60", nrow(df_sev))
names(colCustom_sev) <- rownames(df_sev)

colCustom_sev[df_sev$padj < 0.05] <- "firebrick4"
colCustom_sev[df_sev$padj < 0.05 & abs(df_sev$log2FoldChange) > 0.58496] <- "red2"
colCustom_sev[overlap_genes] <- "dodgerblue3"

## Step 3: volcano plots
EnhancedVolcano(res_amb_mod_short_shrink,
    lab = rep("", nrow(res_amb_mod_short_shrink)),  # no labels
    # lab = ifelse(rownames(res_amb_mod_short_shrink) %in% overlap_genes,
    #              rownames(res_amb_mod_short_shrink), ""),
    x = 'log2FoldChange',
    y = 'padj',
    title = NULL, subtitle = NULL, caption = NULL,
    legendPosition = "none",
    pCutoff = 0.05,
    pCutoffCol = "padj",
    FCcutoff = 0.58496,
    labSize = 3,
    colCustom = colCustom_mod,
    colAlpha = 1,
    xlim = c(-2, 3.25),
    ylim = c(0, 3.1)
)

EnhancedVolcano(res_amb_sev_short_shrink,
    lab = rep("", nrow(res_amb_sev_short_shrink)),  # no labels
    # lab = ifelse(rownames(res_amb_sev_short_shrink) %in% overlap_genes,
    #              rownames(res_amb_sev_short_shrink), ""),
    x = 'log2FoldChange',
    y = 'padj',
    title = NULL, subtitle = NULL, caption = NULL,
    legendPosition = "none",
    pCutoff = 0.05,
    pCutoffCol = "padj",
    FCcutoff = 0.58496,
    labSize = 3,
    colCustom = colCustom_sev,
    colAlpha = 1,
    xlim = c(-2, 3.25),
    ylim = c(0, 3.1)
)

as.data.frame(diffex_amb_mod_short_filt)%>%rownames_to_column("gene_id") %>% filter(gene_id=="GWK47_003819")
as.data.frame(diffex_amb_sev_short_filt)%>%rownames_to_column("gene_id") %>% filter(gene_id=="GWK47_003819")

as.data.frame(diffex_amb_mod_short_filt)%>%rownames_to_column("gene_id") %>% filter(gene_id=="GWK47_003819") #weird, padj is NA 
as.data.frame(res_amb_sev_short)%>%rownames_to_column("gene_id") %>% filter(gene_id=="GWK47_003819") #weird, padj is NA 

mcols(res_amb_sev_short)$filterDecision[res_amb_sev_short$gene_id == "GWK47_003819"]

```

### Export all DEGs to add to Google Spreadsheet

```{r}
# Higher expression in moderate OA, 8hr compared to control 
as.data.frame(res_amb_mod_short_shrink) %>% 
    rownames_to_column("gene_id") %>% 
    filter(gene_id %in% rownames(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange < 0,])) %>% 
    arrange(desc(log2FoldChange)) %>% 
    left_join(snow.annot%>%dplyr::select(-cds, -protein_id, -GO, -gene_uni:-length, -entry_name, -gene_names:-organism, -gene_ontology_cellular_component:-gene_ontology_i_ds) %>%
                  group_by(gene_id)%>%slice_min(evalue) %>% distinct(), by="gene_id") %>% 
    filter(gene_name=="doxa-1")%>%
    write_clip(allow_non_interactive = T)

# Higher expression in moderate OA, 8hr compared to control 
as.data.frame(res_amb_mod_short_shrink) %>% 
    rownames_to_column("gene_id") %>% 
    filter(gene_id %in% rownames(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange > 0,])) %>% 
    arrange(desc(log2FoldChange)) %>% 
    left_join(snow.annot%>%dplyr::select(-cds, -protein_id, -GO, -gene_uni:-length, -entry_name, -gene_names:-organism, -gene_ontology_cellular_component:-gene_ontology_i_ds) %>%
                  group_by(gene_id)%>%slice_min(evalue) %>% distinct(), by="gene_id")  %>% 
    write_clip(allow_non_interactive = T)

# Higher expression in severe OA, 8hr compared to control 
as.data.frame(res_amb_sev_short_shrink) %>% 
    rownames_to_column("gene_id") %>% 
    filter(gene_id %in% rownames(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange < 0,])) %>% 
    arrange(desc(log2FoldChange)) %>% 
    left_join(snow.annot%>%dplyr::select(-cds, -protein_id, -GO, -gene_uni:-length, -entry_name, -gene_names:-organism, -gene_ontology_cellular_component:-gene_ontology_i_ds) %>%
                  group_by(gene_id)%>%slice_min(evalue) %>% distinct(), by="gene_id") %>%
    write_clip(allow_non_interactive = T)

# Lower expression in severe OA, 8hr compared to control 
as.data.frame(res_amb_sev_short_shrink) %>% 
    rownames_to_column("gene_id") %>% 
    filter(gene_id %in% rownames(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange > 0,])) %>% 
    arrange(desc(log2FoldChange)) %>% 
    left_join(snow.annot%>%dplyr::select(-cds, -protein_id, -GO, -gene_uni:-length, -entry_name, -gene_names:-organism, -gene_ontology_cellular_component:-gene_ontology_i_ds) %>%
                  group_by(gene_id)%>%slice_min(evalue) %>% distinct(), by="gene_id")  %>% 
    write_clip(allow_non_interactive = T)
```

Figure out names of genes with high LFC and -LOGP

```{r}
# Moderate OA 
as.data.frame(res_amb_mod_short_shrink) %>% rownames_to_column("gene_id") %>% 
    mutate(lfc=abs(log2FoldChange)) %>% arrange(desc(log2FoldChange)) %>% 
    filter(lfc>1, padj<0.05) %>% 
    mutate(log=-log10(padj)) %>% left_join(snow.annot[c("gene_id","gene_name", "protein")]%>%distinct())

# Severe OA 
as.data.frame(res_amb_sev_short_shrink) %>% rownames_to_column("gene_id") %>% 
    mutate(lfc=abs(log2FoldChange)) %>% arrange(desc(log2FoldChange)) %>% 
    filter(lfc>1, padj<0.05) %>% 
    mutate(log=-log10(padj)) %>% left_join(snow.annot[c("gene_id","gene_name", "protein")]%>%distinct())
```

Make sure I know which l2fc sign reflects higher / lower expression in each OA treatment 

```{r}
# diffex_amb_mod_short_filt #L2FC>0 = lower in moderate OA, L2FC<0 = higher in OA
# diffex_amb_sev_short_filt #L2FC>0 = lower in severe OA, L2FC<0 = higher in OA
```


```{r}
ggvenn(list(
  `Moderate - 8Hr` = row.names(subset(diffex_amb_mod_short_filt, abs(log2FoldChange)>0.5)),
  `Severe - 8Hr`  = row.names(subset(diffex_amb_sev_short_filt, abs(log2FoldChange)>0.5))),
       fill_color = c("yellow", "red", "gray"),
       stroke_size = 0.5, set_name_size = 4, show_percentage = F) + ggtitle("Moderate vs. Severe, 8Hr, all")

ggvenn(list(
  `Moderate - 8Hr` = row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange>0,]),
  `Severe - 8Hr`  = row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange>0,])), 
       fill_color = c("yellow", "red", "gray"),
       stroke_size = 0.5, set_name_size = 4, show_percentage = F) + ggtitle("Moderate vs. Severe, 8Hr, lower expression")

ggvenn(list(
  `Moderate - 8Hr` = row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange<0,]),
  `Severe - 8Hr`  = row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange<0,])), 
       fill_color = c("yellow", "red", "gray"),
       stroke_size = 0.5, set_name_size = 4, show_percentage = F) + ggtitle("Moderate vs. Severe, 8Hr, higher expression")


# diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange>0,] #lower expression in moderate OA compared to ambient
# diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange<0,] #higher expression in moderate OA compared to ambient
```

## Enrichment analyses using enricher

```{r, warning=F, message=F}
# All short term moderate OA genes 
summary(res_short_mod_enrich <- enricher(
  gene      = row.names(diffex_amb_mod_short_filt),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=2)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# Higher expression, moderate short term OA 
summary(res_short_up_mod_enrich <- enricher(
#  gene      = row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange< -0.5,]),
  gene      = row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange<0,]),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=2)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# Lower expression, moderate short term OA 
summary(res_short_down_mod_enrich <- enricher(
  gene      = row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange> 0.5,]),
#  gene      = row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange>0,]),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=2)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

as.data.frame(diffex_amb_mod_short_filt) %>% rownames_to_column("gene_id") %>% filter(gene_id %in% c("GWK47_032391","GWK47_039245","GWK47_046849","GWK47_010349","GWK47_001655")) %>% 
    left_join(snow.annot, by="gene_id")

# Add to DAVID list 
david.lists[["mod_short_all"]] <- data.frame("gene_id"=row.names(diffex_amb_mod_short_filt[abs(diffex_amb_mod_short_filt$log2FoldChange)>0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["mod_short_up"]] <- data.frame("gene_id"=row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange< -0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["mod_short_down"]] <- data.frame("gene_id"=row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange> 0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["short_OA_all"]] <- data.frame(
    "gene_id"=row.names(diffex_OA_short_filt[abs(diffex_OA_short_filt$log2FoldChange)>0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["short_OA_up_sev"]] <- data.frame(
    "gene_id"=row.names(diffex_OA_short_filt[diffex_OA_short_filt$log2FoldChange > 0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["short_OA_down_sev"]] <- data.frame(
    "gene_id"=row.names(diffex_OA_short_filt[diffex_OA_short_filt$log2FoldChange < 0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()
```

```{r}
# --- severe 

# All short term severe OA genes 
summary(res_short_sev_enrich <- enricher(
  gene      = row.names(diffex_amb_sev_short_filt[abs(diffex_amb_sev_short_filt$log2FoldChange)>0,]),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=2)) %>% 
    filter(p.adjust<0.05) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything()) 

# Higher expression, severe short term OA 
summary(res_short_up_sev_enrich <- enricher(
  gene      = row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange< 0,]),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=2)) %>% 
    filter(p.adjust<0.05) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# Lower expression, severe short term OA 
summary(res_short_down_sev_enrich <- enricher(
  gene      = row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange> 0,]),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=2)) %>% 
    filter(p.adjust<0.05) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# Add to DAVID list 
david.lists[["sev_short_all"]] <- data.frame("gene_id"=row.names(diffex_amb_sev_short_filt[abs(diffex_amb_sev_short_filt$log2FoldChange)>0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["sev_short_up"]] <- data.frame("gene_id"=row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange< 0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["sev_short_down"]] <- data.frame("gene_id"=row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange> 0,])) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()
```


```{r}
# Common DEGs 
(res_short_intersect <- data.frame(gene_id=intersect(row.names(diffex_amb_mod_short_filt[abs(diffex_amb_mod_short_filt$log2FoldChange)>0,]),
                                                     row.names(diffex_amb_sev_short_filt[abs(diffex_amb_sev_short_filt$log2FoldChange)>0,]))) %>%
    left_join(snow.annot%>%dplyr::select(-cds,-GO))%>%distinct()) %>% filter(!is.na(gene_name))

(res_short_up_intersect <- data.frame(gene_id=intersect(row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange< 0,]),
                                                        row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange< 0,]))) %>%
    left_join(snow.annot%>%dplyr::select(-cds,-GO))%>%distinct()) %>% filter(!is.na(gene_name))

(res_short_down_intersect <- data.frame(gene_id=intersect(row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange> 0,]),
                                                        row.names(diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange> 0,]))) %>%
    left_join(snow.annot%>%dplyr::select(-cds,-GO))%>%distinct()) %>% filter(!is.na(gene_name))

# All common DEGs
summary(res_short_intersect_enrich <- enricher(
  gene      = res_short_intersect$gene_id,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=1)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# DEGs both upregulated
summary(res_short_up_intersect_enrich <- enricher(
  gene      = res_short_up_intersect$gene_id,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=1)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# DEGs both downregulated
summary(res_short_down_intersect_enrich <- enricher(
  gene      = res_short_down_intersect$gene_id,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=1)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# Add to DAVID list 
david.lists[["intersect_short_all"]] <- data.frame("gene_id"=res_short_intersect$gene_id) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["intersect_short_up"]] <- data.frame("gene_id"=res_short_up_intersect$gene_id) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["intersect_short_down"]] <- data.frame("gene_id"=res_short_down_intersect$gene_id) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()
```


```{r}
# diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange>0,] #lower expression in moderate OA compared to ambient
# diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange<0,] #higher expression in moderate OA compared to ambient
# diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange>0,] #lower expression in moderate OA compared to ambient
# diffex_amb_sev_short_filt[diffex_amb_sev_short_filt$log2FoldChange<0,] #higher expression in moderate OA compared to ambient
# 
# diffex_OA_long_filt[diffex_OA_long_filt$log2FoldChange>0,] #higher expression in moderate 
# diffex_OA_long_filt[diffex_OA_long_filt$log2FoldChange<0,] #higher expression in severe 
```


### Which genes are both DEGs at 8-hr AND 84-days? 

#### Moderate OA 

```{r}
ggvenn(list(
  `Moderate - 8Hr` = row.names(subset(diffex_amb_mod_short_filt, abs(log2FoldChange)>0)),
  `Moderate - 84 days`  = row.names(subset(diffex_amb_mod_long_filt, abs(log2FoldChange)>0))),
       fill_color = c("yellow", "#ffc043"),
       stroke_size = 0.5, set_name_size = 4, show_percentage = F) #+ ggtitle("Moderate vs. Severe, 8Hr")

# Genes differentially expressed in both moderate at 8hr and 84 days
counts_moderate_time <- as.data.frame(
    Y
    #Y_adj
    ) %>% 
    rownames_to_column("gene") %>% 
    filter(gene %in% 
               (data.frame(gene_id=intersect(row.names(diffex_amb_mod_short_filt),row.names(diffex_amb_mod_long_filt)))%>%
    left_join(snow.annot%>%dplyr::select(-cds,-GO)) %>% distinct())$gene_id) %>% 
    left_join(snow.annot%>%dplyr::select(-cds,-GO),by=c("gene"="gene_id")) %>% distinct() %>% 
    filter(!is.na(gene_name)) %>% 
    mutate(label=paste0(gene_name," (",protein,")")) %>% 
#    column_to_rownames("label")
    column_to_rownames("protein")

(res_moderate_intersect <- data.frame(
    gene_id=intersect(row.names(subset(diffex_amb_mod_short_filt, abs(log2FoldChange)>0)), 
                      row.names(subset(diffex_amb_mod_long_filt, abs(log2FoldChange)>0)))) %>%
    left_join(snow.annot%>%dplyr::select(-cds,-GO, -protein_id) %>% group_by(gene_id, gene_name, protein)%>%slice_min(evalue))%>%distinct()) %>% #filter(!is.na(gene_name)) %>% 
    dplyr::select(gene_id, gene_name:protein,spid,gene_ontology_biological_process) #%>% write_clip(allow_non_interactive = T)

summary(res_moderate_intersect_enrich <- enricher(
  gene      = res_moderate_intersect$gene_id,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=0)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

#res_moderate_intersect%>%filter(!is.na(spid))%>%dplyr::select(spid) %>% write_clip(allow_non_interactive = T)

pheatmap(counts_moderate_time %>% dplyr::select((sample_meta_figures %>%
                filter(treatment%in%c("ambient","moderate_short","moderate_long")))$sampleID), 
         cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=TRUE, na.rm=TRUE, annotation_colors = list(treatment=color_scale),
         scale="row", main = "Persistent DEGs in Moderate OA (pH 7.8)", 
         annotation_col=(sample_meta_figures %>%
                             filter(treatment%in%c("ambient","moderate_short","moderate_long")) %>%
                             column_to_rownames("sampleID"))[,"treatment", drop=FALSE], 
         fontsize = 8, cutree_rows = 3, 
         border_color=F)

# Add to DAVID list 
david.lists[["mod_persist_all"]] <- res_moderate_intersect %>% group_by(gene_id)%>%slice_min(evalue)
```

#### Severe OA 

```{r}
ggvenn(list(
  `Severe - 8Hr` = row.names(subset(diffex_amb_sev_short_filt, abs(log2FoldChange)>0)),
  `Severe - 84 days`  = row.names(subset(diffex_amb_sev_long_filt, abs(log2FoldChange)>0))),
       fill_color = c("yellow", "#ffc043"),
       stroke_size = 0.5, set_name_size = 4, show_percentage = F) #+ ggtitle("Moderate vs. Severe, 8Hr")

# Genes differentially expressed in both severe at 8hr and 84 days

counts_severe_time <- as.data.frame(Y_adj) %>% 
    rownames_to_column("gene") %>% 
    filter(gene %in% 
               unique(data.frame(gene_id=intersect(row.names(diffex_amb_sev_short_filt),row.names(diffex_amb_sev_long_filt))) %>%
    left_join(snow.annot%>%dplyr::select(-cds,-GO)) %>% distinct())$gene_id) %>% 
    left_join(snow.annot%>%dplyr::select(-cds,-GO)%>%group_by(gene_id)%>%slice_min(evalue),by=c("gene"="gene_id")) %>% distinct() %>% 
    filter(!is.na(gene_name)) %>% 
    mutate(label=paste0(protein, "(", gene, ")")) %>% 
    dplyr::select(-protein_id:-entry_name) %>% distinct() %>%
#    column_to_rownames("label")
    column_to_rownames("protein")

(res_severe_intersect <- data.frame(
    gene_id=intersect(row.names(subset(diffex_amb_sev_short_filt, abs(log2FoldChange)>0)), 
                      row.names(subset(diffex_amb_sev_long_filt, abs(log2FoldChange)>0)))) %>%
    left_join(snow.annot%>%dplyr::select(-cds,-GO, -protein_id) %>% group_by(gene_id, gene_name, protein)%>%slice_min(evalue))%>%distinct()) %>% #filter(!is.na(gene_name)) %>% 
    dplyr::select(gene_id, gene_name:protein,spid,gene_ontology_biological_process) %>% write_clip(allow_non_interactive = T)

#res_severe_intersect%>%filter(!is.na(spid))%>%dplyr::select(spid) %>% write_clip(allow_non_interactive = T)

pheatmap(counts_severe_time %>% dplyr::select((sample_meta_figures %>%
                filter(treatment%in%c("ambient","severe_short","severe_long")))$sampleID), 
         cluster_rows=TRUE, cluster_cols=FALSE, 
         show_rownames=TRUE, na.rm=TRUE, annotation_colors = list(treatment=color_scale),
         scale="row", main = "Persistent DEGs in Severe OA (pH 7.5)", 
         annotation_col=(sample_meta_figures %>%
                             filter(treatment%in%c("ambient","severe_short","severe_long")) %>%
                             column_to_rownames("sampleID"))[,"treatment", drop=FALSE], 
         fontsize = 8, cutree_rows = 4, 
         border_color=F)

summary(res_severe_intersect_enrich <- enricher(
  gene      = res_severe_intersect$gene_id,
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=0)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything())

# Add to DAVID list 
david.lists[["sev_persist_all"]] <- res_severe_intersect %>% group_by(gene_id)%>%slice_min(evalue)
```

```{r}
ggvenn(list(
  `Severe` = res_severe_intersect$gene_id,
  `Moderate`  = res_moderate_intersect$gene_id), 
       fill_color = c("yellow", "#ffc043"),
       stroke_size = 0.5, set_name_size = 4, show_percentage = F) #+ ggtitle("Moderate vs. Severe, 8Hr")
```

```{r}
# res_OA_long_shrink%>%as.data.frame()%>% arrange(padj) %>% mutate(log=-log10(padj)) %>% 
#     rownames_to_column("gene_id")%>%left_join(snow.annot[c("gene_id","gene_name","protein","spid","evalue", "gene_names","protein_names","gene_ontology_biological_process")]%>%distinct()) %>% 
#     filter(gene_id%in%c("GWK47_024010","GWK47_025037"))
# #    filter(padj<0.05, log2FoldChange>0.5) %>% dplyr::select(gene_name,protein) %>% write_clip(allow_non_interactive = T)
```

### Differences among moderate, severe OA at 84 days - why might they die in severe?  Biomarkers? 

```{r, warning=F, message=F}
# How many DEGs at day 88? 

# LOWER EXPRESSION IN SEVERE OA, higher in moderate OA
subset(diffex_OA_long_filt, log2FoldChange>0) %>% as.data.frame() %>% 
    rownames_to_column("gene_id") %>% 
    left_join(snow.annot%>%dplyr::select(-cds,-GO, -protein_id) %>% group_by(gene_id, gene_name)%>%slice_min(evalue) %>% distinct(), "gene_id") %>% #filter(!is.na(gene_name)) %>% 
    dplyr::select(gene_id:spid,protein_names,gene_ontology_biological_process) 

# lower expression in moderate OA, higher in severe OA 
subset(diffex_OA_long_filt, log2FoldChange<0) %>% as.data.frame() %>% 
    rownames_to_column("gene_id") %>% 
    left_join(snow.annot%>%dplyr::select(-cds,-GO, -protein_id) %>% group_by(gene_id, gene_name)%>%slice_min(evalue) %>% distinct(), "gene_id") %>% #filter(!is.na(gene_name)) %>% 
    dplyr::select(gene_id:spid,protein_names,gene_ontology_biological_process) 
```

### FIGURE 4

```{r, warning=F, message=F }
# res_OA_long_shrink <- lfcShrink(dds_sva,
#     contrast = c("treatment", "moderate_long", "severe_long"), 
#     res=res_amb_mod_short, type = 'normal')

# Relevel to  now include "treatment_moderate_long_vs_severe_long"
dds_sva$treatment <- relevel(dds_sva$treatment, "moderate_long")
dds_sva <- DESeq(dds_sva)
resultsNames(dds_sva)

res_OA_long_shrink <- lfcShrink(dds_sva,
    coef = "treatment_severe_long_vs_moderate_long", type = 'apeglm')

res_OA_long_shrink_df <- as.data.frame(res_OA_long_shrink)

# Make a named vector of colors for each gene
colCustom <- rep("grey70", nrow(res_OA_long_shrink_df))
names(colCustom) <- rownames(res_OA_long_shrink_df)

# ---- Color assignments ----
# Significant, positive
colCustom[res_OA_long_shrink_df$padj < 0.05 & res_OA_long_shrink_df$log2FoldChange > 0.5]  <- "red"         # bright red
colCustom[res_OA_long_shrink_df$padj < 0.05 & res_OA_long_shrink_df$log2FoldChange > 0 & 
           res_OA_long_shrink_df$log2FoldChange <= 0.5]                                     <- "darkred"     # dark red

# Significant, negative
colCustom[res_OA_long_shrink_df$padj < 0.05 & res_OA_long_shrink_df$log2FoldChange < -0.5] <- "blue"  # bright blue
colCustom[res_OA_long_shrink_df$padj < 0.05 & res_OA_long_shrink_df$log2FoldChange < 0 & 
           res_OA_long_shrink_df$log2FoldChange >= -0.5]                                   <- "navy"        # dark blue

# Not significant, large fold change
colCustom[res_OA_long_shrink_df$padj > 0.05 & abs(res_OA_long_shrink_df$log2FoldChange) > 0.5] <- "grey50" # light gray

# Not significant, small fold change
colCustom[res_OA_long_shrink_df$padj > 0.05 & abs(res_OA_long_shrink_df$log2FoldChange) <= 0.5] <- "grey30"   # dark gray

# ---- Volcano plot ----
EnhancedVolcano(
  res_OA_long_shrink_df,
# lab = rep("", nrow(res_OA_long_shrink)), # no labels
  lab = res_OA_long_shrink_df %>%
    rownames_to_column("gene_id") %>%
    left_join(snow.annot[c("gene_id","gene_name")] %>% distinct()) %>%
    pull(gene_name),
  x = 'log2FoldChange',
  y = 'padj',
#  title = 'DEGs among OA Treatments at Day 84',
  title=NULL,
  subtitle = NULL,
  caption = NULL,
  legendPosition = "none",
  pointSize = (res_OA_long_shrink_df %>%
    mutate(size = case_when(
      padj < 0.05 & abs(log2FoldChange) > 0.5 ~ 2.25,
      TRUE ~ 1.5)))$size,
  drawConnectors = TRUE,
  max.overlaps = 50,
  widthConnectors = 0.15,
  colConnectors = "grey30",
  pCutoff = 0.05,
  FCcutoff = 0.5,
  labSize = 4,
  colCustom = colCustom,
  colAlpha = 1,
  xlim = c(-2, 2.25),
  ylim = c(0, 4.5)
)


counts_OA_long <- as.data.frame(Y_adj) %>% 
    rownames_to_column("gene") %>% 
    filter(gene %in% rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange<0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds,-GO)%>%group_by(gene_id)%>%slice_min(evalue),by=c("gene"="gene_id")) %>% distinct() %>% 
    filter(!is.na(gene_name)) %>% 
    mutate(label=paste0(protein," (",gene_name,")")) %>% 
    column_to_rownames("label")
#    column_to_rownames("protein")

pheatmap(counts_OA_long %>% dplyr::select((sample_meta_figures %>%
                filter(treatment%in%c("moderate_long","severe_long")))$sampleID), 
         cluster_rows=TRUE, cluster_cols=FALSE, 
         show_rownames=TRUE, na.rm=TRUE, annotation_colors = list(treatment=color_scale),
         scale="row", main = "DEGs among OA treatments at day 84", 
         annotation_col=(sample_meta_figures %>%
                             filter(treatment%in%c("moderate_long","severe_long")) %>%
                             column_to_rownames("sampleID"))[,"treatment", drop=FALSE], 
         fontsize = 7.5, cutree_rows = 2, 
         border_color=F)


# Genes with positive log2FoldChange are more highly expressed in pH 7.8 (moderate_long).
summary(res_OA_long_enrich <- enricher(
  gene      = rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange> 0)),
#  gene      = rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange> 0.5)),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=1)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything()) #%>% 
#    dplyr::select(ID, term_name)

# Genes with negative log2FoldChange are more highly expressed in pH 7.5 (severe_long).
summary(res_OA_long_enrich <- enricher(
  gene      = rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange< 0)),
#  gene      = rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange< -0.5)),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=1)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything()) #%>% 
#    dplyr::select(ID, term_name) %>% write_clip(allow_non_interactive = T)


# Add to DAVID list 
david.lists[["long_all"]] <- data.frame("gene_id"=rownames(as.data.frame(diffex_OA_long_filt)%>%filter(abs(log2FoldChange)>0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["long_up"]] <- data.frame("gene_id"=rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange< 0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["long_down"]] <- data.frame("gene_id"=rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange> 0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["long_up_0.5"]] <- data.frame("gene_id"=rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange< -0.5))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["long_down_0.5"]] <- data.frame("gene_id"=rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange> 0.5))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

# Export for Google Sheets and Table in Paper 
res_OA_long_shrink %>% write_clip(allow_non_interactive = T)

# LOWER EXPRESSION IN SEVERE OA, higher in moderate OA
as.data.frame(res_OA_long_shrink) %>% 
    rownames_to_column("gene_id") %>% 
    filter(gene_id %in% rownames(subset(diffex_OA_long_filt, log2FoldChange>0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds,-GO, -protein_id) %>% group_by(gene_id, gene_name)%>%slice_min(evalue) %>% distinct(), "gene_id") %>% #filter(!is.na(gene_name)) %>% 
    dplyr::select(gene_id:spid,protein_names,gene_ontology_biological_process) %>% 
    mutate(neg_logp=-log10(padj)) %>% 
    arrange(log2FoldChange) #%>% 
#    write_clip(allow_non_interactive = T)

# HIGHER EXPRESSION IN SEVERE OA, lower in moderate OA
as.data.frame(res_OA_long_shrink) %>% 
    rownames_to_column("gene_id") %>% 
    filter(gene_id %in% rownames(subset(diffex_OA_long_filt, log2FoldChange<0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds,-GO, -protein_id) %>% group_by(gene_id, gene_name)%>%slice_min(evalue) %>% distinct(), "gene_id") %>% #filter(!is.na(gene_name)) %>% 
    dplyr::select(gene_id:spid,protein_names,gene_ontology_biological_process) %>% 
    arrange(desc(log2FoldChange)) #%>% 
#    write_clip(allow_non_interactive = T)

```

### Dose-dependent analysis - not included in paper, but interesting nonetheless

Assess as three ordered levels at 8 h: Ctrl (ambient pH) → Mod (7.8) → Sev (7.5).
Instead of treating these as categorical levels, assign them numeric scores (0, 1, 2). DESeq2 then fits a regression slope for each gene. A Wald test on ph_num asks whether the slope is different from zero — i.e., whether expression systematically increases or decreases with severity.

Expression∼β0​+βph_num​×ph_num+βbatch​×batch+ϵ 

The res_trend gives: 
- log2 fold-change per step increase in severity (Ctrl→Mod, Mod→Sev). 
    - Example: log2FC = +1.0 → each step toward more severe pH doubles expression. So Ctrl→Sev overall is ~2 steps → log2FC ≈ +2.0.
- Genes with positive slope: monotonic up-regulation as pH drops.
- Genes with negative slope: monotonic down-regulation as pH drops.

```{r}
dds_short <- dds_sva[, dds_sva$treatment %in% c("ambient", "moderate_short", "severe_short")]
dds_short$pCO2 <- with(colData(dds_short),
                       ifelse(treatment == "ambient", 473,
                       ifelse(treatment == "moderate_short", 702, 1415)))
dds_short$pCO2 <- as.numeric(dds_short$pCO2)
#design(dds_short) <- ~ SV1 + SV2 + SV3 + SV4 + SV5 + SV6 + SV7 + pCO2
design(dds_short) <- ~ pCO2
dds_short <- DESeq(dds_short)

# results: slope coefficient for pco2
# Each gene’s log2FoldChange represents the change in log2 expression per unit increase in pCO₂ (μatm).
# Positive	Expression increases with rising CO₂ (i.e., more acidification)
# Negative	Expression decreases with rising CO₂

res_pCO2 <- results(dds_short, name = "pCO2")

# Top genes that increase with pCO2
as.data.frame(res_pCO2) %>% arrange(padj) %>% filter(padj<0.01) %>% rownames_to_column("gene_id") %>% 
    filter(stat>0) %>% 
        left_join(snow.annot%>%dplyr::select(-cds, -protein_id, -GO) %>% group_by(gene_id) %>% slice_min(evalue) %>% distinct())  

# Top genes that decrease with pCO2
as.data.frame(res_pCO2) %>% arrange(padj) %>% filter(padj<0.05) %>% rownames_to_column("gene_id") %>% 
    filter(stat<0) %>% 
        left_join(snow.annot%>%dplyr::select(-cds, -protein_id, -GO) %>% group_by(gene_id) %>% slice_min(evalue) %>% distinct())


# Genes with positive stat increase with pco2 (highest in severe OA, lowest in control)
summary(res_short_dose_increase_enrich <- enricher(
  gene      = rownames(as.data.frame(res_pCO2)%>%filter(padj<0.05, stat>0)),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=1)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything()) #%>% 
#    dplyr::select(ID, term_name)

# Genes with negative stat decrease with pco2  (highest in control, lowest in severe OA)
summary(res_short_dose_decrease_enrich <- enricher(
  gene      = rownames(as.data.frame(res_pCO2)%>%filter(padj<0.05, stat<0)),
  universe  = universe,
  TERM2GENE = TERM2GENE,
  pAdjustMethod = "fdr",
  minGSSize=1)) %>% 
    filter(p.adjust<0.1) %>% filter(Count>1) %>% 
    left_join(TERM2NAME, by = c("ID"="GO")) %>% 
    dplyr::select(ID, term_name, everything()) #%>% 
#    dplyr::select(ID, term_name)

# Add to DAVID list 
david.lists[["dose_increase_with_pco2"]] <- data.frame("gene_id"=rownames(as.data.frame(res_pCO2)%>%filter(padj<0.05, stat>0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()

david.lists[["dose_decrease_with_pco2"]] <- data.frame("gene_id"=rownames(as.data.frame(res_pCO2)%>%filter(padj<0.05, stat<0))) %>% 
    left_join(snow.annot%>%dplyr::select(-cds)%>%group_by(gene_id)%>%slice_min(evalue)) %>% 
    select(gene_id:protein,spid,protein_names,evalue) %>% distinct()
```

```{r}
david.lists$sev_short_all %>% filter(grepl("slc",gene_name)) 
david.lists$mod_short_all %>% filter(grepl("slc",gene_name)) 
david.lists$short_OA_all %>% filter(grepl("TCI",gene_name)) 
```

### Save lists of genes as .bed files - to explore in other studies! 

```{r}
(gene_coordinates <- read.gff(file = "../references/snowcrab_ASM1658430v1.gtf", GFF3 = F) %>% filter(feature=="gene") %>%
  mutate(gene_id=gsub("gene_id |;|\"", "", str_extract(attributes, "gene_id (.*?);"))) %>% 
    dplyr::select(gene_id, seqname, start, end, strand))

david.lists$sev_short_up %>% left_join(gene_coordinates, by="gene_id") %>% 
    mutate(start_up2kb=start-2000) %>% mutate(start_up2kb=case_when(start_up2kb<0~0, T~start_up2kb))

snow.OA.gene.lists <- map(david.lists, ~ {
  if (is.null(.x) || nrow(.x) == 0) return(NULL)
  .x %>%
    left_join(gene_coordinates, by = "gene_id") %>%
    mutate(start_up2kb = pmax(start - 2000, 0))})

save(snow.OA.gene.lists, file="../results/snow.OA.gene.lists") #Save as R list object
```

## DAVID Enrichment 

```{r}
david.list.vectors <- list(
        "pc1_pos_loadings"=(david.lists$pc1_pos_loadings %>% filter(!is.na(spid)))$spid,
        "pc1_neg_loadings"=(david.lists$pc1_neg_loadings %>% filter(!is.na(spid)))$spid,
        "pc2_pos_loadings"=(david.lists$pc2_pos_loadings %>% filter(!is.na(spid)))$spid,
        "pc2_neg_loadings"=(david.lists$pc2_neg_loadings %>% filter(!is.na(spid)))$spid,
        "short_OA_all"=(david.lists$short_OA_all %>% filter(!is.na(spid)))$spid,
        "short_OA_up_sev"=(david.lists$short_OA_up_sev %>% filter(!is.na(spid)))$spid,
        "short_OA_down_sev"=(david.lists$short_OA_down_sev %>% filter(!is.na(spid)))$spid,
        "mod_short_all"=(david.lists$mod_short_all %>% filter(!is.na(spid)))$spid,
        "mod_short_up"=(david.lists$mod_short_up %>% filter(!is.na(spid)))$spid,
        "mod_short_down"=(david.lists$mod_short_down %>% filter(!is.na(spid)))$spid,
        "sev_short_all"=(david.lists$sev_short_all %>% filter(!is.na(spid)))$spid,
        "sev_short_up"=(david.lists$sev_short_up %>% filter(!is.na(spid)))$spid,
        "sev_short_down"=(david.lists$sev_short_down %>% filter(!is.na(spid)))$spid,
        "intersect_short_all"=(david.lists$intersect_short_all %>% filter(!is.na(spid)))$spid,
        "intersect_short_up"=(david.lists$intersect_short_up %>% filter(!is.na(spid)))$spid,
        "intersect_short_down"=(david.lists$intersect_short_down %>% filter(!is.na(spid)))$spid,
        "dose_short_increase"=(david.lists$dose_increase_with_pco2 %>% filter(!is.na(spid)))$spid,
        "dose_short_decrease"=(david.lists$dose_decrease_with_pco2 %>% filter(!is.na(spid)))$spid,
        "mod_persist_all"=(david.lists$mod_persist_all %>% filter(!is.na(spid)))$spid,
        "sev_persist_all"=(david.lists$sev_persist_all %>% filter(!is.na(spid)))$spid,
        "long_all"=(david.lists$long_all %>% filter(!is.na(spid)))$spid,
        "long_up"=(david.lists$long_up %>% filter(!is.na(spid)))$spid,
        "long_down"=(david.lists$long_down %>% filter(!is.na(spid)))$spid,
        "long_up_0.5"=(david.lists$long_up_0.5 %>% filter(!is.na(spid)))$spid,
        "long_down_0.5"=(david.lists$long_down_0.5 %>% filter(!is.na(spid)))$spid
        )#,
#        "background"=david.lists$background %>% filter(!is.na(spid)))$spid)

# Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs 
write_delim(x = plyr::ldply(david.list.vectors, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", file = "../results/enrichment/David_uniprot_all_DEGs.txt")

# Background list - copy to clipboard, paste into DAVID
(david.lists$background %>% filter(!is.na(spid)))$spid %>% write_clip(allow_non_interactive = T)
```

### Here I used the multilist file upload in DAVID with Uniprot SPID as identifiers, and the background set of genes copied to clipboard then pasted into DAVID 

## DAVID Results 

```{bash}
ls ../results/enrichment/DAVID*.csv > ../results/enrichment/DAVID_files.txt
```

```{r}
david_files <- read_delim("../results/enrichment/DAVID_files.txt", col_names = c("filename"), delim="\t") %>%
    mutate(gene_set=gsub("../results/enrichment/DAVIDChartReport_|_2025-11-03.csv", "", filename)) %>% 
    mutate(time=case_when(grepl("short", gene_set)~"short", grepl("long", gene_set)~"long", grepl("persist", gene_set)~"persist", TRUE~NA)) %>% 
    mutate(OA=case_when(grepl("mod", gene_set)~"mod", grepl("sev", gene_set)~"sev", TRUE~NA)) %>% 
    mutate(response=case_when(grepl("up", gene_set)~"up", grepl("down", gene_set)~"down", grepl("all", gene_set)~"all", 
                              grepl("increase", gene_set)~"increase", grepl("decrease", gene_set)~"decrease", TRUE~NA))
    
```

```{r, warning=F, message=F}
file_list <- vector(mode = "list", length = nrow(david_files))
names(file_list) <- c(david_files$gene_set)

for (i in 1:nrow(david_files)) {
#    file_list[[i]] <- data.frame(read_delim(file=paste("../results/enrichment/L2FC_0.5/", filenames[i,"filename"], sep=""), delim = ",")) %>% 
    file_list[[i]] <- data.frame(read_delim(file=david_files$filename[i], delim = ",")) %>% 
        mutate(time=david_files$time[i], OA=david_files$OA[i], response=david_files$response[i])}

david <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    #dplyr::rename(percent=x) %>%
    separate(term, into = c("term_temp", "process_temp"), sep = "~") %>%
    mutate(process=case_when(
      is.na(process_temp) == FALSE ~ process_temp,
      is.na(process_temp) == TRUE ~ str_remove(term_temp, ".*:"))) %>% 
      mutate(term=case_when(
      is.na(process_temp) == FALSE ~ term_temp,
      is.na(process_temp) == TRUE ~ str_remove(term_temp, ":.*"))) %>%
#    separate(gene_set, sep = "_", into = c("stress", "filtering"), remove = F) %>%
    # mutate(stress=factor(stress, ordered = TRUE, levels=c("Cold","Cool", "Warm")),
    #        filtering=as.factor(filtering)) %>% 
  tidyr::complete(gene_set, category)  # add rows for gene_sets that are missing from dataframe

save(david, file="../results/enrichment/david")
```

### PC Loadings 

```{r}
david %>% 
  filter(grepl("pc", gene_set)) %>% 
  filter(fdr<0.1) %>% 
  filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT")) %>% #,"UP_KW_CELLULAR_COMPONENT","UP_KW_MOLECULAR_FUNCTION", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT"
#  filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT","UP_KW_CELLULAR_COMPONENT","UP_KW_MOLECULAR_FUNCTION", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT")) %>% 
    mutate(across(c(p_value, fdr), signif, 2)) %>%
  dplyr::select(gene_set, category, fdr, term, count, genes) %>% ungroup() %>% 
  dplyr::arrange(gene_set, fdr)
```

### View or copy all enrichment results to clipboard for Google Sheet

```{r}
david %>%
  filter(fdr<0.1) %>% 
#  filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT")) %>%
    filter(category %in% c("UP_KW_CELLULAR_COMPONENT","UP_KW_MOLECULAR_FUNCTION", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT")) %>% 
    write_clip(allow_non_interactive = T)
```

### Short term response 

```{r}
david %>% 
#  filter(grepl("mod_short", gene_set)) %>% 
#  filter(grepl("sev_short_down|sev_short_up", gene_set)) %>% 
  filter(grepl("intersect", gene_set)) %>% 
  filter(fdr<0.1) %>% 
  filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT")) %>% #,"UP_KW_CELLULAR_COMPONENT","UP_KW_MOLECULAR_FUNCTION", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT"
#  filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT","UP_KW_CELLULAR_COMPONENT","UP_KW_MOLECULAR_FUNCTION", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT")) %>% 
    mutate(across(c(p_value, fdr), signif, 2)) %>%
  dplyr::select(gene_set, category, fdr, term, count, genes) %>% ungroup() %>% 
  dplyr::arrange(gene_set, fdr)
```

### Long term differences

```{r}
david %>%
  filter(grepl("long_up", gene_set)) %>% 
  filter(fdr<0.1) %>% 
  filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT")) %>% #,"UP_KW_CELLULAR_COMPONENT","UP_KW_MOLECULAR_FUNCTION", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT"
#  filter(category%in%c("UP_KW_BIOLOGICAL_PROCESS","GOTERM_BP_DIRECT","UP_KW_CELLULAR_COMPONENT","UP_KW_MOLECULAR_FUNCTION", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT")) %>% 
    mutate(across(c(p_value, fdr), signif, 2)) %>%
  dplyr::select(gene_set, category, fdr, term, count, genes) %>% ungroup() %>% 
  dplyr::arrange(gene_set, fdr) %>% View()
```

```{r}
david %>%
  filter(!grepl("all", gene_set)) %>% 
  filter(count>=2) %>% 
  filter(fdr<0.1) %>%
  filter(category %in% c("GOTERM_BP_DIRECT", "UP_KW_BIOLOGICAL_PROCESS")) %>%  
#  filter(category %in% c("GOTERM_BP_DIRECT")) %>%  
#  filter(category %in% c("UP_KW_BIOLOGICAL_PROCESS")) %>% 
  group_by(gene_set, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(gene_set, category, term, process, count, p_value, fold_enrichment, fdr, genes), 
            by = c("gene_set", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, filtering, p_value) %>% group_by(stress, filtering) %>% dplyr::slice(1:15) %>%
  ungroup() %>% arrange(rev(gene_set), p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>% #,
#         gene_set=factor(gene_set, ordered = TRUE, levels=c("Warm", "Cool", "Cold"))) %>%

ggplot(aes(y = process, x=gene_set, col=gene_set)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 # scale_color_manual(name="Temperature", values=c("Cold"="royalblue1","Cool"="darkgreen", "Warm"="firebrick4"),
 #                                guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=9), 
        plot.title = element_text(size=9),
        legend.text=element_text(size=7), 
        legend.title = element_blank(),
#        strip.text.x = element_blank(),
        strip.background = element_blank()) +
  ggtitle("Enriched Biological Processes")

```

# Boxplots, Venn Diagrams, Heatmaps 

## Save objects for interactive Shiny App

```{r}
saveRDS(assay(vsd.treatment), file = "../snow_crab_expression_explorer/vsd_matrix.rds") #variance-stabilization normalized counts 
saveRDS(Y_adj, file = "../snow_crab_expression_explorer/Y_adj.rds") #surrogate variable - regressed variance-stabilization normalized counts 
saveRDS(sample.info, file = "../snow_crab_expression_explorer/sample_info.rds") #sample metadata 
saveRDS(snow.annot, file = "../snow_crab_expression_explorer/snow_annot.rds") #snow crab genome annotation info 

# Build combined DEG info table with short, readable comparison labels
deg_info <- tibble(
  gene_id = c(
    rownames(diffex_amb_mod_short_filt),
    rownames(diffex_amb_sev_short_filt),
    rownames(diffex_OA_short_filt),
    rownames(diffex_amb_mod_long_filt),
    rownames(diffex_amb_sev_long_filt),
    rownames(diffex_OA_long_filt)
  ),
  comparison = c(
    rep("Ctrl. vs. Mod OA (8h)", nrow(diffex_amb_mod_short_filt)),
    rep("Ctrl. vs. Sev OA (8h)", nrow(diffex_amb_sev_short_filt)),
    rep("Mod OA vs. Sev OA (8h)", nrow(diffex_OA_short_filt)),
    rep("Ctrl. vs. Mod OA (88d)", nrow(diffex_amb_mod_long_filt)),
    rep("Ctrl. vs. Sev OA (88d)", nrow(diffex_amb_sev_long_filt)),
    rep("Mod OA vs. Sev OA (88d)", nrow(diffex_OA_long_filt))
  )
) %>%
  group_by(gene_id) %>%
  summarise(comparison = paste(unique(comparison), collapse = "\n")) %>%
  ungroup()

saveRDS(deg_info, "../snow_crab_expression_explorer/deg_info.rds")
```

## Figure 5 - boxplots of standout genes 

```{r}
 #   "GWK47_015214", # Slc5a6_0 – Sodium-dependent multivitamin transporter; Chronic OA Tolerant (e)

standout.genes <- c(
  "GWK47_029205", # HSP90AA1 – Heat shock protein HSP 90-alpha; top -PC1 loadings (a)
  "GWK47_006688", # CUPA3_2 – Cuticle protein AM1199; top -PC1 loadings (a)
  "GWK47_019251", # MUCB1 – Integumentary mucin B.1; top +PC1 loadings (b)
  "GWK47_032544", # SLC15A1 – Solute carrier family 15 member 1; top +PC1 loadings (b)
  "GWK47_027395", # MRPL30 – 39S ribosomal protein L30, mitochondrial; Common DEG, 8 hr (c)
  "GWK47_010201", # MOIH – MOIH_3; Distinct DEG, 8 hr, also high at Day 88 (c)
  "GWK47_018958", # Cad87A – Cadherin-87A; Common & persistent DEG (d)
  "GWK47_010369", # NUP43 – Nucleoporin Nup43; Common & persistent DEG (d)
  "GWK47_049593", # Pnlip_1 – Pancreatic triacylglycerol lipase; Chronic OA Tolerant (e)
  "GWK47_039354", # OtopLc – Proton channel OtopLc; Chronic OA Tolerant (e)
  "GWK47_027385", # l(2)dtl – lethal (2) denticleless; Chronic OA Stressed (f)
  "GWK47_031192"  # CA7 – Carbonic anhydrase 7; Chronic OA Stressed (f)
)

assay(vsd.treatment) %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  filter(gene_id %in% standout.genes) %>%
  mutate(gene_id = factor(gene_id, levels = standout.genes)) %>%
  pivot_longer(starts_with("S")) %>%
  left_join(sample.info[c("sampleID", "treatment")],
            by = c("name" = "sampleID")) %>%
  filter(!is.na(treatment)) %>%
  # 👇 Add dummy spacer level "gap" between short-term and long-term treatments
  mutate(
    treatment = factor(
      treatment,
      ordered = TRUE,
      levels = c("ambient", "moderate_short", "severe_short",
                 "gap",                   # ← spacer level
                 "moderate_long", "severe_long")
    )
  ) %>%
  left_join(
    snow.annot[c("gene_id", "gene_name", "protein",
                 "gene_uni", "protein_names")] %>% distinct()
  ) %>%
  mutate(
    label = paste0(protein, "\n(", gene_name, ")"),
    label = str_wrap(label, width = 23),
    label = factor(label, levels = label[match(standout.genes, gene_id)])
  ) %>%
  ggplot(aes(x = treatment, y = value, fill = treatment)) +
  # 🔹 Make boxplots thinner and closer together
  geom_boxplot(
    width = 0.6,   # thinner boxes
    alpha = 0.5,
    outlier.shape = NA
  ) +
  geom_jitter(width = 0.1, size = 0.5) +
  # Optional vertical line remains as divider
  geom_vline(xintercept = 4, color = "gray40", linetype = "dashed") +
  facet_wrap(~label, scales = "free", ncol = 4) +
  scale_fill_manual(
    name = NULL,
    labels = c("Control, 0-hr",
               "Moderate OA, 8-hr", "Severe OA, 8-hr",
               "Moderate OA, 88-d", "Severe OA, 88-d"),
    values = color_scale
  ) +
  theme_minimal(base_size = 13) +
  ylab("Expression Value (normalized)") +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y= element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    # 🔹 Adjust spacing between facets (increase if you want more white space)
    panel.spacing.x = unit(0.6, "lines"),  # try 0.6–1 for visible separation
    panel.spacing.y = unit(1, "lines")
  ) +
  guides(fill = guide_legend(nrow = 2, byrow = TRUE)) +
  # keep the dummy spacer level (creates the Day1 vs Day88 gap)
  scale_x_discrete(drop = FALSE)


```







# BONEYARD 

<!-- ```{r} -->
<!-- ggvenn(list( -->
<!--   `Moderate - 8Hr` = row.names(diffex_amb_mod_short_filt), -->
<!--   `Moderate - 84 days`  = row.names(diffex_amb_mod_long_filt)),  -->
<!--        fill_color = c("yellow", "#ffc043"), -->
<!--        stroke_size = 0.5, set_name_size = 4) + ggtitle("Moderate OA") -->

<!-- ggvenn(list( -->
<!--   `Severe - 8Hr` = row.names(diffex_amb_sev_short_filt), -->
<!--   `Severe - 84 days` = row.names(diffex_amb_sev_long_filt)),  -->
<!--        fill_color = c("red", "darkred"), -->
<!--        stroke_size = 0.5, set_name_size = 4)+ ggtitle("Severe OA") -->

<!-- ggvenn(list( -->
<!--   `Moderate\n8Hr` = row.names(diffex_amb_mod_short_filt), -->
<!--   `Severe - 8Hr` = row.names(diffex_amb_sev_short_filt), -->
<!--   `Moderate - 84 days`  = row.names(diffex_amb_mod_long_filt), -->
<!--   `Severe\n84 days` = row.names(diffex_amb_sev_long_filt)),  -->
<!--        fill_color = c("yellow", "#ffc043","red", "darkred"), -->
<!--        stroke_size = 0.5, set_name_size = 4, show_percentage = F)  -->


<!-- data.frame(gene_id=intersect(row.names(diffex_amb_sev_short_filt),row.names(diffex_amb_sev_long_filt)))%>% -->
<!--     left_join(snow.annot) -->

<!-- data.frame(gene_id=intersect(row.names(diffex_amb_mod_short_filt),row.names(diffex_amb_mod_long_filt)))%>% -->
<!--     left_join(snow.annot) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- gene_sets <- list( -->
<!--   `Moderate 8 Hr AND 84 days` = intersect(row.names(diffex_amb_mod_short_filt), row.names(diffex_amb_mod_long_filt)), -->
<!--   `Severe 8 Hr AND 84 days` = intersect(row.names(diffex_amb_sev_short_filt), row.names(diffex_amb_sev_long_filt)), -->
<!--   `Moderate 8 Hr OR 84 days` = setdiff(row.names(diffex_amb_mod_short_filt), row.names(diffex_amb_mod_long_filt)), # This is not right -->
<!--   `Severe 8 Hr OR 84 days` = setdiff(row.names(diffex_amb_sev_short_filt), row.names(diffex_amb_sev_long_filt))) # This is not right -->
<!-- ``` -->


<!-- ### Heat maps of DEGs  -->

<!-- ```{r} -->
<!-- # Generate counts matrix With DEGs among any treatment  -->
<!-- sample_meta_figures <- sample.info %>% filter(sampleID%!in%c("S45","S60")) %>%  -->
<!--     mutate(treatment=factor(treatment, ordered = T, -->
<!--                           levels=c("ambient","moderate_short","severe_short","moderate_long","severe_long"))) %>%  -->
<!--                               arrange(treatment) -->

<!-- # DEGs among Ambient, Moderate  -->
<!-- counts_mod_degs <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene") %>%  -->
<!--     filter(gene %in% unique(row.names(diffex_amb_mod_short_filt), row.names(diffex_amb_mod_long_filt))) %>%  -->
<!--     column_to_rownames("gene") -->
<!-- pheatmap(counts_mod_degs %>% dplyr::select((sample_meta_figures%>%filter(OA%in%c("ambient","moderate")))$sampleID),  -->
<!--          cluster_rows=TRUE, cluster_cols=FALSE, -->
<!--          show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=color_scale), -->
<!--          scale="row", main = "Differentially expressed genes (DEGs), moderate OA treatment",  -->
<!--          annotation_col=(sample_meta_figures %>%filter(OA%in%c("ambient","moderate")) %>% column_to_rownames("sampleID"))[,"treatment", drop=FALSE],  -->
<!--          fontsize = 8, cutree_rows = 5, border_color=F) -->

<!-- # DEGs among Ambient, Severe  -->
<!-- counts_sev_degs <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene") %>%  -->
<!--     filter(gene %in% unique(row.names(diffex_amb_sev_short_filt), row.names(diffex_amb_sev_long_filt))) %>%  -->
<!--     column_to_rownames("gene") -->

<!-- pheatmap(counts_sev_degs %>% dplyr::select((sample_meta_figures%>%filter(OA%in%c("ambient","severe")))$sampleID),  -->
<!--          cluster_rows=TRUE, cluster_cols=FALSE, -->
<!--          show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=color_scale), -->
<!--          scale="row", main = "Differentially expressed genes (DEGs), severe OA treatment",  -->
<!--          annotation_col=(sample_meta_figures %>%filter(OA%in%c("ambient","severe")) %>% column_to_rownames("sampleID"))[,"treatment", drop=FALSE],  -->
<!--          fontsize = 8, cutree_rows = 6, border_color = NA) -->



<!-- ## <--- Group DEGs by those that are involved in short-term response only, long-term response only, or both -----> -->
<!-- ## Moderate OA  -->

<!-- # From DESeq2 intersect results - genes that are DEGs at 8hr but not 84 days  -->
<!-- counts_mod_8hr_OR_84d <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene") %>%  -->
<!--     filter(gene %in% gene_sets$`Moderate 8 Hr OR 84 days`) %>%  -->
<!--     column_to_rownames("gene") -->
<!-- pheatmap(counts_mod_8hr_OR_84d %>% dplyr::select((sample_meta_figures%>%filter(OA%in%c("ambient","moderate")))$sampleID),  -->
<!--          cluster_rows=TRUE, cluster_cols=FALSE, -->
<!--          show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=color_scale), -->
<!--          scale="row", main = "Moderate OA - Either Short OR Long Acclimation (not both)",  -->
<!--          annotation_col=(sample_meta_figures %>%filter(OA%in%c("ambient","moderate")) %>% column_to_rownames("sampleID"))[,"treatment", drop=FALSE],  -->
<!--          fontsize = 8, cutree_rows = 4,  -->
<!--          border_color = NA) -->

<!-- # From DESeq2 intersect results - genes that are DEGs at 8hr AND also at 84 days  -->
<!-- counts_mod_8hr_AND_84d <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene") %>%  -->
<!--     filter(gene %in% gene_sets$`Moderate 8 Hr AND 84 days`) %>%  -->
<!--     column_to_rownames("gene") -->
<!-- pheatmap(counts_mod_8hr_AND_84d %>% dplyr::select((sample_meta_figures%>%filter(OA%in%c("ambient","moderate")))$sampleID),  -->
<!--          cluster_rows=TRUE, cluster_cols=FALSE, -->
<!--          show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=color_scale), -->
<!--          scale="row", main = "Moderate OA - Short AND Long acclimation",  -->
<!--          annotation_col=(sample_meta_figures %>%filter(OA%in%c("ambient","moderate")) %>% column_to_rownames("sampleID"))[,"treatment", drop=FALSE],  -->
<!--          fontsize = 8, cutree_rows = 3,  -->
<!--          border_color = NA) -->

<!-- ## <--- Group DEGs by those that are involved in short-term response only, long-term response only, or both -----> -->
<!-- ## Severe OA  -->

<!-- # From DESeq2 intersect results - genes that are DEGs at 8hr but not 84 days  -->
<!-- counts_sev_8hr_OR_84d <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene") %>%  -->
<!--     filter(gene %in% gene_sets$`Severe 8 Hr OR 84 days`) %>%  -->
<!--     column_to_rownames("gene") -->
<!-- pheatmap(counts_sev_8hr_OR_84d %>% dplyr::select((sample_meta_figures%>%filter(OA%in%c("ambient","severe")))$sampleID),  -->
<!--          cluster_rows=TRUE, cluster_cols=FALSE, -->
<!--          show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=color_scale), -->
<!--          scale="row", main = "Severe OA - Either Short OR Long (not both)",  -->
<!--          annotation_col=(sample_meta_figures %>%filter(OA%in%c("ambient","severe")) %>% column_to_rownames("sampleID"))[,"treatment", drop=FALSE],  -->
<!--          fontsize = 8, cutree_rows = 4,  -->
<!--          border_color = NA) -->

<!-- # From DESeq2 intersect results - genes that are DEGs at 8hr AND also at 84 days  -->
<!-- counts_sev_8hr_AND_84d <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene") %>%  -->
<!--     filter(gene %in% gene_sets$`Severe 8 Hr AND 84 days`) %>%  -->
<!--     column_to_rownames("gene") -->
<!-- pheatmap(counts_sev_8hr_AND_84d %>% dplyr::select((sample_meta_figures%>%filter(OA%in%c("ambient","severe")))$sampleID),  -->
<!--          cluster_rows=TRUE, cluster_cols=FALSE, -->
<!--          show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=color_scale), -->
<!--          scale="row", main = "Severe OA - Short AND Long Acclimation",  -->
<!--          annotation_col=(sample_meta_figures %>%filter(OA%in%c("ambient","severe")) %>% column_to_rownames("sampleID"))[,"treatment", drop=FALSE],  -->
<!--          fontsize = 8, cutree_rows = 3,  -->
<!--          border_color = NA) -->
<!-- ``` -->

<!-- ## Genes that are persistent DEGs in both short and long term - 2 only -->

<!-- ```{r} -->
<!-- counts_acclimatory_all <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene") %>% -->
<!--     filter(gene %in% intersect(gene_sets$`Moderate 8 Hr AND 84 days`, gene_sets$`Severe 8 Hr AND 84 days`)) %>% -->
<!--     column_to_rownames("gene") -->
<!-- # pheatmap(counts_acclimatory_all %>% dplyr::select((sample_meta_figures)$sample), -->
<!-- #          cluster_rows=TRUE, cluster_cols=FALSE, -->
<!-- #          show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=color_scale), -->
<!-- #          scale="row", main = "Severe OA - Short AND Long Acclimation", -->
<!-- #          annotation_col=(sample_meta_figures %>% column_to_rownames("sampleID"))[,"treatment", drop=FALSE], -->
<!-- #          fontsize = 8, cutree_rows = 3, -->
<!-- #          border_color = NA) -->

<!-- persist.intersect.a <- print(counts_acclimatory_all %>% rownames_to_column("gene") %>% -->
<!--     pivot_longer(starts_with("S"),values_to = "vst",names_to = "sampleID") %>% -->
<!--     left_join(snow.annot, by=c("gene"="gene_id")) %>% -->
<!--     left_join(sample.info%>%dplyr::select(sampleID,treatment), by="sampleID") %>% -->
<!--     filter(treatment%in%c("ambient","severe_short","severe_long")) %>% -->
<!--         mutate(treatment=factor(treatment, ordered=T, levels=c("ambient","severe_short","severe_long"))) %>%  -->
<!--     group_by(gene, protein_names, treatment) %>% dplyr::summarize(mean=mean(vst), sd=sd(vst)) %>% -->
<!--     ggplot(aes(x=treatment,y=mean,group=gene,label=protein_names))+geom_line() + geom_point() + -->
<!--           geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=.2, position=position_dodge(0.05)) +  -->
<!--         theme_minimal() + -->
<!--     theme(legend.position = "none") + xlab(NULL) + facet_wrap(~protein_names, scales="free", nrow=1)) -->


<!-- persist.intsersect.b <- print(counts_acclimatory_all %>% rownames_to_column("gene") %>% -->
<!--     pivot_longer(starts_with("S"),values_to = "vst",names_to = "sampleID") %>% -->
<!--     left_join(snow.annot, by=c("gene"="gene_id")) %>% -->
<!--     left_join(sample.info%>%dplyr::select(sampleID,treatment), by="sampleID") %>% -->
<!--     filter(treatment%in%c("ambient","moderate_short","moderate_long")) %>% -->
<!--         mutate(treatment=factor(treatment, ordered=T, levels=c("ambient","moderate_short","moderate_long"))) %>%  -->
<!--     group_by(gene, protein_names, treatment) %>% dplyr::summarize(mean=mean(vst), sd=sd(vst)) %>% -->
<!--     ggplot(aes(x=treatment,y=mean,group=gene,label=protein_names))+geom_line() + geom_point() + -->
<!--           geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=.2, position=position_dodge(0.05)) +  -->
<!--         theme_minimal() + -->
<!--     theme(legend.position = "none") + xlab(NULL) + facet_wrap(~protein_names, scales="free", nrow=1)) -->

<!-- persist.intsersect.b / persist.intersect.a -->

<!-- snow.annot %>% filter(gene_id %in% rownames(counts_acclimatory_all)) -->
<!-- ``` -->
<!-- ### Any genes represent important acclimation mechanisms in moderate OA?  -->

<!-- ```{r} -->
<!-- library(stringr) -->

<!-- ggvenn(list( -->
<!--   `Chronic OA DEGs (among OA)` = row.names(subset(diffex_OA_long_filt, log2FoldChange>0)), -->
<!--   `Moderate - 8hr vs. Control`  = row.names(diffex_amb_mod_short_filt), -->
<!--   `Moderate - 88d vs. Control`  = row.names(subset(diffex_amb_mod_long_filt, log2FoldChange<0))), -->
<!--        fill_color = c("yellow", "#ffc043", "red3"), -->
<!--        stroke_size = 0.5, set_name_size = 4, show_percentage = F) #+ ggtitle("Moderate vs. Severe, 8Hr") -->

<!-- # One common among all  -->
<!-- a <- Reduce(intersect, list( -->
<!--          row.names(diffex_OA_long_filt), -->
<!--          row.names(diffex_amb_mod_short_filt), -->
<!--          row.names(diffex_amb_mod_long_filt))) -->

<!-- # Higher in moderate OA vs severe OA @ 88=days AND modreat OA vs. control at 8-hr -->
<!-- b <- Reduce(intersect, list( -->
<!--          row.names(diffex_OA_long_filt), -->
<!--          row.names(diffex_amb_mod_short_filt))) -->

<!-- # One common among all  -->
<!-- c <- Reduce(intersect, list( -->
<!--          row.names(subset(diffex_OA_long_filt, log2FoldChange>0)), -->
<!--          row.names(subset(diffex_amb_mod_long_filt, log2FoldChange<0)))) -->


<!-- # Hand-curated list of interesting genes from the above sets - higher in chronic moderate OA compared to all other treatments:  -->
<!-- d <- c("GWK47_044097", "GWK47_046958", "GWK47_026454", "GWK47_027793",  "GWK47_026211", "GWK47_025886","GWK47_022288","GWK47_017946","GWK47_008775","GWK47_008774","GWK47_008314", "GWK47_007022", "GWK47_050617","GWK47_048344","GWK47_041012", "GWK47_039981","GWK47_037994","GWK47_027940") -->
<!-- # Also interesting, but not as convincing: c("GWK47_043379","GWK47_055060", "GWK47_055061","GWK47_009486","GWK47_006933","GWK47_003044",) -->

<!-- e <- rownames(as.data.frame(subset(diffex_OA_long_filt, log2FoldChange < -0.5)))  -->

<!-- # Hand curated list of interesting genes that COULD be biomarkers  -->
<!-- f <- c("GWK47_029205", "GWK47_029842", "GWK47_031192", "GWK47_033734", "GWK47_042166", -->
<!--   "GWK47_037129", "GWK47_051181","GWK47_008600","GWK47_016794", "GWK47_027385") -->

<!-- g<- c("GWK47_027385","GWK47_016794", "GWK47_051181","GWK47_031192") -->
<!-- h <- c("GWK47_033734", "GWK47_042166","GWK47_029842","GWK47_037129","GWK47_029205") # responds to other things too? Or develomental differences?  -->

<!-- for (i in 1:length(g)) { -->
<!-- print( -->
<!-- assay(vsd.treatment) %>% -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% g[i]) %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "evalue")]) %>% -->
<!--     group_by(gene_id) %>% slice_min(evalue) %>%  -->
<!--   mutate(label = paste0(protein, "\n(", gene_id, ")"), -->
<!--          label = str_wrap(label, width = 35)) %>%  # <-- wrap long labels here -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.4) + -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 3) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 10) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank(), -->
<!--         legend.position = "bottom"))  -->
<!-- } -->
<!-- length(d) -->

<!-- snow.annot %>% View() filter(gene_id %in% g) %>% dplyr::select(gene_id:protein, spid,evalue,protein_names) %>% distinct() %>% group_by(gene_id) %>% slice_min(evalue) -->
<!-- ``` -->

<!-- Figure 5 - boxplots of interesting genes:  -->
<!-- - Top candidate chronic OA stress biomarkers  -->
<!-- - Top candidate chronic OA tolerance biomarkers  -->

<!-- ### Top OA stress genes  -->

<!-- ```{r} -->
<!-- # library(cowplot) -->
<!-- # library(patchwork) -->

<!-- print( -->
<!-- assay(vsd.treatment) %>% -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% g) %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein")]) %>% -->
<!--   mutate( -->
<!--       label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       # label = paste0(protein, "\n(", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 25)) %>%  # <-- wrap long labels here -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.4, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 2) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank(), -->
<!--         legend.position = "none") + -->
<!--   #       legend.position = c(0.99, 0),  # x,y position (0=left/bottom, 1=right/top) -->
<!--   # legend.justification = c("right", "bottom"))) + -->
<!--     ggtitle("(a)")) -->

<!-- print( -->
<!-- assay(vsd.treatment) %>% -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% h) %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "evalue")]) %>% -->
<!--   mutate( -->
<!--       label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       # label = paste0(protein, "\n(", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 25)) %>%  # <-- wrap long labels here -->
<!--     group_by(gene_id) %>% slice_min(evalue) %>%  -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.4, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 2) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank(), -->
<!--         legend.position = "none") + -->
<!--   #       legend.position = c(0.99, 0),  # x,y position (0=left/bottom, 1=right/top) -->
<!--   # legend.justification = c("right", "bottom"))) + -->
<!--     ggtitle("(b)")) -->


<!-- ### For paper!!!!  -->

<!-- # All together -->
<!-- assay(vsd.treatment) %>% -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% c(g,h)) %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "evalue")]) %>% -->
<!--   mutate( -->
<!--       label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       # label = paste0(protein, "\n(", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 20)) %>%  # <-- wrap long labels here -->
<!--     group_by(gene_id) %>% slice_min(evalue) %>%  -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.8, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 3) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Ctrl (0-hr)",  -->
<!--                              "pH 7.8, 8-hr", "pH 7.5, 8-hr", -->
<!--                              "pH 7.8, 88-d", "pH 7.5, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank(),  -->
<!--         legend.position = "bottom") #+ -->
<!-- # guides(fill = guide_legend(nrow = 2, byrow = TRUE)) + -->
<!-- # theme( -->
<!-- #   legend.position = "bottom", -->
<!-- #   legend.title = element_blank(), -->
<!-- #   legend.spacing.y = unit(0.1, "cm"),      # reduces vertical spacing -->
<!-- #   legend.key.height = unit(0.3, "cm"))      # shrinks key boxes -->


<!-- # SVA-adjusted -->
<!-- Y_adj %>% -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% c(g,h)) %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "evalue")]) %>% -->
<!--   mutate( -->
<!--       label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       # label = paste0(protein, "\n(", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 20)) %>%  # <-- wrap long labels here -->
<!--     group_by(gene_id) %>% slice_min(evalue) %>%  -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.8, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 3) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Ctrl (0-hr)",  -->
<!--                              "pH 7.8, 8-hr", "pH 7.5, 8-hr", -->
<!--                              "pH 7.8, 88-d", "pH 7.5, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank(),  -->
<!--         legend.position = "bottom") -->
<!-- ``` -->

<!-- ### Top 4 OA tolerance genes  -->

<!-- ```{r} -->
<!-- assay(vsd.treatment) %>% -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% c("GWK47_027940","GWK47_026454","GWK47_044097","GWK47_046958")) %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "evalue")]) %>% -->
<!--   mutate( -->
<!--       label=case_when(protein %in% c("hypothetical protein")~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       # label = paste0(protein, "\n(", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 35)) %>%  # <-- wrap long labels here -->
<!--     group_by(gene_id) %>% slice_min(evalue) %>% distinct() %>%  -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.4, outlier.shape = NA) + geom_jitter(width=0.2, size=1) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 2) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control (0-hr)",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank()) + -->
<!--   guides(fill = guide_legend( -->
<!--     nrow = 2, byrow = TRUE)) + -->
<!--   theme( -->
<!--     legend.position = "bottom", -->
<!--     legend.title = element_blank()) -->
<!-- ``` -->

<!-- ### Persistent DEGs in both OA treatments  -->

<!-- ```{r} -->
<!-- assay(vsd.treatment) %>% -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% c("GWK47_018958","GWK47_010369")) %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "evalue")]) %>% -->
<!--   mutate( -->
<!--       label=case_when(protein %in% c("hypothetical protein")~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       # label = paste0(protein, "\n(", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 35)) %>%  # <-- wrap long labels here -->
<!--     group_by(gene_id) %>% slice_min(evalue) %>% distinct() %>%  -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.6, outlier.shape = NA) + geom_jitter(width=0.2, size=1) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 2) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control (0-hr)",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank()) + -->
<!--   guides(fill = guide_legend( -->
<!--     nrow = 2, byrow = TRUE)) + -->
<!--   theme( -->
<!--     legend.position = "bottom", -->
<!--     legend.title = element_blank()) -->
<!-- ``` -->
<!-- #### Cuticle or chitin proteins  -->

<!-- ```{r} -->
<!-- assay(vsd.treatment) %>% -->
<!-- #Y_adj %>%  -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!-- #  filter(gene_id %in% (snow.annot%>%filter(grepl("CUPA3",gene_name)))$gene_id[1:3]) %>% -->
<!-- #  filter(gene_id %in% (snow.annot%>%filter(grepl("Chitin|chitin|Cuticle|cuticle",protein_names)))$gene_id) %>% -->
<!-- #  filter(gene_id %in% (snow.annot%>%filter(grepl("V-type proton ATPase|Sodium/potassium-transporting ATPase",protein_names)))$gene_id) %>% -->
<!-- #    filter(gene_id %in% c(all_degs, pc1_genes_pos, pc1_genes_neg, pc2_genes_pos, pc2_genes_neg)) %>% -->
<!-- #    filter(gene_id != "GWK47_025690") %>%  -->
<!--   filter(gene_id %in% "GWK47_010201") %>% -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "protein_names")]) %>% -->
<!--   mutate( -->
<!-- #      label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       label = paste0(protein, "\n(", gene_id, "\n", ")"), -->
<!--       label = str_wrap(label, width = 35)) %>%  # <-- wrap long labels here -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.6, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 6) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank()) +  -->
<!--   guides(fill = guide_legend( -->
<!--     nrow = 2, byrow = TRUE)) + -->
<!--   theme( -->
<!--     legend.position = "bottom", -->
<!--     legend.title = element_blank()) -->


<!-- ``` -->

<!-- #### Vitamin K, blood clotting, other interseting gene sets  -->

<!-- ```{r} -->
<!-- assay(vsd.treatment) %>% -->
<!-- #Y_adj %>%  -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!-- #  filter(gene_id %in% (snow.annot%>%filter(grepl("CUPA3",gene_name)))$gene_id[1:3]) %>% -->
<!-- #  filter(gene_id %in% (snow.annot%>%filter(grepl("GO:0042373|GO:0004252",gene_ontology_go)))$gene_id) %>% -->
<!-- #  filter(gene_id %in% c(all_degs, pc1_genes_pos, pc1_genes_neg, pc2_genes_pos, pc2_genes_neg)) %>% -->
<!-- #    filter(gene_id %in% c("GWK47_012107", "GWK47_003819","GWK47_016350", "GWK47_001205")) %>% #ion regulation  -->
<!-- #    filter(gene_id %in% c("GWK47_010377", "GWK47_054260")) %>% #redox homeostasis GWK47_003902/GWK47_024712 -->
<!-- #    filter(gene_id %in% c("GWK47_003902","GWK47_024712")) %>% #actin capping -->
<!-- #  filter(gene_id %in% (snow.annot%>%filter(grepl("Twitchin",protein)))$gene_id) %>%  -->
<!-- #  filter(gene_id %in% (snow.annot%>%filter(grepl("arbonic anhydrase",protein)))$gene_id) %>%  -->
<!--     filter(gene_id=="GWK47_024010") %>%  -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "gene_uni", "protein_names")]) %>% -->
<!--   mutate( -->
<!-- #      label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       label = paste0(protein, "\n(", gene_name, " - ", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 25)) %>%  # <-- wrap long labels here -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.6, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 5) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control, 0-hr",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank()) +  -->
<!--   theme(legend.position = "right") -->

<!--   # guides(fill = guide_legend( -->
<!--   #   nrow = 2, byrow = TRUE)) + -->
<!--   # theme( -->
<!--   #   legend.position = "bottom", -->
<!--   #   legend.title = element_blank()) -->
<!-- ``` -->
<!-- ### OA Tolerance Genes?  -->

<!-- ```{r} -->
<!-- # LOWER EXPRESSION IN SEVERE OA, higher in moderate OA -->
<!-- Chronic_oa_tolerance <- rownames(subset(diffex_OA_long_filt, log2FoldChange>0)) -->

<!-- chronic_oa_tolerance_unique <- c( -->
<!--   "GWK47_002621","GWK47_044097","GWK47_019251","GWK47_039428","GWK47_049593", -->
<!--   "GWK47_039354","GWK47_012393","GWK47_044956","GWK47_015214","GWK47_046958", -->
<!--   "GWK47_037894","GWK47_018438","GWK47_027940","GWK47_026454","GWK47_007022", -->
<!--   "GWK47_008314","GWK47_050617","GWK47_048344","GWK47_024947","GWK47_041012", -->
<!--   "GWK47_007937","GWK47_008775","GWK47_000598","GWK47_008774","GWK47_002108", -->
<!--   "GWK47_027793","GWK47_030901","GWK47_014333","GWK47_039981","GWK47_038916", -->
<!--   "GWK47_026211","GWK47_025886") -->

<!-- # PCA, not SV-controlled  -->
<!-- plotPCA(vsd.treatment[rownames(vsd.treatment) %in% chronic_oa_tolerance_unique],  -->
<!--         intgroup="treatment") + -->
<!--     geom_point(size=1.5) + -->
<!--       scale_color_manual(values=color_scale) + -->
<!--       theme_minimal()+ stat_ellipse()  -->

<!-- assay(vsd.treatment) %>% -->
<!-- #Y_adj %>%  -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% chronic_oa_tolerance_unique) %>% # Genes highlighted in Figure 3 -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "protein_names")]) %>% -->

<!--   mutate( -->
<!-- #      label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       label = paste0(protein, "\n(", gene_name, " - ", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 25)) %>%  # <-- wrap long labels here -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.6, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 6) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control, 80-hr",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank()) +  -->
<!--   theme(legend.position = "none") -->
<!-- ``` -->

<!-- ### OA Stress Genes?  -->

<!-- ```{r} -->
<!-- # HIGHER EXPRESSION IN SEVERE OA -->
<!-- Chronic_oa_stress <- rownames(subset(diffex_OA_long_filt, log2FoldChange<0)) -->

<!-- # PCA, not SV-controlled  -->
<!-- plotPCA(vsd.treatment[rownames(vsd.treatment) %in% c(Chronic_oa_stress)],  -->
<!--         intgroup="treatment") + -->
<!--     geom_point(size=1.5) + -->
<!--       scale_color_manual(values=color_scale) + -->
<!--       theme_minimal()+ stat_ellipse()  -->

<!-- assay(vsd.treatment) %>% -->
<!-- #Y_adj %>%  -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% Chronic_oa_stress[1:28]) %>% # Genes highlighted in Figure 3 -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "protein_names")]) %>% -->

<!--   mutate( -->
<!-- #      label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       label = paste0(protein, "\n(", gene_name, " - ", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 25)) %>%  # <-- wrap long labels here -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.6, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 6) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control, 80-hr",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank()) +  -->
<!--   theme(legend.position = "none") -->
<!-- ``` -->



<!-- ### Look for stress response genes based on Poweret al. 2023 study  -->

<!-- ```{r} -->
<!-- transport_stress_genes <- c("GWK47_043225","GWK47_032395","GWK47_035755","GWK47_022009","GWK47_044774","GWK47_002223", -->
<!--                             "GWK47_009484","GWK47_041179","GWK47_015185","GWK47_009678","GWK47_002381","GWK47_009172", -->
<!--                             "GWK47_033734","GWK47_033791","GWK47_054367","GWK47_027099") -->

<!-- # PCA, not SV-controlled  -->
<!-- plotPCA(vsd.treatment[rownames(vsd.treatment) %in% transport_stress_genes],  -->
<!--         intgroup="treatment") + -->
<!--     geom_point(size=1.5) + -->
<!--       scale_color_manual(values=color_scale) + -->
<!--       theme_minimal()+ stat_ellipse()  -->


<!-- #assay(vsd.treatment) %>% -->
<!-- Y_adj %>%  -->
<!--   as.data.frame() %>% -->
<!--   rownames_to_column("gene_id") %>% -->
<!--   filter(gene_id %in% c(all_degs, pc1_genes_pos, pc1_genes_neg, pc2_genes_pos, pc2_genes_neg)) %>% -->
<!--   filter(gene_id %in% transport_stress_genes) %>% # Genes highlighted in Figure 3 -->
<!--   pivot_longer(starts_with("S")) %>% -->
<!--   left_join(sample.info[c("sampleID", "treatment")], by = c("name" = "sampleID")) %>% -->
<!--   filter(!is.na(treatment)) %>% -->
<!--   mutate(treatment = factor( -->
<!--     treatment, -->
<!--     ordered = TRUE, -->
<!--     levels = c("ambient", "moderate_short", "severe_short", -->
<!--                "moderate_long", "severe_long") -->
<!--   )) %>% -->
<!--   left_join(snow.annot[c("gene_id", "gene_name", "protein", "protein_names")]) %>% -->
<!--   mutate( -->
<!-- #      label=case_when(protein=="hypothetical protein"~paste0(protein, "\n(", gene_id, ")"), TRUE~protein), -->
<!--       label = paste0(protein, "\n(", gene_name, " - ", gene_id, ")"), -->
<!--       label = str_wrap(label, width = 25)) %>%  # <-- wrap long labels here -->
<!--   ggplot(aes(x = treatment, y = value, fill = treatment)) + -->
<!--   geom_boxplot(alpha = 0.6, outlier.shape = NA) + geom_jitter(width=0.2, size=0.5) +  -->
<!--   geom_vline(xintercept = 3.5, color = "gray40", linetype = "dashed") +  # ← between 3rd & 4th -->
<!--   facet_wrap(~label, scales = "free", ncol = 3) +   # remove the labeller argument -->
<!--   scale_fill_manual(name = NULL,  -->
<!--                     labels=c("Control, 80-hr",  -->
<!--                              "Moderate OA, 8-hr", "Severe OA, 8-hr", -->
<!--                              "Moderate OA, 88-d", "Severe OA, 88-d"), -->
<!--                     values = color_scale) + -->
<!--   theme_minimal(base_size = 12) + -->
<!--     ylab("Expression Value (normalized)") +  -->
<!--   theme(axis.text.x = element_blank(), -->
<!--         axis.title.x = element_blank()) +  -->
<!--   theme(legend.position = "none") -->
<!-- ``` -->


<!-- # ------------------------------------  -->
<!-- ### Plot interesting genes -->

<!-- ```{r, warning=F, message=F} -->


<!-- gene.counts <- as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene_id") %>%  -->
<!--     left_join(snow.annot[c("gene_id","gene_name", "protein","protein_names", "spid", "gene_ontology_biological_process")]%>%distinct()) -->

<!-- c <- gene.counts %>%  filter(gene_id %in% row.names(diffex_amb_mod_short_filt[diffex_amb_mod_short_filt$log2FoldChange> 0.5,])) -->
<!--                                  c("GWK47_018133","GWK47_018134")) -->
<!--     #filter(gene_id %in% res_moderate_intersect$gene_id) -->
<!--     #filter(gene_id%in% c("GWK47_008839","GWK47_010377","GWK47_021798","GWK47_021798","GWK47_030664")) -->
<!-- #                                c("GWK47_035555","GWK47_042301")) # -->

<!-- #                                c("GWK47_036262","GWK47_046958","GWK47_051844","GWK47_027739")) #serine pepidase -->
<!-- #                                 filter(gene_id %in% rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange<0))) # higher in severe OA long term -->
<!-- #                                 filter(gene_id %in% rownames(as.data.frame(diffex_OA_long_filt)%>%filter(log2FoldChange>0))) # lower in severe OA long term -->
<!-- #                                filter(gene_id %in% c("GWK47_006688", "GWK47_035115", "GWK47_042440")) #cuticle, PC1 negative loadings -->

<!-- #                                filter(gene_id%in% -->
<!-- #                                c(rownames(as.data.frame(diffex_OA_long_filt)))) %>% #varied by OA at 84days   -->
<!-- #                                filter(grepl("Heat shock|hsp|HSP|Hsp", protein)) -->
<!-- #                                filter(protein!="hypothetical protein") %>%  -->
<!-- #                                filter(grepl("cortisol",gene_ontology_biological_process)) -->
<!-- #                                filter(gene_id=="GWK47_022272") -->
<!-- #                                filter(gene_id=="GWK47_009453") -->
<!-- #                                filter(grepl("Chitin|chitin", protein_names)) -->
<!--                                 #filter(grepl("Chitinase|chitinase", protein_names)) -->
<!-- #                                 filter(gene_id%in%c("GWK47_018133","GWK47_018134"))     -->
<!-- #                                 filter(gene_id %in% c("GWK47_036262","GWK47_046958","GWK47_051844","GWK47_027739", "GWK47_009825","GWK47_002658")) %>%  #serine endopeptidases -->
<!-- #                                 filter(grepl("Proclotting|Vitamin K",protein_names)) -->
<!-- #                                filter(gene_id%in%c("GWK47_010369", "GWK47_018958")) #two genes both "persistent degs" in mod and sev OA -->
<!--     # Plot top 20 differentially expressed genes in the selected treatment  -->

<!-- #c %>% dplyr::select(gene_name:gene_ontology_biological_process) #%>% write_clip(allow_non_interactive = T) -->

<!-- c -->

<!-- #for (i in 1:nrow(c)) {  -->
<!-- for (i in 1:25) {  -->
<!-- a <-  plotCounts(dds_sva, gene=c$gene_id[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>% -->
<!--   rownames_to_column("sampleID") %>% -->
<!--   left_join(sample.info[,c("sampleID", "treatment")]) %>% -->
<!-- #  filter(treatment %in% b) %>% -->
<!--   mutate(treatment=factor(treatment, ordered = TRUE,  -->
<!--                           levels=c("ambient", "moderate_short","severe_short", "moderate_long", "severe_long"))) -->

<!-- print( -->
<!--   ggplot(a%>%filter(), -->
<!--        aes(x=treatment, y=count, color=treatment, label=sampleID)) + -->
<!--   geom_boxplot(outlier.shape = NA) + -->
<!--    geom_jitter(width=0.25) +  -->
<!-- #  geom_text(size=2) +  -->
<!-- #  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) + -->
<!--   theme_bw() + -->
<!--     ggtitle(str_wrap(paste0(c$gene_id[i], " - ", c$protein_names[i]))) + -->
<!--     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") + -->
<!--    scale_color_manual(name="Treatment", values=color_scale)) -->
<!-- } -->
<!-- ``` -->






<!-- ```{r, message=F, warning=F, include=F} -->
<!-- ##### Run the function `DESeq` to assess differential expression  -->
<!-- # NOTE, could consider including 'minReplicatesForReplace=Inf' to keep DESeq2 from replacing outlier values with trimmed means.  -->
<!-- ## The default is for this to happen when there are more than 7 replicates.  -->

<!-- dds.DESeq.treatment <- DESeq(dds.treatment) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # these are the treatment "levels" that will be used to run pairwise contrasts  -->
<!-- levels(dds.DESeq.treatment$treatment) -->
<!-- ``` -->

<!-- ### Identify differentially expressed genes among contrasts  -->

<!-- #### Question: Which genes are differentially expressed in acute moderate OA exposure (ambient vs. pH 7.8 @ 8 hrs) -->

<!-- ```{r} -->
<!-- summary(res.mod.short <- results(dds.DESeq.treatment, contrast=c("treatment", "ambient", "moderate_short"), alpha=0.05)) -->
<!-- paste("No. of genes differentially expressed (padj<0.05), Ambient pH vs. moderate-short:",  sum(res.mod.short$padj < 0.05, na.rm=TRUE)) -->

<!-- ``` -->
<!-- #### Question: Which genes are differentially expressed in acute severe OA exposure (ambient vs. pH 7.5 @ 8 hrs) -->

<!-- ```{r} -->
<!-- summary(res.sev.short <- results(dds.DESeq.treatment, contrast=c("treatment", "ambient", "severe_short"), alpha=0.05)) -->
<!-- paste("No. of genes differentially expressed (padj<0.05), Ambient pH vs. severe-short:",  sum(res.sev.short$padj < 0.05, na.rm=TRUE)) -->
<!-- ``` -->
<!-- #### Question: Which genes differ in response to moderate vs. severe OA acute exposure (@ 8hrs) -->

<!-- ```{r} -->
<!-- summary(res.short <- results(dds.DESeq.treatment, contrast=c("treatment", "moderate_short", "severe_short"), alpha=0.05)) -->
<!-- paste("No. of genes differentially expressed (padj<0.05), Ambient pH vs. severe-short:",  sum(res.short$padj < 0.05, na.rm=TRUE)) -->
<!-- ``` -->

<!-- ### How does energy budget differ in crab held in long-term moderate vs. severe OA?  -->

<!-- ```{r} -->
<!-- print("Comparison: Ambient pH vs. moderate-long") -->
<!-- summary(res.long <- results(dds.DESeq.treatment, contrast=c("treatment", "severe_long", "moderate_long"), alpha=0.05)) -->
<!-- paste("No. of genes differentially expressed (padj<0.05) severe-long pH vs. moderate-long:",  sum(res.long$padj < 0.05, na.rm=TRUE)) -->
<!-- ``` -->


<!-- #### Total number of genes analyzed = 95,779 -->

<!-- ```{r} -->
<!-- nrow(res.long) -->
<!-- ``` -->
<!-- #### Save differential expression results to objects  -->

<!-- ```{r} -->
<!-- save(res.mod.short, file = "../results/deseq2/res.mod.short") -->
<!-- save(res.sev.short, file = "../results/deseq2/res.sev.short") -->
<!-- save(res.short, file = "../results/deseq2/res.short") -->
<!-- save(res.long, file = "../results/deseq2/res.long") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- #### Subset the DESeq results for only those statistically sign. ones -->
<!-- diffex.mod.short <- subset(res.mod.short, padj < 0.05& abs(log2FoldChange)>0.5) #  -->
<!-- diffex.sev.short <- subset(res.sev.short, padj < 0.05& abs(log2FoldChange)>0.5) -->
<!-- diffex.short <- subset(res.short, padj < 0.05& abs(log2FoldChange)>0.5) -->
<!-- diffex.long <- subset(res.long, padj < 0.05& abs(log2FoldChange)>0.5) -->
<!-- ``` -->



### Examine top differentially expressed genes in each treatment

Plot counts
NOTE: using the DeSeq2 function `plotCounts` plots original counts, i.e. those that are normalized, but outlier values are NOT replaced with trimmed means.  I'm really interested to see how the counts look that DESeq2 is analyzing, i.e. I want to see the outlier-replaced counts. So, I must include the argument `replaced=TRUE` in the plotCounts function.

It's good to look at the gene expression value - if counts are low (e.g. less than 50 or so) that gene isn't as interesting as one with mucn larger counts! 

#### <span style="color:darkgreen">Top 15 annotated DEGs, MODERATE OA (pH 7.8) SHORT Exposure (green)</span>

```{r, warning=FALSE, message=F, echo=F}
contrast <- diffex.mod.short.filt
#contrast <- diffex.sev.short.filt
#contrast <- diffex.short.filt
#contrast <- diffex.long.filt

# select treatments to display, or show all 
#b <- c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long") #all treatments 
b <- c("moderate_short", "ambient") #all treatments 

gene.counts <- 
#    counts.annot.tanner %>%
  as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene_id") %>%  
  left_join(contrast %>%
    as.data.frame() %>% rownames_to_column("gene_id")) %>%
  filter(!is.na(padj)) %>% 
    mutate(abs_lfc=abs(log2FoldChange)) 
    #%>% #, !is.na(spid)
#  dplyr::select(gene_id, padj, protein_names, everything()) %>% arrange(padj)

c <- gene.counts %>% arrange(desc(abs_lfc)) %>% head(n=15) #%>% dplyr::select(gene_id, padj, spid, evalue, protein_names)

# Plot top 20 differentially expressed genes in the selected treatment 

for (i in 1:nrow(c)) { 
a <-  plotCounts(dds.DESeq.treatment, gene=c$gene_id[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sampleID") %>%
  left_join(sample.info[,c("sampleID", "treatment")]) %>%
  filter(treatment %in% b) %>%
  mutate(treatment=factor(treatment, ordered = TRUE, 
                          levels=c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long")))

print(
  ggplot(a,
       aes(x=treatment, y=count, color=treatment, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_text() + 
#  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  theme_bw() +
    ggtitle(c$gene_id[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
   scale_color_manual(name="Treatment", values=color_scale))
}
```

#### <span style="color:firebrick4">Top 15 annotated DEGs, SEVERE OA (pH 7.5) vs. Ambient (firebrick4)</span>

```{r, warning=FALSE, message=F, echo=F}
#contrast <- diffex.mod.short.filt
contrast <- diffex.sev.short.filt
#contrast <- diffex.short.filt
#contrast <- diffex.long.filt

# select treatments to display, or show all 
#b <- c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long") #all treatments 
b <- c("severe_short", "ambient") #all treatments 

gene.counts <- 
#    counts.annot.tanner %>%
  as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene_id") %>%  
  left_join(contrast %>%
    as.data.frame() %>% rownames_to_column("gene_id")) %>%
  filter(!is.na(padj)) %>% 
    mutate(abs_lfc=abs(log2FoldChange)) 
    #%>% #, !is.na(spid)
#  dplyr::select(gene_id, padj, protein_names, everything()) %>% arrange(padj)

c <- gene.counts %>% arrange(desc(abs_lfc)) %>% head(n=15) #%>% dplyr::select(gene_id, padj, spid, evalue, protein_names)

# Plot top 20 differentially expressed genes in the selected treatment 

for (i in 1:nrow(c)) { 
a <-  plotCounts(dds.DESeq.treatment, gene=c$gene_id[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sampleID") %>%
  left_join(sample.info[,c("sampleID", "treatment")]) %>%
  filter(treatment %in% b) %>%
  mutate(treatment=factor(treatment, ordered = TRUE, 
                          levels=c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long")))

print(
  ggplot(a,
       aes(x=treatment, y=count, color=treatment, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_text() + 
#  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  theme_bw() +
    ggtitle(c$gene_id[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
   scale_color_manual(name="Treatment", values=color_scale))
}
```

#### <span style="color:sienna2">Top 15 annotated DEGs, MODERATE OA (7.8) vs. SEVERE OA (pH 7.5) SHORT Exposure (orange)</span>

```{r, warning=FALSE, message=F, echo=F}
#contrast <- diffex.mod.short.filt
#contrast <- diffex.sev.short.filt
contrast <- diffex.short.filt
#contrast <- diffex.long.filt

# select treatments to display, or show all 
#b <- c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long") #all treatments 
b <- c("moderate_short", "severe_short") #all treatments 

gene.counts <- 
#    counts.annot.tanner %>%
  as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene_id") %>%  
  left_join(contrast %>%
    as.data.frame() %>% rownames_to_column("gene_id")) %>%
  filter(!is.na(padj)) %>% 
    mutate(abs_lfc=abs(log2FoldChange)) 
    #%>% #, !is.na(spid)
#  dplyr::select(gene_id, padj, protein_names, everything()) %>% arrange(padj)

c <- gene.counts %>% arrange(desc(abs_lfc)) %>% head(n=15) #%>% dplyr::select(gene_id, padj, spid, evalue, protein_names)

# Plot top 20 differentially expressed genes in the selected treatment 

for (i in 1:nrow(c)) { 
a <-  plotCounts(dds.DESeq.treatment, gene=c$gene_id[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sampleID") %>%
  left_join(sample.info[,c("sampleID", "treatment")]) %>%
  filter(treatment %in% b) %>%
  mutate(treatment=factor(treatment, ordered = TRUE, 
                          levels=c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long")))

print(
  ggplot(a,
       aes(x=treatment, y=count, color=treatment, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_text() + 
#  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  theme_bw() +
    ggtitle(c$gene_id[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
   scale_color_manual(name="Treatment", values=color_scale))
}
```


#### <span style="color:firebrick4">Top 15 annotated DEGs, SEVERE OA (pH 7.5) LONG Exposure (red)</span>

```{r, warning=FALSE, message=F, echo=F}
#contrast <- diffex.mod.short.filt
#contrast <- diffex.sev.short.filt
#contrast <- diffex.short.filt
contrast <- diffex.long.filt

# select treatments to display, or show all 
#b <- c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long") #all treatments 
b <- c("moderate_long", "severe_long") #all treatments 

gene.counts <- 
#    counts.annot.tanner %>%
  as.data.frame(assay(vsd.treatment)) %>% rownames_to_column("gene_id") %>%  
  left_join(contrast %>%
    as.data.frame() %>% rownames_to_column("gene_id")) %>%
  filter(!is.na(padj)) %>% 
    mutate(abs_lfc=abs(log2FoldChange)) 
    #%>% #, !is.na(spid)
#  dplyr::select(gene_id, padj, protein_names, everything()) %>% arrange(padj)

c <- gene.counts %>% arrange(desc(abs_lfc)) %>% head(n=15) #%>% dplyr::select(gene_id, padj, spid, evalue, protein_names)

# Plot top 20 differentially expressed genes in the selected treatment 

for (i in 1:nrow(c)) { 
a <-  plotCounts(dds.DESeq.treatment, gene=c$gene_id[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sampleID") %>%
  left_join(sample.info[,c("sampleID", "treatment")]) %>%
  filter(treatment %in% b) %>%
  mutate(treatment=factor(treatment, ordered = TRUE, 
                          levels=c("ambient", "moderate_short", "moderate_long", "severe_short", "severe_long")))

print(
  ggplot(a,
       aes(x=treatment, y=count, color=treatment, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_text() + 
#  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  theme_bw() +
    ggtitle(c$gene_id[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
   scale_color_manual(name="Treatment", values=color_scale))
}
```

### Summary stats for the number of Differentially Expressed Genes (DEGs)

#### Bar plot of the number of DEGs identified in each contrast 

```{r}
# write simple function to return the name of an object as a string

# Here is a list containing all DEG sets - this will be used in subsequent code chunks, too! 
degs <- list(
diffex.mod.short.filt,
diffex.sev.short.filt,
diffex.short.filt, 
diffex.long.filt)

names(degs) <- c("Moderate OA - Short", "Severe OA - Short", 
                 "Moderate vs Severe - Short", "Moderate vs Severe - Long")

n.degs <- data.frame(matrix(NA, nrow = length(degs), ncol = 3))
names(n.degs) <- c("contrast", "n.degs", "perc")
for (i in (1:length(degs))) {
  n.degs[i,1] <- names(degs[i])
  n.degs[i,2] <- nrow(degs[[i]])
  n.degs[i,3] <- 100*round(nrow(degs[[i]])/ncol(counts.ts), digits = 3)
}

n.degs$contrast <- factor(n.degs$contrast, levels=n.degs$contrast, ordered = T)

save(degs, file = "../results/deseq2/degs")
```

Barplot of the number of DEGs 

```{r}
# single stressors vs. multiple stressors 
#pdf(file="../results/deseq2/DEGs_barplot_single-vs-double-stressor.pdf", height = 2, width = 7)
ggplot(n.degs, 
       # %>% 
       #   filter(contrast %in% c("Amb vs. Low pH @6C",
       #                          "3 vs. 6C @Amb pH", "6 amb vs. 3C low",
       #                          "6 vs. 10C @Amb pH", "6 amb vs. 10 low")) %>%
       #   mutate(effect_of=case_when(
       #     grepl("Amb vs. Low pH @6C", contrast) ~ "High pCO2",
       #     grepl("3 vs. 6C", contrast) ~ "Cold temperature",
       #     grepl("6 amb vs. 3C low", contrast) ~ "Cold temperature + High pCO2",
       #     grepl("6 amb vs. 10 low", contrast) ~ "Warm temperature + High pCO2",
       #     grepl("6 vs. 10C", contrast) ~ "Warm temperature")) %>% 
       #  mutate(effect_of=fct_relevel(effect_of,
       #     c("Cold temperature + High pCO2", "Cold temperature",
       #       "Warm temperature + High pCO2", "Warm temperature",
       #       "High pCO2"))), 
       aes(x=contrast, y=n.degs, fill=contrast)) + 
  geom_bar(alpha=0.8, stat = "identity", color="gray20", size=0.1) +
  geom_text(hjust=-0.1, size=3.5, aes(label=paste(comma(n.degs), " (", perc, "%)", sep=""))) + 
  xlab("Contrast") + ylab("Number Differentially Expressed Genes (% of all genes)") + 
  xlab(NULL) + ylab(NULL) + coord_flip() + ggtitle("Number Differentially Expressed Genes (% of all genes)") +
  theme_minimal() + ylim(c(0,2150)) + 
  theme(legend.position = "none", plot.title = element_text(size = 10)) +
  scale_fill_manual(name="Treatment",
                     values=c("Moderate OA - Short"="darkgreen",
                              "Severe OA - Short"="yellow3",
                              "Moderate vs Severe - Short"="sienna2",
                              "Moderate vs Severe - Long"="firebrick4"),
                     labels=c("Moderate OA - Short"="pH 7.8 vs. Control, Immediate Response",
                              "Severe OA - Short"="pH 7.5 vs. Control, Immediate Response",
                              "Moderate vs Severe - Short"="pH 7.5 vs. pH 7.8, Immediate response",
                              "Moderate vs Severe - Long"="pH 7.5 vs. pH 7.8, Long-term response")) +
  guides(fill=guide_legend(nrow=3,byrow=TRUE,reverse=TRUE)) 
#dev.off()
```

Regenerate barplot for paper (Figure 3) with the number of upregulated / downregualted DEGs by treatment 

```{r}
# summary(res.mod.short)
# summary(res.mod.long)
# summary(res.sev.short)
# summary(res.sev.long)

degs <- list(
diffex.mod.short.filt,
diffex.sev.short.filt,
diffex.short.filt, 
diffex.long.filt)

degs.updown <- list(
diffex.mod.short.filt %>% data.frame() %>% filter(log2FoldChange > 0),
diffex.mod.short.filt %>% data.frame() %>% filter(log2FoldChange < 0),
diffex.mod.long.filt %>% data.frame() %>% filter(log2FoldChange > 0),
diffex.mod.long.filt %>% data.frame() %>% filter(log2FoldChange < 0),
diffex.sev.short.filt %>% data.frame() %>% filter(log2FoldChange > 0),
diffex.sev.short.filt %>% data.frame() %>% filter(log2FoldChange < 0),
diffex.sev.long.filt %>% data.frame() %>% filter(log2FoldChange > 0),
diffex.sev.long.filt %>% data.frame() %>% filter(log2FoldChange < 0))
```

```{r}
names(degs.updown) <- c("Moderate OA - Short", "Moderate OA - Short", 
                        "Moderate OA - Long", "Moderate OA - Long",
                        "Severe OA - Short", "Severe OA - Short",
                        "Severe OA - Long", "Severe OA - Long")

n.degs.updown <- data.frame(matrix(NA, nrow = length(degs.updown), ncol = 3))
names(n.degs.updown) <- c("stressor", "n.degs", "perc")
for (i in (1:length(degs.updown))) {
  n.degs.updown[i,1] <- names(degs.updown[i])
  n.degs.updown[i,2] <- nrow(degs.updown[[i]])
  n.degs.updown[i,3] <- 100*round(nrow(degs.updown[[i]])/ncol(counts.ts), digits = 2)
}

n.degs.updown <- n.degs.updown %>%
  mutate(stressor=factor(stressor, levels=
                           c("Moderate OA - Short", "Moderate OA - Long", 
                 "Severe OA - Short", "Severe OA - Long")),
         response=rep(c("up", "down"), times=4))

# Barplot with no. of DEGs by single stressors vs. multiple stressors, including up/down regulation 
#pdf(file="../results/deseq2/DEGs_barplot_single-vs-double-stressor-up-down.pdf", height = 2, width = 7)
ggplot(n.degs.updown, 
       aes(x=stressor, y=n.degs, fill=stressor)) + 
  geom_bar_pattern(aes(pattern=response), alpha=0.75, stat = "identity", color="gray20", size=0.05,
                   pattern_fill="black", pattern_alpha=0.2, pattern_density=0.01, pattern_spacing=0.025) +
  xlab("Contrast") + ylab("Number Differentially Expressed Genes (% of all genes)") + 
  xlab(NULL) + ylab(NULL) + coord_flip() + ggtitle("Number Differentially Expressed Genes (% of all genes)") +
  theme_pubr() + scale_y_continuous(label=comma, limits = c(0,2000)) +
  theme(legend.position = "none", plot.title = element_text(size = 10)) +
  scale_fill_manual(name="Treatment",
                     values=c("Moderate OA - Short"="darkgreen",
                              "Moderate OA - Long"="yellow3",
                              "Severe OA - Short"="sienna2",
                              "Severe OA - Long"="firebrick4"),
                     labels=c("Moderate OA - Short"="pH 7.8, Short Exposure",
                              "Moderate OA - Long"="pH 7.8, Long Exposure",
                              "Severe OA - Short"="pH 7.5, Short Exposure",
                              "Severe OA - Long"="pH 7.5, Long Exposure")) +
  scale_pattern_manual(values = c("stripe","none")) +
  guides(fill=guide_legend(nrow=3,byrow=TRUE,reverse=TRUE)) 
#dev.off()

```

### Heat maps of DEGs 

#### Single heatmap with all DEGs among any treatment 

```{r}
# Generate counts matrix With DEGs among any treatment 

diffex.counts <- subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% unique(c(rownames(diffex.mod.short.filt),rownames(diffex.mod.long.filt),
         rownames(diffex.sev.short.filt), rownames(diffex.sev.long.filt))))

# Create a dataframe to annotate heatmap with treatments 
dds.df <- sample.info[match(colnames(diffex.counts), 
                               sample.info$sampleID),c("sampleID", "treatment")] %>%
  remove_rownames() %>% column_to_rownames(var = "sampleID")

# Make sure treatment order is correct
all(colnames(diffex.counts) == rownames(dds.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(diffex.counts, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=
             c("ambient"= "navyblue", "moderate_short"="darkgreen",
                       "moderate_long"="yellow3","severe_short"="sienna2",
                      "severe_long"="firebrick4")),
         scale="row", main = "Differentially expressed genes (DEGs) among any pH treatment", 
         annotation_col=dds.df[,"treatment", drop=FALSE], fontsize = 8,
         #cutree_rows = 2,  
         border_color=F)
```

### Venn Diagrams of DEGs by pairwise contrast

```{r, include=T, echo=F, message=F}
`Moderate OA - Short` <- row.names(degs$`Moderate OA - Short`)
`Moderate OA - Long`  <- row.names(degs$`Moderate OA - Long`)
`Severe OA - Short` <- row.names(degs$`Severe OA - Short`)
`Severe OA - Long` <- row.names(degs$`Severe OA - Long`)

venn.stressors <- data.frame(dat = 
             unique(c(`Moderate OA - Short`, `Moderate OA - Long`,
                      `Severe OA - Short`, `Severe OA - Long`))) %>%
  mutate(`Moderate OA - Short` = dat %in% `Moderate OA - Short`,
         `Moderate OA - Long` = dat %in% `Moderate OA - Long`,
         `Severe OA - Short` = dat %in% `Severe OA - Short`,
         `Severe OA - Long` = dat %in% `Severe OA - Long`) %>%
  dplyr::select(-dat) %>%
  euler()

#pdf(file="../results/deseq2/venn.pdf", height = 5.2, width = 6)
eulerr:::plot.euler(venn.stressors, counts = T, quantities = T, 
                      legend = T, edges=F, fontzie=8,
                      font=1, cex=1, alpha=0.65,
                      fills=list(fill=c("#428C42","#DBDB33", "#FF9772", "#BB4D4D")))
#dev.off()

# lighten("darkgreen", amount = .25)
# lighten("yellow3", amount = .25)
# lighten("sienna2", amount = .25)
# lighten("firebrick4", amount = .25)
```

###  DAVID enrichment analysis

It is a little clunky, but the DAVID functional analysis tool is one of the best ways to perform enrichment analysis (that I have found to date). To use, you copy a list of Uniprot IDs (or other gene identifier) to clipboard, then paste it into their tool. You then do the same for all genes in the analysis, providing an accurate background list of genes.  Here is code to copy all genes, then upregulated and downregulated DEGs identified by each contrast. For background on what an enrichment analysis is, check out [this great video](https://www.youtube.com/watch?v=H1cUs6pql9s). 

#### GENERATE LIST OF DEGS WITH ABS(L2FC)>0.5 (all DEGs, not up.down)

```{r, include=F, echo=F, message=F}
names(degs)

# BACKGROUND - all genes submitted to DESeq2 analysis 
enrich.background <- res.mod.long %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  left_join(tanner.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid, protein_names, protein_names), by = "gene_id") %>%
  dplyr::select(spid) %>% na.omit() %>% unlist() %>% as.vector()

# ------- "Effect of Moderate OA, Short-term exposure"
print(paste("DEGs, ", names(degs[1])))
enrich.mod.short <- degs[1] %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(tanner.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid, protein_names), by = "gene_id")

# ------- "Effect of Moderate OA - Long exposure"
print(paste("DEGs, ", names(degs[2])))
enrich.mod.long <- degs[2] %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(tanner.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid, protein_names), by = "gene_id")

# ------- "Effect of Severe OA - Short exposure"
print(paste("DEGs, ", names(degs[3])))
enrich.sev.short <- degs[3] %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(tanner.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid, protein_names), by = "gene_id")

# ------- "Effect of Severe OA - Long exposure"
print(paste("DEGs, ", names(degs[4])))
enrich.sev.long <- degs[4] %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(tanner.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid, protein_names), by = "gene_id")
```

```{r}
DEGs.4David <- list(
  
  "moderate_short" = 
    enrich.mod.short %>%
    dplyr::select(spid) %>% na.omit() %>% unlist() %>% as.vector(),
  
  "moderate_long" = 
    enrich.mod.long %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),

    "severe_short" = 
    enrich.sev.short %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "severe_long" = 
    enrich.sev.long %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector()

  ) 

# Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs 
write_delim(x = plyr::ldply(DEGs.4David, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", 
            file = "../results/deseq2/DEGs-4David.txt")
```

#### Run DAVID analysis in browser!  
- Navigate to DAVID on a browser - https://david.ncifcrf.gov/tools.jsp  
- On the left you'll see "Upload Gene List"  
- For Step 1, use the "B:Choose from a file" and upload the DEGs-updown-4David.txt file that was created in the last code chunk.  
- For the Step 2 dropdown list, select "UNIPROT_ACCESSION"  
- For Step 3, select "Gene List"  
- Select "Submit List". 
- You then need to upload a background list of genes. This is important. If you don't, the analysis will assume that you're comparing the DEG lists to the human genome (You'll see "Current Background: Homo sapiens"). Instead, we need to upload the list of all genes that we analyzed. To do this, use the below code chunk to copy Uniprot IDs for all genes assessed and paste into the Upload space on DAVID.   

```{r}
# BACKGROUND - all genes submitted to DESeq2 analysis 
enrich.background %>% write_clip(allow_non_interactive = TRUE)
```

#### Read in DAVID Enrichment Analysis results 

NOTE: outside of R I created a .tab file with the first column containing file names for enrichment results, and second column containing the contrast + which treatment was upregulated (for grouping in R). 
Example: the following row refers to the file contains enrichment results for genes upregulated in 10C in the contrast 6C vs. 10C at Low pH

```{r, warning=F, message=F}
#### Read in enriched biological functions for all contrasts

filenames <- read_delim(file="../results/deseq2/GO_filenames.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/deseq2/", filenames[i,"filename"], sep=""), delim = "\t"))}


david <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~") %>%
    separate(gene_set, sep = "_", into = c("OA", "duration"), remove = F) %>%
    mutate(treatment=factor(gene_set, ordered = TRUE, levels=c("moderate_short","moderate_long", "severe_short", "severe_long")),
           OA=as.factor(OA), duration=as.factor(duration)) %>% 
  tidyr::complete(OA, duration, category) # add rows for gene_sets that are missing from dataframe

save(david, file="../results/deseq2/david")
```

#### Copy DAVID results (Uniprot Kewords and Gene Ontology terms) to clipboard to paste into Google Spreadsheet

```{r}
# Uniprot keywords
david %>%
  filter(p_value<0.05) %>%
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  arrange(OA, duration, p_value) %>%
  mutate_at(c("p_value", "fdr"), funs(signif(.,2))) %>%
  dplyr::select(treatment, OA, duration, term, process, count, p_value, fdr, genes) %>%
  write_clip(allow_non_interactive = TRUE)

# Gene Ontology terms
david %>%
  filter(p_value<0.05) %>%
  filter(category=="GOTERM_BP_DIRECT") %>%
  arrange(OA, duration, p_value) %>%
  mutate_at(c("p_value", "fdr"), funs(signif(.,2))) %>%
  dplyr::select(treatment, OA, duration, term, process, count, p_value, fdr, genes) %>%
  write_clip(allow_non_interactive = TRUE)
```

### DAVID enrichment analysis results figure, all DEGs for each treatment (not separated by up/down-regulated processes)

```{r}
#pdf(file="../results/deseq2/KW-bubble.pdf", height = 3.05, width = 4)

david %>% #View() 
  #filter(count>=3) %>% # use this to filter for processes that have at minimum n genes
  filter(p_value<0.05) %>% #filter enrichment results for p-value

  # Option 1: Use this line for gene ontology biological processes  
  filter(category=="GOTERM_BP_DIRECT") %>%

  # Option 2: Use this line for Uniprot Keywords
  #filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
    
    # Can filter for other options in the "category" column, but might need to adjust the y= in the ggplot(aes). 
    # For example would need to use y=term for KEGG_PATHWAY

  group_by(treatment, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(treatment, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("treatment", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%
#  filter(response=="Upregulated") %>% 

  ungroup() %>% arrange(treatment, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=treatment, col=treatment)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Treatment",
                     values=c("moderate_short"=lighten("darkgreen", 0.25),
                              "moderate_long"=lighten("yellow3", 0.25),
                              "severe_short"=lighten("sienna2"),
                              "severe_long"=lighten("firebrick4")),
                     labels=c("moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure"),
                     guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes")
#dev.off()
```

#### Re-run DAVID enrichment analysis on genes that are expressed higher ("upregulated") and lower ("downregulated") in each treatment 

First, determine which log-fold change (LFC) direction (>0 or <0) refers to higher/lower expression for treatments in each contrast

```{r}
test <- ((
#  enrich.mod.short %>% 
#  enrich.mod.long %>% 
#  enrich.sev.short %>% 
  enrich.sev.long %>% 
    
#    filter(select(., contains("log2FoldChange"))>.5))$gene_id)
    filter(select(., contains("log2FoldChange"))<(-0.5)))$gene_id)

# Here are treatments 
# "ambient"        "moderate_long"  "moderate_short" "severe_long"    "severe_short"  

a <- c("ambient", "moderate_short")
c <- data.frame(gene=character(), treatment=character(), mean=numeric())
d <- list()

for (i in 1:length(test)) {
b <-  plotCounts(dds.DESeq.treatment, gene=test[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sampleID") %>%
  left_join(sample.info[,c("sampleID", "treatment")]) %>%
  filter(treatment %in% a)

c[1:2,1] <- test[i]
c[1:2,2:3] <- summarySE(measurevar = "count", groupvars = "treatment", data = b)[,c("treatment", "count")] %>% 
  mutate(treatment=as.character(treatment))

d[[i]] <- c
}

bind_rows(d, .id = "i") %>% 
  mutate(treat.num=factor(treatment, levels = a, ordered = T),
         gene=as.factor(gene)) %>%
  ggplot(aes(x=treatment, y=log(mean, base = 2), group=gene)) + geom_line(size=.25, color="gray50") +
  theme_minimal() + 
  #scale_x_continuous(breaks=c(1,2), labels=a) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2))

# RESULTS
# enrich.mod.short - LFC > 0.5 - LOWER IN OA
# enrich.mod.short - LFC < (-0.5) - HIGHER IN OA
# enrich.mod.long - LFC > 0.5 - LOWER IN OA
# enrich.mod.long - LFC < (-0.5) - HIGHER IN OA
# enrich.sev.short - LFC > 0.5 - LOWER IN OA
# enrich.sev.short - LFC < (-0.5) - HIGHER IN OA
# enrich.sev.long - LFC > 0.5 - LOWER IN OA
# enrich.sev.long - LFC < (-0.5) - HIGHER IN OA
```

### Generate .tab input file for DAVID enrichment analysis 
This file lists DEGs that are upregulated/downregulated in response to each treatment for DAVID enrichment analysis

```{r}
# RESULTS

DEGs.updown.4David <- list(
  
  "moderate_short_downregulated" = 
    # enrich.mod.short - LFC > 0.5 - LOWER IN OA
    enrich.mod.short %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>% na.omit() %>% unlist() %>% as.vector(),
  
  "moderate_short_upregulated" =  
    # enrich.mod.short - LFC < (-0.5) - HIGHER IN OA
    enrich.mod.short %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(), 

  "moderate_long_downregulated" = 
    # enrich.mod.long - LFC > 0.5 - LOWER IN OA
    enrich.mod.long %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "moderate_long_upregulated"  =
    # enrich.mod.long - LFC < (-0.5) - HIGHER IN OA
    enrich.mod.long %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "severe_short_downregulated" = 
    # enrich.sev.short - LFC > 0.5 - LOWER IN OA
    enrich.sev.short %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "severe_short_upregulated" = 
    # enrich.sev.short - LFC < (-0.5) - HIGHER IN OA
    enrich.sev.short %>%
     filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "severe_long_downregulated" = 
    # enrich.sev.long - LFC > 0.5 - LOWER IN OA
    enrich.sev.long %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "severe_long_upregulated"  =
    # enrich.sev.long - LFC < (-0.5) - HIGHER IN OA
    enrich.sev.long %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector()

  ) 

# Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs 
write_delim(x = ldply(DEGs.updown.4David, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", 
            file = "../results/deseq2/DEGs-updown-4David.txt")
```

- Navigate to DAVID on a browser - https://david.ncifcrf.gov/tools.jsp
- On the left you'll see "Upload Gene List"
- For Step 1, use the "B:Choose from a file" and upload the DEGs-updown-4David.txt file that was created in the last code chunk.  
- For the Step 2 dropdown list, select "UNIPROT_ACCESSION"  
- For Step 3, select "Gene List"  
- Select "Submit List". 
- You then need to upload a background list of genes. This is important. If you don't, the analysis will assume that you're comparing the DEG lists to the human genome (You'll see "Current Background: Homo sapiens"). Instead, we need to upload the list of all genes that we analyzed. To do this, use the below code chunk to copy Uniprot IDs for all genes assessed and paste into the Upload space on DAVID.  

```{r}
# Copy the BACKGROUND - all genes submitted to DESeq2 analysis 
enrich.background %>% write_clip(allow_non_interactive = TRUE)
```

- Paste into the white box titled "A:Paste a list" in the "Upload" tab of DAVID. 
- In Step2 select "UNIPROT_ACCESSION"  
- For Step 3 select Background 
- Submit List 
- Select the "Functional Annotation Tool" then the "Functional Annotation Chart" on the next page. A table will pop up.  On top-right of table right click "Download File", then "Save Link As..." and save to file.  I used the filename "../results/deseq2/GO_filenames_updown.txt". 

#### Read in enriched biological functions that were up/down-regulated in treatments

```{r, warning=F, message=F}
filenames_updown <- read_delim(file="../results/deseq2/GO_filenames_updown.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list_updown <- vector(mode = "list", length = nrow(filenames_updown))
names(file_list_updown) <- c(filenames_updown$gene_set)

for (i in 1:nrow(filenames_updown)) {
    file_list_updown[[i]] <- data.frame(read_delim(file=paste("../results/deseq2/", filenames_updown[i,"filename"], sep=""), delim = "\t"))}

david_updown <- bind_rows(file_list_updown, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~") %>%
    separate(gene_set, sep = "_", into = c("OA", "duration", "response"), remove = F) %>%
    mutate(treatment= gsub("_upregulated|_downregulated", "", gene_set) %>% factor(., ordered = TRUE, levels=c("moderate_short","moderate_long", "severe_short", "severe_long")),
           OA=as.factor(OA), duration=as.factor(duration), response=as.factor(response)) %>% 
  tidyr::complete(OA, duration, response, category) # add rows for gene_sets that are missing from dataframe

save(david_updown, file="../results/deseq2/david_updown")
```

### Plot enriched terms in UPREGULATED genes

```{r}
# Bubble plot, Upregulated processes
#pdf(file="../results/GO-enrichment/Final-DAVID/Upregulated-GO-bubble.pdf", height = 9, width = 7.5)
david_updown %>% 
#  filter(count>=3) %>% 
  filter(p_value<0.05) %>% filter(count>=3) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(treatment, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david_updown %>% 
              dplyr::select(treatment, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("treatment", "response", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%
  filter(response=="upregulated") %>% 

  ungroup() %>% arrange(treatment, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=treatment, col=treatment)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Treatment",
                     values=c("moderate_short"=lighten("darkgreen", 0.25),
                              "moderate_long"=lighten("yellow3", 0.25),
                              "severe_short"=lighten("sienna2"),
                              "severe_long"=lighten("firebrick4")),
                     labels=c("moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure"),
                     guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes\nin UPREGULATED genes")

```

Enriched terms in DOWNREGULATED genes

```{r}
# Bubble plot, Upregulated processes
#pdf(file="../results/GO-enrichment/Final-DAVID/Upregulated-GO-bubble.pdf", height = 9, width = 7.5)
david_updown %>% 
#  filter(count>=3) %>% 
  filter(p_value<0.05) %>% filter(count>=3) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
  #filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(treatment, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david_updown %>% 
              dplyr::select(treatment, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("treatment", "response", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%
  filter(response=="downregulated") %>% 

  ungroup() %>% arrange(treatment, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=treatment, col=treatment)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Treatment",
                     values=c("moderate_short"=lighten("darkgreen", 0.25),
                              "moderate_long"=lighten("yellow3", 0.25),
                              "severe_short"=lighten("sienna2"),
                              "severe_long"=lighten("firebrick4")),
                     labels=c("moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure"),
                     guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes\nin DOWNREGULATED genes")

```


## PCA FOR AMSS PRESENTATION 

```{r}
range(tab.expr.sva$PC1)
range(tab.expr.sva$PC2)
```

```{r}
# Keep axis limits, only show short-term treatments 

ggscatter(tab.expr.sva %>% filter(treatment %!in% c("moderate_long", "severe_long")), 
          x="PC1", y="PC2", col="treatment", shape="OA", size=3, alpha=0.85, 
          ellipse = F, star.plot = T) +
  theme_minimal() + #ggtitle("Gene expression PC1xPC2") + 
  xlim(c(-22,25)) + ylim(c(-16,30)) + 
  xlab(paste("PC1 (", round(pc.percent.sva[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.sva[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_fill_manual(guide="none", name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(values=c("ambient"=19, "moderate"=15, "severe"=17),
                     labels=c("Control", "Moderate OA", "Severe OA"))


ggscatter(tab.expr.sva, 
          x="PC1", y="PC2", col="treatment", shape="OA", size=3, alpha=0.85, 
          ellipse = F, star.plot = T) +
  xlim(c(-22,25)) + ylim(c(-16,30)) + 
  theme_minimal() + #ggtitle("Gene expression PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.sva[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.sva[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_fill_manual(guide="none", name="Treatment",
                     values=color_scale,
                     labels=c("ambient"= "Control",
                              "moderate_short"="pH 7.8, Short Exposure",
                              "moderate_long"="pH 7.8, Long Exposure",
                              "severe_short"="pH 7.5, Short Exposure",
                              "severe_long"="pH 7.5, Long Exposure")) +
  scale_shape_manual(values=c("ambient"=19, "moderate"=15, "severe"=17),
                     labels=c("Control", "Moderate OA", "Severe OA"))
```

















<!-- ## Unsupervised analyses  -->

<!-- #### Heat map with top 10% most variable genes, VSD-transformed counts -->

<!-- ```{r} -->
<!-- # calculate CV for each gene across all samples  -->
<!-- temp <-  assay(vsd.treatment) %>% as.data.frame() %>% -->
<!--   rownames_to_column(var = "gene") %>% -->
<!--   rowwise() %>%  -->
<!--   dplyr::mutate( -->
<!--      sd = sd(c_across(-gene)), -->
<!--      mean = mean(c_across(-gene)), -->
<!--      cv = sd/mean) %>% -->
<!--   #column_to_rownames("gene") %>% -->
<!--   arrange(dplyr::desc(cv)) -->

<!-- # subset top 10% of most variable genes  -->
<!-- most.variable <- temp %>% -->
<!--   head(n=round(0.1*nrow(assay(vsd.treatment)))) %>% -->
<!--   select(gene) %>% unlist() %>% as.vector() -->

<!-- # Generate heat map / cluster those top 10% most variable genes - any pattern in the clustering?  -->
<!-- cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[,most.variable],  -->
<!--          Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column", -->
<!--                      show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE, -->
<!--          annotation_colors = list(treatment= -->
<!--           c(`ambient`= "navyblue",  -->
<!--             `moderate_short`="darkgreen", `moderate_long`="yellow3", -->
<!--             `severe_short`="sienna2", `severe_long`="firebrick4")), -->
<!--          annotation_row=as.data.frame(colData(vsd.treatment)[,c("treatment"), drop=FALSE]), -->
<!--          main = "gene counts - top 10% most variable\n(VSD-transformed)") -->
<!-- ``` -->

<!-- ### Hierarchical clustering -->

<!-- NOTE: this tree was generated by the pheatmap function in the previous code chunk -->

<!-- ```{r} -->
<!-- # Just plot hierarchical cluster, color-code sample by treatment, or by tank number -->
<!-- cluster.dendo <- as.dendrogram(cluster.most.variable$tree_row)  -->

<!-- ##bare bones dendogram -->
<!-- #plot(cluster.dendo) -->

<!-- # #A sub tree - looking at branches -->
<!-- # par(cex = 1) -->
<!-- # plot(cluster.dendo[[1]], horiz = TRUE) -->

<!-- order.dendrogram(cluster.dendo) # to find out the order of the sample labels  -->

<!-- # Create a dataframe ordered the same as the dendogram for annotation purposes  -->
<!-- dendo.df <- sample.info[match(cluster.most.variable$tree_row$labels, sample.info$sampleID),c("sampleID", "treatment")] -->
<!-- all(dendo.df$sample == cluster.most.variable$tree_row$labels) # Should == TRUE -->

<!-- # Create color coding vectors  -->
<!-- colorCodes.treat <- c(`ambient`= "navyblue",  -->
<!--             `moderate_short`="darkgreen", `moderate_long`="yellow3", -->
<!--             `severe_short`="sienna2", `severe_long`="firebrick4") -->

<!-- # Plot dendogram color-code by treatment -->
<!-- labels(cluster.dendo) <- dendo.df$sampleID[order.dendrogram(cluster.dendo)] -->
<!-- labels_colors(cluster.dendo) <- colorCodes.treat[as.character(dendo.df$treatment[order.dendrogram(cluster.dendo)])] -->
<!-- plot(cluster.dendo, xlab="Sample Number", cex.lab=0.9) -->
<!-- legend("topright", title = "Treatment", bty="n", legend = -->
<!--          c("ambient","moderate_short","moderate_long","severe_short","severe_long"),  -->
<!--        fill = c("navyblue","darkgreen","yellow3","sienna2", "firebrick4")) -->
<!-- ``` -->

<!-- ## Library summary stats -->

<!-- #### No. of fragments per sample?  -->

<!-- ```{r} -->
<!-- data.frame(rowSums(counts.ts)) %>%  -->
<!--                   dplyr::rename(read.total = 1) %>%  -->
<!--                   rownames_to_column(var="sampleID") %>% #summary()  -->
<!-- dplyr::summarise(mean=round(mean(read.total, na.rm=T)),  -->
<!--           sd=round(sd(read.total, na.rm=T)),  -->
<!--           se=round(sd/sqrt(length(sample))), -->
<!--           median=median(read.total),  -->
<!--           min=min(read.total),  -->
<!--           max=max(read.total)) %>% t() %>% as.data.frame() %>%  -->
<!--   dplyr::mutate(V1=prettyNum(V1, big.mark = ",")) %>%  -->
<!--   dplyr::rename("fragments.per.sample"="V1") #use this to average across all samples  -->
<!-- ``` -->

<!-- ##### Statistically test whether no. of fragments / treatment differs  -->

<!-- ```{r} -->
<!-- hist(log(fragments$read.total)) -->
<!-- shapiro.test(log(fragments$read.total)) -->
<!-- summary(aov(log(read.total) ~ treatment, fragments)) -->
<!-- TukeyHSD(aov(log(read.total) ~ treatment, fragments)) -->

<!-- anova(lm(log(read.total) ~ OA, fragments)) -->
<!-- anova(lm(log(read.total) ~ duration, fragments)) -->
<!-- anova(lm(log(read.total) ~ OA*duration, fragments)) -->

<!-- ggplot(fragments, aes(x=OA, y=read.total)) + geom_boxplot() + theme_minimal() -->
<!-- ggplot(fragments, aes(x=duration, y=read.total)) + geom_boxplot() + theme_minimal() -->
<!-- ggplot(fragments, aes(x=treatment, y=read.total)) + geom_boxplot() + theme_minimal() -->
<!-- ``` -->

<!-- #### How many samples per treatment, tank, etc? -->

<!-- ```{r} -->
<!-- fragments %>% group_by(treatment) %>% tally() -->
<!-- fragments %>% group_by(OA) %>% tally() -->
<!-- fragments %>% group_by(duration) %>% tally() -->
<!-- fragments %>% group_by(treatment) %>% tally() #%>% arrange(treatment, tank) -->
<!-- ``` -->

<!-- #### How many genes total in the filtered count matrix? -->

<!-- ```{r} -->
<!-- prettyNum(ncol(counts.ts),big.mark = ",") -->
<!-- ``` -->

<!-- #### What's the average no. of genes per sample, etc.?  -->

<!-- ```{r} -->
<!-- data.frame(rowSums(counts.ts != 0)) %>%  -->
<!--                   dplyr::rename(count.total = 1) %>%  -->
<!--                   rownames_to_column(var="sampleID") %>%  -->
<!--   summarise(mean=round(mean(count.total, na.rm=T)),  -->
<!--             sd=round(sd(count.total, na.rm=T)),  -->
<!--             se=round(sd/sqrt(length(sample))), -->
<!--             median=median(count.total, na.rm=T), -->
<!--             min=min(count.total, na.rm=T),  -->
<!--             max=max(count.total, na.rm=T)) %>% t() %>% as.data.frame() %>%  -->
<!--   mutate(V1=prettyNum(V1, big.mark = ",")) %>%  -->
<!--   dplyr::rename("genes.per.sample"="V1") -->
<!-- ``` -->

<!-- #### Total no. of genes identified in each sample -->

<!-- ```{r} -->
<!-- #ggplotly( -->
<!-- ggplot(data = data.frame(rowSums(counts.t != 0, na.rm=TRUE)) %>%  -->
<!--                   dplyr::rename(count.total = 1) %>%  -->
<!--                   rownames_to_column(var="sampleID")) + -->
<!--            geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total # genes by sample") +  -->
<!--              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())#)  -->
<!-- ``` -->

<!-- ### Prep data for analyses -->

<!-- #### Merge sample key info to count data, then sort, and generate heat map for initial inspection by treatment  -->

<!-- ```{r} -->
<!-- # IMPORTANT NOTE: to use data for ALL genes, use object "counts.t". To remove low-frequency genes, use object "counts.ts" (recommended option) -->

<!-- # merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day  -->
<!-- counts.tk <- merge(x=sample.info[,c("sampleID", "OA", "duration", "treatment")], by.x="sampleID", y=counts.ts, by.y="row.names") %>%  -->
<!--   arrange(treatment)  %>% column_to_rownames(var="sampleID") %>% droplevels() -->
<!-- save(counts.tk, file = "../results/counts.tk") -->
<!-- head(counts.tk[1:24]) #check out results of merge/arrange -->
<!-- ``` -->
<!-- #### Reformat for DESeq, ensure correct sample order for  -->

<!-- ```{r} -->
<!-- # NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order. -->

<!-- all(rownames(counts.tk) == counts.tk[-1:-3] %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE'  -->
<!-- ``` -->
<!-- #### Round counts - if using counts which counted fractionally (e.g. salmon), need to round to the nearest integer for DESeq2.  -->

<!-- ```{r} -->
<!-- #counts.tk <- counts.tk %>% mutate_if(is.numeric, round) -->
<!-- counts.tk[ , sapply(counts.tk, is.numeric)] <- round(counts.tk[ , sapply(counts.tk, is.numeric)]) -->

<!-- ``` -->

<!-- ### perMANOVA - multivariate analysis of variance -->

<!-- ```{r} -->
<!-- # Effect of temperature -->
<!-- vsd.t <- t(assay(vsd.treatment)) %>% as.data.frame() -->
<!-- perm <- vegan::adonis2(vsd.t ~ treatment, data=left_join(data.frame(sample=colnames(txi.gen$counts)), sample.info)%>%  -->
<!--             filter(sample!="sample_45") %>% column_to_rownames("sampleID")%>%dplyr::select(OA,treatment), permutations = 1000, method="bray") -->
<!-- perm #interesting- it found a difference -->

<!-- # Pairwise comparisons -->
<!-- vsd.t <- left_join(data.frame(sample=colnames(txi.gen$counts)), sample.info)%>%  -->
<!--             filter(sample!="sample_45") %>%dplyr::select(sample, OA,treatment) %>%  -->
<!--           left_join(t(assay(vsd.treatment)) %>% as.data.frame() %>% rownames_to_column("sampleID"),  -->
<!--                     by="sampleID")  -->

<!-- # Clustering by treatment?  -->
<!-- pairwise.adonis(vsd.t[,-1:-3], vsd.t$treatment, perm = 1000) %>% arrange(p.adjusted) %>% -->
<!--   filter(p.adjusted<0.05) %>% dplyr::select(pairs, p.adjusted) -->

<!-- # Clustering by temp x ph temperature? Yes! temperatures that differ are below.  -->
<!-- pairwise.adonis(vsd.t[,-1:-3], vsd.t$OA, perm = 1000) %>% arrange(p.adjusted) %>% -->
<!--   filter(p.adjusted<0.05) %>% dplyr::select(pairs, p.adjusted) -->
<!-- ``` -->

<!-- ### Data may not be linear, run NMDS instead  -->

<!-- ```{r} -->
<!-- library(vegan) -->

<!-- # expr: rows = samples, columns = genes -->
<!-- # Use a distance measure appropriate for gene expression -->
<!-- # Euclidean is fine for transformed expression values -->
<!-- dist_mat <- dist(t(assay(vsd.treatment)), method = "euclidean") -->

<!-- # Run NMDS -->
<!-- set.seed(123) # for reproducibility -->
<!-- nmds <- metaMDS(dist_mat, k = 3, trymax = 100) -->

<!-- # View stress (lower is better, <0.2 is usually acceptable) -->
<!-- nmds$stress  -->

<!-- # Extract NMDS coordinates -->
<!-- nmds_points <- as.data.frame(nmds$points) -->
<!-- nmds_points$Sample <- rownames(nmds_points) -->
<!-- nmds_points <- merge(nmds_points, sample.info, by.x = "sampleID", by.y = "sampleID") -->

<!-- # Plot with ggplot2 -->
<!-- ggplot(nmds_points, aes(x = MDS1, y = MDS2, color = treatment)) + -->
<!--   geom_point(size = 3) + -->
<!--   theme_minimal() + -->
<!--   labs(title = "NMDS of global gene expression", -->
<!--        x = "NMDS1", y = "NMDS2") + -->
<!--       scale_color_manual(name="Treatment", -->
<!--                      values=color_scale, -->
<!--                      labels=c("ambient"= "Control", -->
<!--                               "moderate_short"="pH 7.8, Short Exposure", -->
<!--                               "moderate_long"="pH 7.8, Long Exposure", -->
<!--                               "severe_short"="pH 7.5, Short Exposure", -->
<!--                               "severe_long"="pH 7.5, Long Exposure"))  -->

<!-- dca_res <- decorana(t(assay(vsd.treatment))) -->
<!-- scores_df <- as.data.frame(scores(dca_res, display = "sites"))  # site = sample -->
<!-- scores_df$Sample <- rownames(scores_df) -->
<!-- scores_df <- merge(scores_df, sample.info,  -->
<!--                    by.x = "sampleID", by.y = "sampleID") -->
<!-- ggplot(scores_df, aes(x = DCA1, y = DCA2, color = treatment)) + -->
<!--   geom_point(size = 3) + -->
<!--   theme_minimal() + -->
<!--   labs(title = "Detrended Correspondence Analysis of gene expression", -->
<!--        x = "DCA1", y = "DCA2") + -->
<!--           scale_color_manual(name="Treatment", -->
<!--                      values=color_scale, -->
<!--                      labels=c("ambient"= "Control", -->
<!--                               "moderate_short"="pH 7.8, Short Exposure", -->
<!--                               "moderate_long"="pH 7.8, Long Exposure", -->
<!--                               "severe_short"="pH 7.5, Short Exposure", -->
<!--                               "severe_long"="pH 7.5, Long Exposure"))  -->

<!-- ``` -->
